{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment II\n",
    "* Leoni Mota Loris"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* At first, Load the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import namedtuple\n",
    "\n",
    "Atom = namedtuple('Atom', ['name', 'element', 'aminoacid', 'x', 'y', 'z', 'h', 'residue_seq_number'])\n",
    "\n",
    "Atom.parse_name = lambda string: string[12:16].strip()\n",
    "Atom.parse_element = lambda string: string[12] if len(string.strip()) < 78 \\\n",
    "                     else string[76:78].strip()\n",
    "Atom.parse_aminoacid = lambda string: string[17:20].strip()\n",
    "Atom.parse_x = lambda string: float(string[30:38].strip())\n",
    "Atom.parse_y = lambda string: float(string[38:46].strip())\n",
    "Atom.parse_z = lambda string: float(string[46:54].strip())\n",
    "Atom.parse_residue_seq_number = lambda string: int(string[23:30].strip())\n",
    "\n",
    "Atom.parse = lambda string: Atom(\n",
    "    Atom.parse_name(string), Atom.parse_element(string), Atom.parse_aminoacid(string),\n",
    "    Atom.parse_x(string), Atom.parse_y(string), Atom.parse_z(string), 1,Atom.parse_residue_seq_number(string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_molecule(path):\n",
    "    atoms = []\n",
    "    with open(path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('ATOM'):\n",
    "                try:\n",
    "                    atoms.append(Atom.parse(line))\n",
    "                except Exception as e:\n",
    "                    print(e)\n",
    "    return pd.DataFrame(atoms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>17.047</td>\n",
       "      <td>14.099</td>\n",
       "      <td>3.625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>16.967</td>\n",
       "      <td>12.784</td>\n",
       "      <td>4.338</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.685</td>\n",
       "      <td>12.755</td>\n",
       "      <td>5.133</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.268</td>\n",
       "      <td>13.825</td>\n",
       "      <td>5.594</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.170</td>\n",
       "      <td>12.703</td>\n",
       "      <td>5.337</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>19.334</td>\n",
       "      <td>12.829</td>\n",
       "      <td>4.463</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.150</td>\n",
       "      <td>11.546</td>\n",
       "      <td>6.304</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.115</td>\n",
       "      <td>11.555</td>\n",
       "      <td>5.265</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>13.856</td>\n",
       "      <td>11.469</td>\n",
       "      <td>6.066</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>14.164</td>\n",
       "      <td>10.785</td>\n",
       "      <td>7.379</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name element aminoacid       x       y      z  h  residue_seq_number\n",
       "0    N       N       THR  17.047  14.099  3.625  1                   1\n",
       "1   CA       C       THR  16.967  12.784  4.338  1                   1\n",
       "2    C       C       THR  15.685  12.755  5.133  1                   1\n",
       "3    O       O       THR  15.268  13.825  5.594  1                   1\n",
       "4   CB       C       THR  18.170  12.703  5.337  1                   1\n",
       "5  OG1       O       THR  19.334  12.829  4.463  1                   1\n",
       "6  CG2       C       THR  18.150  11.546  6.304  1                   1\n",
       "7    N       N       THR  15.115  11.555  5.265  1                   2\n",
       "8   CA       C       THR  13.856  11.469  6.066  1                   2\n",
       "9    C       C       THR  14.164  10.785  7.379  1                   2"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crambin = read_molecule('assets/1crn.pdb')\n",
    "crambin[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* I) First we need to map the sequence of characters as in:\n",
    "\n",
    "**TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN**\n",
    "to a sequence of amino acids \n",
    "\n",
    "* II) And then, infer the position of each atom of the amino acid from a standard reference ($\\alpha$ carbon) \n",
    "\n",
    "\n",
    "* III) Infer the position of multiple amino acids based on the peptide bond."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### I) map the sequence of characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_mapping = {'T': 'THR', 'C': 'CYS', 'P': 'PRO', 'S': 'SER', 'I': 'ILE',\n",
    "                     'V': 'VAL', 'A': 'ALA', 'R': 'ARG', 'N': 'ASN', 'F': 'PHE',\n",
    "                     'L': 'LEU', 'G': 'GLY', 'E': 'GLU', 'D': 'ASP', 'Y': 'TYR'}\n",
    "    \n",
    "def map_to_aminoacids(sequence):\n",
    "    return list(map(lambda char: aminoacid_mapping[char], sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_sequence = 'TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids_names = map_to_aminoacids(aminoacid_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### II) infer the position of each atom of the amino acid. For this part, it is not well defined on how to infer the positions.\n",
    "\n",
    "\n",
    "* We could map this sequence to known sequences of the PDB and hope it is a known protein. This is called \n",
    "**Homology Modelling** (**SWISS-MODEL** is an example).\n",
    "\n",
    "\n",
    "* *Or*, we could predict the protein structure using an any number of softwares such as **I-TASSER (Iterative Threading ASSEmbly Refinement)**\n",
    "\n",
    "\n",
    "We're are going to use both and compare the results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I-TASSER:\n",
    "![assets/TASSER.png](assets/TASSER.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* But, if we're not using something already pre-defined, we should get the pdb file for each of those amino acids and infer the position of each amino acid by the peptide bond.\n",
    "\n",
    "*Obs.:* PDB's for each amino acid were downloaded here:\n",
    "https://www.nyu.edu/pages/mathmol/library/life/life1.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_aminoacids():\n",
    "    return list(map(lambda aminoacid_name: read_molecule('assets/' + aminoacid_name + '.pdb'), aminoacids_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids = load_aminoacids()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>0.080</td>\n",
       "      <td>-0.083</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.539</td>\n",
       "      <td>-0.033</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.032</td>\n",
       "      <td>1.394</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.268</td>\n",
       "      <td>2.339</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.116</td>\n",
       "      <td>-0.818</td>\n",
       "      <td>-1.226</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.743</td>\n",
       "      <td>-2.188</td>\n",
       "      <td>-1.162</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.654</td>\n",
       "      <td>-0.842</td>\n",
       "      <td>-1.361</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1H</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.523</td>\n",
       "      <td>0.813</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2H</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.461</td>\n",
       "      <td>-1.018</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.904</td>\n",
       "      <td>-0.508</td>\n",
       "      <td>0.929</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.691</td>\n",
       "      <td>-0.378</td>\n",
       "      <td>-2.154</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.066</td>\n",
       "      <td>-2.592</td>\n",
       "      <td>-1.973</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1HG2</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>3.986</td>\n",
       "      <td>-1.433</td>\n",
       "      <td>-2.235</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2HG2</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.083</td>\n",
       "      <td>0.168</td>\n",
       "      <td>-1.502</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3HG2</td>\n",
       "      <td>3</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.138</td>\n",
       "      <td>-1.279</td>\n",
       "      <td>-0.467</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.452</td>\n",
       "      <td>1.567</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.911</td>\n",
       "      <td>0.645</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name element aminoacid      x      y      z  h  residue_seq_number\n",
       "0      N               THR  0.080 -0.083  0.000  1                   1\n",
       "1     CA               THR  1.539 -0.033  0.000  1                   1\n",
       "2      C               THR  2.032  1.394  0.000  1                   1\n",
       "3      O               THR  1.268  2.339  0.000  1                   1\n",
       "4     CB               THR  2.116 -0.818 -1.226  1                   1\n",
       "5    OG1               THR  1.743 -2.188 -1.162  1                   1\n",
       "6    CG2               THR  3.654 -0.842 -1.361  1                   1\n",
       "7     1H       1       THR -0.523  0.813  0.000  1                   1\n",
       "8     2H       2       THR -0.461 -1.018  0.000  1                   1\n",
       "9     HA               THR  1.904 -0.508  0.929  1                   1\n",
       "10    HB               THR  1.691 -0.378 -2.154  1                   1\n",
       "11   HG1               THR  2.066 -2.592 -1.973  1                   1\n",
       "12  1HG2       1       THR  3.986 -1.433 -2.235  1                   1\n",
       "13  2HG2       2       THR  4.083  0.168 -1.502  1                   1\n",
       "14  3HG2       3       THR  4.138 -1.279 -0.467  1                   1\n",
       "15    OC               THR  3.452  1.567  0.000  1                   1\n",
       "16    HC               THR  3.911  0.645  0.000  1                   1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(aminoacids[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### III) Infer the position of multiple amino acids based on the peptide bond.\n",
    "\n",
    "\n",
    "* After the nucleotilic attack the resonance delocalization of electrons in the peptide bond forces the creation of a plane where the atoms will be located (amide plan)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* First, We define how translations and rotations are made.\n",
    "\n",
    "\n",
    "* Homogeneous coordinates were used throughout this work.\n",
    "![assets/rotation.png](assets/rotation.png)\n",
    "![assets/translation.png](assets/translation.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def translation_matrix(delta_x, delta_y, delta_z, ref_x=.0, ref_y=.0, ref_z=.0):\n",
    "    T = np.eye(4)\n",
    "    T[0, -1], T[1, -1], T[2, -1] = delta_x - ref_x, delta_y - ref_y, delta_z - ref_z\n",
    "    return T\n",
    "\n",
    "def rotation_matrix(x_rotation_rad, y_rotation_rad, z_rotation_rad):\n",
    "    Rx, Ry, Rz = np.eye(4), np.eye(4), np.eye(4)\n",
    "\n",
    "    Rx[1, 1], Rx[1, 2], Rx[2, 1], Rx[2, 2] =\\\n",
    "        np.cos(x_rotation_rad), -np.sin(x_rotation_rad), np.sin(x_rotation_rad), np.cos(x_rotation_rad)\n",
    "\n",
    "    Ry[0, 0], Ry[0, 2], Ry[2, 0], Rx[2, 2] =\\\n",
    "        np.cos(y_rotation_rad), np.sin(y_rotation_rad), -np.sin(y_rotation_rad), np.cos(y_rotation_rad)\n",
    "\n",
    "    Rz[0, 0], Rx[0, 1], Rx[1, 0], Rx[1, 1] =\\\n",
    "        np.cos(z_rotation_rad), -np.sin(z_rotation_rad), np.sin(z_rotation_rad), np.cos(z_rotation_rad)\n",
    "\n",
    "    return np.dot(np.dot(Rx, Ry), Rz)\n",
    "\n",
    "def apply_transformation(transformation, df):\n",
    "    df[['x', 'y', 'z', 'h']] = transformation.dot(df[['x', 'y', 'z', 'h']].T).T\n",
    "\n",
    "def apply_rotation(transformation, df, atom_reference='N'):\n",
    "    position = df[df.name == atom_reference]\n",
    "    apply_transformation(translation_matrix(-position.x, -position.y, -position.z), df)\n",
    "    apply_transformation(transformation, df)\n",
    "    apply_transformation(translation_matrix(position.x, position.y, position.z), df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# rotations_z = np.arange(0, 190, 10) * np.pi / 180\n",
    "\n",
    "# for rotation_z in rotations_z:\n",
    "#     a = aminoacids[0].copy()\n",
    "#     apply_rotation(rotation_matrix(rotation_z, 0, 0), a)    \n",
    "#     plt.scatter(a[:10].x, a[:10].z, c=np.arange(10), s=70)\n",
    "#     plt.xlabel('X!!!!')\n",
    "#     print(rotation_z *180 / np.pi)\n",
    "#     plt.xlim(-5, 5)\n",
    "#     plt.ylim(-5, 5)\n",
    "#     plt.grid()\n",
    "#     plt.show()\n",
    "#     plt.pause(0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dehydration(aminoacid1, aminoacid2):\n",
    "    aminoacid1 = aminoacid1[~aminoacid1.name.isin(['OC', 'HC'])]    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some utility functions for calculating the euclidean distance and the angle between two vectors:\n",
    "\n",
    "\n",
    "$$distance = \\sqrt{|(A - B)|}$$\n",
    "\n",
    "\n",
    "$$cos(\\alpha) = \\frac{A \\cdot B}{|A||B|}$$ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [],
   "source": [
    "def euclidean_distance(p1, p2):\n",
    "    return np.sqrt(np.sum((p1 - p2) ** 2))\n",
    "\n",
    "def l2_norm(p):\n",
    "    return np.sqrt(np.sum(p ** 2))\n",
    "\n",
    "def get_angle_rad(a, b, normal_reference):\n",
    "    negative = np.all(np.dot(np.cross(a, b), normal_reference) < 0)\n",
    "    angle = np.arccos(((a).dot(b))/(l2_norm(a) * l2_norm(b))) \n",
    "    angle = angle if not np.isnan(angle) else .0\n",
    "    return angle if not negative else -angle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will consider the omega torsion angle of `Proline` to be close to zero degrees for the `cis` configuration, or most often and 180 degrees for the trans configuration.\n",
    "\n",
    "\n",
    "IMAGEM AQUI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_omega(current_aminoacid):\n",
    "    return 0 if 'PRO' in current_aminoacid.aminoacid.values else 180\n",
    "\n",
    "def get_desired_OCN_angle(current_aminoacid):\n",
    "    return 123 if 'PRO' in current_aminoacid.aminoacid.values else 123.5\n",
    "\n",
    "def get_desired_CNCa_angle(current_aminoacid):\n",
    "    return 126 if 'PRO' in current_aminoacid.aminoacid.values else 122\n",
    "\n",
    "def get_desired_bond_length(current_aminoacid):\n",
    "    return 1.32 if 'PRO' in current_aminoacid.aminoacid.values else 1.33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_nitrogen_location(aminoacid):\n",
    "    alpha_carbon = aminoacid[aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    carboxyl_carbon = aminoacid[aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    oxygen_double_bonded = aminoacid[aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "\n",
    "    CO = oxygen_double_bonded - carboxyl_carbon\n",
    "    CA = alpha_carbon - carboxyl_carbon\n",
    "\n",
    "    plane_normal = np.cross(CO, CA)\n",
    "    plane_vector_basis = np.array([\n",
    "        np.cross(CO, plane_normal) / l2_norm(np.cross(CO, plane_normal)), \n",
    "        CO / l2_norm(CO), \n",
    "        plane_normal / l2_norm(plane_normal)]).T\n",
    "\n",
    "    CO_new_norm = CO * get_desired_bond_length(aminoacid) / l2_norm(CO)\n",
    "    CO_new_norm_plane = np.linalg.solve(plane_vector_basis, CO_new_norm)[:2]\n",
    "    \n",
    "    rotation_angle = np.radians(get_desired_OCN_angle(aminoacid))\n",
    "    rotation_angle_matrix = np.array(\n",
    "        [[np.cos(rotation_angle), -np.sin(rotation_angle)], \n",
    "         [np.sin(rotation_angle), np.cos(rotation_angle)]])\n",
    "\n",
    "    CN_plane = CO_new_norm_plane @ rotation_angle_matrix\n",
    "    CN = np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array([CN_plane[0], CN_plane[1], 0]))\n",
    "    return CN + carboxyl_carbon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C'):\n",
    "    if terminal == 'C':\n",
    "        alpha_carbon = previous_aminoacid[previous_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        oxygen_double_bonded = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C1 = oxygen_double_bonded - carboxyl_carbon\n",
    "        C2 = alpha_carbon - carboxyl_carbon\n",
    "    if terminal == 'N':\n",
    "        last_carbon = current_aminoacid[current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        nitrogen_amide = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C2 = last_carbon - nitrogen_amide\n",
    "        C1 = carboxyl_carbon - nitrogen_amide\n",
    "    return np.cross(C1, C2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    desired_nitrogen_location = get_desired_nitrogen_location(previous_aminoacid)\n",
    "    current_nitrogen_location = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        \n",
    "    required_tranlation = desired_nitrogen_location - current_nitrogen_location\n",
    "    \n",
    "    required_tranlation_matrix = translation_matrix(required_tranlation[0], required_tranlation[1], required_tranlation[2])\n",
    "    apply_transformation(required_tranlation_matrix, current_aminoacid)\n",
    "    \n",
    "def rotate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    omega = get_desired_omega(current_aminoacid)\n",
    "    previous_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C')\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_x = get_angle_rad(current_normal[[1, 2]], previous_normal[[1, 2]], \n",
    "                                     -previous_normal[[1, 2]])\n",
    "    apply_rotation(rotation_matrix(rotation_angle_x, 0, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_y = get_angle_rad(current_normal[[0, 2]], previous_normal[[0, 2]], \n",
    "                                     -previous_normal[[0, 2]])\n",
    "    apply_rotation(rotation_matrix(0, rotation_angle_y, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_z = get_angle_rad(current_normal[[0, 1]], previous_normal[[0, 1]],\n",
    "                                     -previous_normal[[0, 1]])\n",
    "    apply_rotation(rotation_matrix(0, 0, rotation_angle_z), current_aminoacid)\n",
    "\n",
    "def rotate_for_CNCa_angle(previous_aminoacid, current_aminoacid):\n",
    "    carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    nitrogen_amide = current_aminoacid[current_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    alpha_carbon = current_aminoacid[current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "\n",
    "    NC = carboxyl_carbon - nitrogen_amide\n",
    "    NCa = alpha_carbon - nitrogen_amide\n",
    "    print('NCa', NCa)\n",
    "    plane_normal = np.cross(NCa, NC)\n",
    "    plane_vector_basis = np.array([\n",
    "        NC / l2_norm(NC),\n",
    "        np.cross(NC, plane_normal) / l2_norm(np.cross(NC, plane_normal)), \n",
    "        plane_normal / l2_norm(plane_normal)]).T\n",
    "\n",
    "    current_CNCa_angle_rad = get_angle_rad(NC, NCa, -plane_normal)\n",
    "    desired_CNCa_angle_rad = np.radians(get_desired_CNCa_angle(current_aminoacid))\n",
    "    \n",
    "    needed_rotation_rad = desired_CNCa_angle_rad - current_CNCa_angle_rad\n",
    "    print('needed_rotation', np.degrees(needed_rotation_rad))\n",
    "    \n",
    "    new_current_aminoacid = current_aminoacid.copy()\n",
    "    for _, atom in new_current_aminoacid.iterrows():\n",
    "        atom.loc[['x', 'y', 'z']] =\\\n",
    "            np.linalg.solve(plane_vector_basis, np.array(atom.loc[['x', 'y', 'z']].values.reshape(-1,), dtype='float'))\n",
    "    \n",
    "    apply_rotation(rotation_matrix(0, 0, needed_rotation_rad), new_current_aminoacid)\n",
    "\n",
    "    for _, atom in new_current_aminoacid.iterrows():\n",
    "        atom.loc[['x', 'y', 'z']] =\\\n",
    "            np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array(atom.loc[['x', 'y', 'z']].values.reshape(-1,), dtype='float'))\n",
    "    \n",
    "    new_alpha_carbon = new_current_aminoacid[new_current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    NC = carboxyl_carbon - nitrogen_amide\n",
    "    NCa = new_alpha_carbon - nitrogen_amide\n",
    "    print('NCa', NCa)\n",
    "    print(np.degrees(get_angle_rad(NC, NCa, -plane_normal)))\n",
    "#     CN = np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array([CN_plane[0], CN_plane[1], 0]))\n",
    "#     return CN + carboxyl_carbon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_peptide_bond_length(previous_aminoacid, current_aminoacid):\n",
    "    a = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    b = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('euclidean_distance for the peptide bond between %s and %s: %.2f' % \n",
    "          (previous_aminoacid.aminoacid[0], current_aminoacid.aminoacid[0], euclidean_distance(a, b)))\n",
    "    \n",
    "def check_OCN_angle(previous_aminoacid, current_aminoacid):\n",
    "    c = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    n = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    o = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('O-C-N Angle for the peptide bond between %s and %s:        %.2f' % \n",
    "          (previous_aminoacid.aminoacid[0], current_aminoacid.aminoacid[0], abs(180*get_angle_rad(o-c, n-c, (o-c))/np.pi)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:22: RuntimeWarning: invalid value encountered in double_scalars\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "euclidean_distance for the peptide bond between THR and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and THR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between THR and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and CYS:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between CYS and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and CYS:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between CYS and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and PRO:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between PRO and SER: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and SER:        123.00\n",
      "\n",
      "euclidean_distance for the peptide bond between SER and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between SER and ILE:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ILE and VAL: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and VAL:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between VAL and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between VAL and ALA:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ALA and ARG: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and ARG:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ARG and SER: 1.33\n",
      "O-C-N Angle for the peptide bond between ARG and SER:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between SER and ASN: 1.33\n",
      "O-C-N Angle for the peptide bond between SER and ASN:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ASN and PHE: 1.33\n",
      "O-C-N Angle for the peptide bond between ASN and PHE:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between PHE and ASN: 1.33\n",
      "O-C-N Angle for the peptide bond between PHE and ASN:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ASN and VAL: 1.33\n",
      "O-C-N Angle for the peptide bond between ASN and VAL:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between VAL and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between VAL and CYS:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between CYS and ARG: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and ARG:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ARG and LEU: 1.33\n",
      "O-C-N Angle for the peptide bond between ARG and LEU:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between LEU and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between LEU and PRO:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between PRO and GLY: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLY:        123.00\n",
      "\n",
      "euclidean_distance for the peptide bond between GLY and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and THR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between THR and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and PRO:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between PRO and GLU: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLU:        123.00\n",
      "\n",
      "euclidean_distance for the peptide bond between GLU and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between GLU and ALA:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ALA and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and ILE:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ILE and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and CYS:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between CYS and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and ALA:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ALA and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and THR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between THR and TYR: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and TYR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between TYR and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between TYR and THR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between THR and GLY: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and GLY:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between GLY and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and CYS:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between CYS and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and ILE:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ILE and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and ILE:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ILE and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and ILE:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ILE and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and PRO:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between PRO and GLY: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLY:        123.00\n",
      "\n",
      "euclidean_distance for the peptide bond between GLY and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and ALA:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ALA and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and THR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between THR and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and THR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between THR and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and CYS:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between CYS and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and PRO:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between PRO and GLY: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLY:        123.00\n",
      "\n",
      "euclidean_distance for the peptide bond between GLY and ASP: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and ASP:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ASP and TYR: 1.33\n",
      "O-C-N Angle for the peptide bond between ASP and TYR:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between TYR and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between TYR and ALA:        123.50\n",
      "\n",
      "euclidean_distance for the peptide bond between ALA and ASN: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and ASN:        123.50\n",
      "\n"
     ]
    }
   ],
   "source": [
    "aminoacids_df = load_aminoacids()\n",
    "for i in range(1, len(aminoacids_df)):\n",
    "    current_aminoacid = aminoacids_df[i]\n",
    "    previous_aminoacid = aminoacids_df[i - 1]\n",
    "    \n",
    "    dehydration(previous_aminoacid, current_aminoacid)\n",
    "    translate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "    rotate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "#     rotate_for_CNCa_angle(previous_aminoacid, current_aminoacid)\n",
    "    \n",
    "    check_peptide_bond_length(previous_aminoacid, current_aminoacid)\n",
    "    check_OCN_angle(previous_aminoacid, current_aminoacid)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We have all the atoms aligned accordingly. So, let's create string rows for the pdb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_atom_to_file(atom_row, aminoacid_seq_number, fd):\n",
    "    global atom_seq_number\n",
    "    atom_seq_number += 1\n",
    "    if 'H' not in atom_row[0]:\n",
    "        atom_row.element = ''.join(filter(str.isalpha, atom_row[0]))[0]\n",
    "        template_row =\\\n",
    "        'ATOM    '+\\\n",
    "        ('%d'%atom_seq_number).rjust(3)+\\\n",
    "        '  '+\\\n",
    "        ('%s'%atom_row[0]).rjust(3)+\\\n",
    "        ' '+\\\n",
    "        ('%s'%atom_row.aminoacid).rjust(3)+\\\n",
    "        ' A'+\\\n",
    "        ('%d'%aminoacid_seq_number).rjust(4)+\\\n",
    "        '    '+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.y).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        '  1.00  5.99           '+\\\n",
    "        ('%s\\n'%atom_row.element)\n",
    "        f.write(template_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "metadata": {},
   "outputs": [],
   "source": [
    "atom_seq_number = 0\n",
    "\n",
    "with open('assets/final_inferred_molecule.pdb', 'w') as f:\n",
    "    for idx, aminoacid in enumerate(aminoacids_df):\n",
    "        for atom_row in aminoacid.iterrows():\n",
    "            write_atom_to_file(atom_row[1], idx, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## b) Receive PDB file and list $\\phi$ and $\\psi$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$ ψ  = Normal(N_{i}, C_{\\alpha i}, C_{i}) \\^ Normal(C_{\\alpha i}, C_{i}, N_{i+1})  $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$φ = Normal(C_{i-1}, N_{i}, C_{\\alpha i}) \\^ Normal(N_{i}, C_{\\alpha i}, C_{i}) $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![assets/phi_psi.png](assets/phi_psi.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_psi(ni, cai, ci, ni1):\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    normal_ci = np.cross((cai - ci), (ni1 - ci))\n",
    "    return np.degrees(get_angle_rad(normal_cai, normal_ci, -(ci-cai)))\n",
    "\n",
    "def compute_phi(ci_prev, ni, cai, ci):\n",
    "    normal_ni = np.cross((ci_prev - ni), (cai - ni))\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    return np.degrees(get_angle_rad(normal_ni, normal_cai, -(ni-cai)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_phy_psi(molecule):\n",
    "    molecule['phi'] = None\n",
    "    molecule['psi'] = None\n",
    "\n",
    "    data = []\n",
    "\n",
    "    for i in range(1, len(molecule.residue_seq_number.unique())):\n",
    "        aminoacid_i = molecule[molecule.residue_seq_number == (i)]\n",
    "        ni = aminoacid_i[aminoacid_i.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        cai = aminoacid_i[aminoacid_i.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        ci = aminoacid_i[aminoacid_i.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        angle_data = {'residue_name': aminoacid_i.aminoacid.values[0], 'phi': None, 'psi': None}\n",
    "        try:\n",
    "            aminoacid_prev = molecule[molecule.residue_seq_number == (i-1)]\n",
    "            ci_prev = aminoacid_prev[aminoacid_prev.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['phi'] = compute_phi(ci_prev, ni, cai, ci)\n",
    "        except: pass\n",
    "\n",
    "        try:\n",
    "            aminoacid_i1 = molecule[molecule.residue_seq_number == (i+1)]\n",
    "            ni1 = aminoacid_i1[aminoacid_i1.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['psi'] = compute_psi(ni, cai, ci, ni1)\n",
    "        except: pass\n",
    "\n",
    "        data.append(angle_data)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:22: RuntimeWarning: invalid value encountered in arccos\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>LEU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>8.537736e-07</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>8.537736e-07</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1.207418e-06</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1.207418e-06</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1.207418e-06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>8.537736e-07</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ASP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>8.537736e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             phi         psi residue_name\n",
       "0   1.800000e+02  180.000000          THR\n",
       "1   1.800000e+02  179.999999          CYS\n",
       "2   0.000000e+00    0.000000          CYS\n",
       "3   1.800000e+02  180.000000          PRO\n",
       "4   0.000000e+00  180.000000          SER\n",
       "5   1.800000e+02  179.999999          ILE\n",
       "6   1.800000e+02  180.000000          VAL\n",
       "7   1.800000e+02  180.000000          ALA\n",
       "8   0.000000e+00  179.999999          ARG\n",
       "9   1.800000e+02    0.000000          SER\n",
       "10  1.800000e+02  180.000000          ASN\n",
       "11  0.000000e+00    0.000000          PHE\n",
       "12  0.000000e+00  180.000000          ASN\n",
       "13  0.000000e+00  180.000000          VAL\n",
       "14  0.000000e+00    0.000000          CYS\n",
       "15  1.800000e+02  180.000000          ARG\n",
       "16  1.800000e+02  180.000000          LEU\n",
       "17  0.000000e+00    0.000000          PRO\n",
       "18  0.000000e+00    0.000000          GLY\n",
       "19  0.000000e+00  180.000000          THR\n",
       "20  1.800000e+02  179.999999          PRO\n",
       "21  8.537736e-07  180.000000          GLU\n",
       "22  8.537736e-07  180.000000          ALA\n",
       "23  1.207418e-06  180.000000          ILE\n",
       "24  0.000000e+00  180.000000          CYS\n",
       "25  0.000000e+00  180.000000          ALA\n",
       "26  0.000000e+00    0.000000          THR\n",
       "27  0.000000e+00  180.000000          TYR\n",
       "28  1.800000e+02  180.000000          THR\n",
       "29  1.800000e+02  180.000000          GLY\n",
       "30  1.800000e+02  180.000000          CYS\n",
       "31  1.207418e-06  179.999999          ILE\n",
       "32  1.800000e+02  179.999999          ILE\n",
       "33  0.000000e+00    0.000000          ILE\n",
       "34  1.800000e+02  180.000000          PRO\n",
       "35  0.000000e+00  180.000000          GLY\n",
       "36  1.800000e+02  180.000000          ALA\n",
       "37  1.800000e+02  180.000000          THR\n",
       "38  1.800000e+02  180.000000          THR\n",
       "39  1.207418e-06    0.000000          CYS\n",
       "40  0.000000e+00  180.000000          PRO\n",
       "41  8.537736e-07  180.000000          GLY\n",
       "42  1.800000e+02  180.000000          ASP\n",
       "43  0.000000e+00  180.000000          TYR\n",
       "44  1.800000e+02  180.000000          ALA\n",
       "45  8.537736e-07         NaN          ASN"
      ]
     },
     "execution_count": 464,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/final_inferred_molecule.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## c) Ramachandran "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>147.660026</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107.832573</td>\n",
       "      <td>144.346241</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>131.170389</td>\n",
       "      <td>133.291143</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>118.896523</td>\n",
       "      <td>151.202717</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>76.181506</td>\n",
       "      <td>-18.979386</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>157.924288</td>\n",
       "      <td>166.027161</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>63.579524</td>\n",
       "      <td>-42.100632</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>55.930329</td>\n",
       "      <td>-44.561332</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>61.364845</td>\n",
       "      <td>-43.779138</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>63.193412</td>\n",
       "      <td>-43.303486</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>61.208800</td>\n",
       "      <td>-42.430336</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>64.920400</td>\n",
       "      <td>-39.489513</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>59.102393</td>\n",
       "      <td>-47.211965</td>\n",
       "      <td>PHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>62.823308</td>\n",
       "      <td>-35.231846</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>69.239667</td>\n",
       "      <td>-41.218779</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>56.671614</td>\n",
       "      <td>-36.000832</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>77.068519</td>\n",
       "      <td>-16.126707</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>53.208796</td>\n",
       "      <td>-46.167676</td>\n",
       "      <td>LEU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>77.189679</td>\n",
       "      <td>-7.597741</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>-106.257207</td>\n",
       "      <td>7.311793</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>52.674731</td>\n",
       "      <td>136.344219</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>56.977153</td>\n",
       "      <td>146.617514</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>56.412840</td>\n",
       "      <td>-36.191815</td>\n",
       "      <td>GLU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>63.430344</td>\n",
       "      <td>-34.860660</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>74.766620</td>\n",
       "      <td>-37.894498</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>64.948033</td>\n",
       "      <td>-31.686198</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>62.036226</td>\n",
       "      <td>-54.031875</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>68.778289</td>\n",
       "      <td>-25.486562</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>67.592836</td>\n",
       "      <td>-36.300194</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>108.956767</td>\n",
       "      <td>-18.473152</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>-91.815951</td>\n",
       "      <td>-3.067675</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>69.515392</td>\n",
       "      <td>164.375577</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>129.761553</td>\n",
       "      <td>157.032715</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>111.563161</td>\n",
       "      <td>129.588363</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>124.130558</td>\n",
       "      <td>158.127320</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>78.150912</td>\n",
       "      <td>-24.351838</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>89.739760</td>\n",
       "      <td>-161.711226</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>120.806412</td>\n",
       "      <td>1.238816</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>114.625917</td>\n",
       "      <td>104.511918</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>75.525053</td>\n",
       "      <td>145.356696</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>71.200027</td>\n",
       "      <td>162.749132</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>61.887021</td>\n",
       "      <td>-23.112126</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>89.704786</td>\n",
       "      <td>-2.008555</td>\n",
       "      <td>ASP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>130.332598</td>\n",
       "      <td>60.007908</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>88.769547</td>\n",
       "      <td>-2.529881</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           phi         psi residue_name\n",
       "0          NaN  147.660026          THR\n",
       "1   107.832573  144.346241          THR\n",
       "2   131.170389  133.291143          CYS\n",
       "3   118.896523  151.202717          CYS\n",
       "4    76.181506  -18.979386          PRO\n",
       "5   157.924288  166.027161          SER\n",
       "6    63.579524  -42.100632          ILE\n",
       "7    55.930329  -44.561332          VAL\n",
       "8    61.364845  -43.779138          ALA\n",
       "9    63.193412  -43.303486          ARG\n",
       "10   61.208800  -42.430336          SER\n",
       "11   64.920400  -39.489513          ASN\n",
       "12   59.102393  -47.211965          PHE\n",
       "13   62.823308  -35.231846          ASN\n",
       "14   69.239667  -41.218779          VAL\n",
       "15   56.671614  -36.000832          CYS\n",
       "16   77.068519  -16.126707          ARG\n",
       "17   53.208796  -46.167676          LEU\n",
       "18   77.189679   -7.597741          PRO\n",
       "19 -106.257207    7.311793          GLY\n",
       "20   52.674731  136.344219          THR\n",
       "21   56.977153  146.617514          PRO\n",
       "22   56.412840  -36.191815          GLU\n",
       "23   63.430344  -34.860660          ALA\n",
       "24   74.766620  -37.894498          ILE\n",
       "25   64.948033  -31.686198          CYS\n",
       "26   62.036226  -54.031875          ALA\n",
       "27   68.778289  -25.486562          THR\n",
       "28   67.592836  -36.300194          TYR\n",
       "29  108.956767  -18.473152          THR\n",
       "30  -91.815951   -3.067675          GLY\n",
       "31   69.515392  164.375577          CYS\n",
       "32  129.761553  157.032715          ILE\n",
       "33  111.563161  129.588363          ILE\n",
       "34  124.130558  158.127320          ILE\n",
       "35   78.150912  -24.351838          PRO\n",
       "36   89.739760 -161.711226          GLY\n",
       "37  120.806412    1.238816          ALA\n",
       "38  114.625917  104.511918          THR\n",
       "39   75.525053  145.356696          CYS\n",
       "40   71.200027  162.749132          PRO\n",
       "41   61.887021  -23.112126          GLY\n",
       "42   89.704786   -2.008555          ASP\n",
       "43  130.332598   60.007908          TYR\n",
       "44   88.769547   -2.529881          ALA"
      ]
     },
     "execution_count": 465,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/1crn.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "metadata": {},
   "outputs": [],
   "source": [
    "phi_psi = phi_psi.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnIAAAJSCAYAAABHgiijAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHcBJREFUeJzt3WGMZfdZ3/Hfg1e4NE6UIDsbYw+1qYxQAsIio5BSgcYiEGd54TiSW6dqkwLVxjRpK+iLJgoqtJGBokKk0GBYiomjClyr1IlFFkISaUgVOSR2cRM71LDrBHZjKyYJQ9hgTDz598Vcl5Ezszt3sjPnPjufjzSamXPPzD779/Hd755zZ0+NMQIAQD9fM/UAAADsjpADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQ1KGpB9gvl1566bjqqqumHmNuX/ziF/OsZz1r6jHasW7zs2bze+TPvpj19fVc84LnTD1KK4613bFu8+u8Zvfff/9nxxiXnWu/AxNyV111Ve67776px5jb6upqVlZWph6jHes2P2s2v3/8y/dmbW0t7/13r5h6lFYca7tj3ebXec2q6k92sp9LqwAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFOHph4AAM6n48eP72i/I0eO7PEkfZ06dWrH+y4tLe3hJOfHhfb72cwZOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQ1KGpBwCA8+nIkSNTj3Cg7PSG9FPejH7KX3uvOSMHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFMLEXJVdXtVPV5VD27a9pNV9emqemD2dmTTY2+qqhNV9XBVvXyaqQEAprUQIZfkHUmu32L7W8cY187ejidJVb0wyc1JXjT7ml+sqov2bVIAgAWxECE3xvhgks/vcPcbktw5xnhyjPHJJCeSvGTPhgMAWFCLfouuN1TVa5Lcl+TfjjH+PMkVST68aZ/Ts21foaqOJjmaJIcPH87q6ureTrsHzpw503LuqVm3+Vmz+a2tPZH19XXrNifH2u7s57p96UtfOu/f8+TJk+f9e57LQTjWFjnkbkvyliRj9v7nkvxQktpi37HVNxhjHEtyLEmWl5fHysrKngy6l1ZXV9Nx7qlZt/lZs/nd9vC9WVtbs25zcqztzn6u207vnzqPKe53ehCOtYUNuTHGZ57+uKp+JclvzT49nWTz0XBlkkf3cTSAnDp1Kk8++WTGGGf9Q+9Cvln3+TDlDdc73Ox9p87376XD75kNC/Eaua1U1eWbPr0xydM/0XpPkpur6uKqujrJNUk+st/zAQBMbSHOyFXVbyRZSXJpVZ1O8hNJVqrq2mxcNv1UktclyRjjoaq6K8knkjyV5PVjjPUp5gYAmNJChNwY49VbbP7Vs+x/a5Jb924iAIDFt7CXVgEAODshBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTC/EPAgN0s7S0lIsvPp0nnnjCfSmbupD+u035e7mQ7lnbkTNyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU+61CsBk3H/z/Jjyfqf+G07LGTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0NShqQcAAL46blx/cDkjBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTh6YeAADYH6dOndrRfktLS3s8CeeLM3IAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0tRMhV1e1V9XhVPbhp29dX1fuq6o9n7583215V9baqOlFVH6uq75hucgCA6SxEyCV5R5Lrn7HtjUk+MMa4JskHZp8nySuSXDN7O5rktn2aEQBgoSxEyI0xPpjk88/YfEOSO2Yf35HklZu2v3Ns+HCS51bV5fszKQDA4jg09QBncXiM8ViSjDEeq6rnz7ZfkWTzXX9Pz7Y99sxvUFVHs3HWLocPH87q6uqeDrwXzpw503LuqVm3+Vmz+a2tPZH19XXrNifH2u7s57qdPHlyX36dvXYQjrVFDrnt1BbbxlY7jjGOJTmWJMvLy2NlZWUPx9obq6ur6Tj31Kzb/KzZ/G57+N6sra1Ztzk51nbHus3vIKzZQlxa3cZnnr5kOnv/+Gz76SRLm/a7Msmj+zwbAMDkFjnk7kny2tnHr03y7k3bXzP76dWXJvmLpy/BAgAcJAtxabWqfiPJSpJLq+p0kp9I8jNJ7qqqH07yp0lumu1+PMmRJCeS/FWSH9z3gQEAFsBChNwY49XbPPS9W+w7krx+bycCAFh8i3xpFQCAsxByAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKYOTT0AANDXqVOndrTf0tLSHk9yMDkjBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaOjT1AABAX0tLS1OPcKA5IwcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0NTC39mhqj6V5C+TrCd5aoyxXFVfn+S/J7kqyaeS/KMxxp9PNSMAwBS6nJG7boxx7Rhjefb5G5N8YIxxTZIPzD4HADhQuoTcM92Q5I7Zx3ckeeWEswAATKJDyI0kv1tV91fV0dm2w2OMx5Jk9v75k00HADCRGmNMPcNZVdU3jDEerarnJ3lfkn+V5J4xxnM37fPnY4znbfG1R5McTZLDhw+/+M4779yvsc+bM2fO5JJLLpl6jHas2/ys2fx++vefyPr6en78u6zbPBxru2Pd5td5za677rr7N72kbFsL/8MOY4xHZ+8fr6q7k7wkyWeq6vIxxmNVdXmSx7f52mNJjiXJ8vLyWFlZ2aepz5/V1dV0nHtq1m1+1mx+tz18b9bW1qzbnBxru2Pd5ncQ1myhL61W1bOq6tlPf5zk+5M8mOSeJK+d7fbaJO+eZkIAgOks+hm5w0nurqpkY9ZfH2P8TlV9NMldVfXDSf40yU0TzggAMImFDrkxxiNJvn2L7Z9L8r37PxEAwOJY6EurAABsT8gBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKCpQ1MPADt16tSpqUcAgIXijBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgqbYhV1XXV9XDVXWiqt449TwAAPvtnPdararXzD48k+SRMcYDezvSuVXVRUnenuT7kpxO8tGqumeM8YlpJ2MvLS0t7XjfkydP7mi/nd6/dZ5fGziYjh8/vqP9jhw5sseTfPXmube158dpnTPkklw3e/93k3xLVY0k/2KMcd/ejXVOL0lyYozxSJJU1Z1Jbkgi5ACAA+OcITfG+MHNn1fVDyR5V1V9+xjjc3s22dldkWTzXxdOJ/nOs37FZ/84+bUf2MuZ9sS1a2vJJ5879Rjt7HTdLnvyr3f2DS/+O1/lRIvPsTa/f/+5v8hTTz2V/Np/mXqUVi7UY+07P7/DPxI/8/Zdff/9XLcdPzcmC/38eKEea5vt5NLqj22xeS3Jf6qqT4wxfv78j3VOtcW28RU7VR1NcjRJvvXyi7O2trbXc5136+vrLeee2k7X7cvjyzv6fn/9xBxPak051ub31FPryYh1m9OFeqyN8RV/DG1pt7/3/Vy3nT43Jov9/HihHmub7eTS6rO32PY12bjUesn5HWfHTifZfFH+yiSPPnOnMcaxJMeSZHl5eTz3Rz+0P9OdR6urq1lZWZl6jHZ2um5eI/e3HGvze90v35u1tbW890dfMfUorVyox9pev0ZuP9ftQnmNXOtj7ce2Omf1lXZyafU/bP68qr4vG2e5/s0Y4892NdxX76NJrqmqq5N8OsnNSf7JRLMAAExiJ5dWb599+HVJvmX2NTdNGHEZYzxVVW9I8t4kFyW5fYzx0FTzAABMYSeXVn9v9v6LSR5J8gdjpy8E2ENjjONJdnYeGwDgArSTS6t37McgAADMp+2dHQAADjohBwDQlJADAGhKyAEANLWTn1qFC9Yi/0OWwN463/8g+G7/od9F5LmxD2fkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCAptxrFYAD6eMf//iO9utw39Hzfd9Y+nBGDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaMq9VgFYeDu9l+g8vu3bvu28f8+puIfqweWMHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBo6tDUAwBwcB0/fnxH+33oQx/a8fe85ZZbdjsOtOOMHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JR7rQKw8H7qp35qx/vu9F6rS0tLux0HFoYzcgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFPutQrAZI4cObKj/d7znvfs+Hu6hyoHycKekauqn6yqT1fVA7O3I5see1NVnaiqh6vq5VPOCQAwlUU/I/fWMcZ/3ryhql6Y5OYkL0ryDUneX1XfPMZYn2JAAICpLOwZubO4IcmdY4wnxxifTHIiyUsmngkAYN8tesi9oao+VlW3V9XzZtuuSHJq0z6nZ9sAAA6UGmNM94tXvT/JC7Z46M1JPpzks0lGkrckuXyM8UNV9fYk944x/tvse/xqkuNjjN/c4vsfTXI0SQ4fPvziO++8c29+I3vozJkzueSSS6Yeox3rNj9rNr+f/v0nsr6+nh//Lus2j90ca1/4whd2vO9znvOceUdqwf+j8+u8Ztddd939Y4zlc+036Wvkxhgv28l+VfUrSX5r9unpJJt/JOnKJI9u8/2PJTmWJMvLy2NlZWXXs05ldXU1HeeemnWbnzWb320P35u1tTXrNqfdHGvHjx/f8b4X6n8P/4/O7yCs2cJeWq2qyzd9emOSB2cf35Pk5qq6uKquTnJNko/s93wAAFNb5J9a/dmqujYbl1Y/leR1STLGeKiq7kryiSRPJXm9n1gFAA6ihQ25McY/O8tjtya5dR/HAQBYOAt7aRUAgLMTcgAATQk5AICmFvY1cgDwtCNHjpx7JziAnJEDAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoKlJQ66qbqqqh6rqy1W1/IzH3lRVJ6rq4ap6+abt18+2naiqN+7/1AAAi2HqM3IPJnlVkg9u3lhVL0xyc5IXJbk+yS9W1UVVdVGStyd5RZIXJnn1bF8AgAPn0JS/+BjjD5Okqp750A1J7hxjPJnkk1V1IslLZo+dGGM8Mvu6O2f7fmJ/JgYAWByThtxZXJHkw5s+Pz3bliSnnrH9O7f7JlV1NMnRJDl8+HBWV1fP75T74MyZMy3nnpp1m581m9/a2hNZX1+3bnNyrO2OdZvfQVizPQ+5qnp/khds8dCbxxjv3u7Lttg2svWl4LHdrz3GOJbkWJIsLy+PlZWVsw+7gFZXV9Nx7qlZt/lZs/nd9vC9WVtbs25zcqztjnWb30FYsz0PuTHGy3bxZaeTLG36/Mokj84+3m47AMCBMvUPO2znniQ3V9XFVXV1kmuSfCTJR5NcU1VXV9XXZuMHIu6ZcE4AgMlM+hq5qroxyS8kuSzJe6rqgTHGy8cYD1XVXdn4IYankrx+jLE++5o3JHlvkouS3D7GeGii8QEAJjX1T63eneTubR67NcmtW2w/nuT4Ho8GALDwFvXSKgAA5yDkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACamjTkquqmqnqoqr5cVcubtl9VVU9U1QOzt1/a9NiLq+rjVXWiqt5WVTXN9AAA05r6jNyDSV6V5INbPHZyjHHt7O2WTdtvS3I0yTWzt+v3fkwAgMUzaciNMf5wjPHwTvevqsuTPGeMce8YYyR5Z5JX7tmAAAALbOozcmdzdVX9QVX9XlV992zbFUlOb9rn9GwbAMCBc2ivf4Gqen+SF2zx0JvHGO/e5sseS/KNY4zPVdWLk7yrql6UZKvXw42z/NpHs3EZNocPH87q6upcsy+CM2fOtJx7atZtftZsfmtrT2R9fd26zcmxtjvWbX4HYc32POTGGC/bxdc8meTJ2cf3V9XJJN+cjTNwV27a9cokj57l+xxLcixJlpeXx8rKyryjTG51dTUd556adZufNZvfbQ/fm7W1Nes2J8fa7li3+R2ENVvIS6tVdVlVXTT7+Juy8UMNj4wxHkvyl1X10tlPq74myXZn9QAALmhT//MjN1bV6ST/IMl7quq9s4e+J8nHqur/JPkfSW4ZY3x+9tiPJPmvSU4kOZnkt/d5bACAhbDnl1bPZoxxd5K7t9j+m0l+c5uvuS/Jt+7xaAAAC2/SkAOAC8WpU6d2tN/S0tIeT8JBspCvkQMA4NyEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0NShqQcAgAvB0tLS1CNwADkjBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGiqxhhTz7AvqurPkvzJ1HPswqVJPjv1EA1Zt/lZs92xbvOzZrtj3ebXec3+3hjjsnPtdGBCrququm+MsTz1HN1Yt/lZs92xbvOzZrtj3eZ3ENbMpVUAgKaEHABAU0Ju8R2beoCmrNv8rNnuWLf5WbPdsW7zu+DXzGvkAACackYOAKApIbcgquqmqnqoqr5cVcubtl9VVU9U1QOzt1/a9NiLq+rjVXWiqt5WVTXN9NPZbt1mj71ptjYPV9XLN22/frbtRFW9cf+nXixV9ZNV9elNx9iRTY9tuYY4juZRVZ+aPVc9UFX3zbZ9fVW9r6r+ePb+eVPPOaWqur2qHq+qBzdt23KNasPbZsfex6rqO6abfFrbrNuBek4TcovjwSSvSvLBLR47Oca4dvZ2y6bttyU5muSa2dv1ez/mwtly3arqhUluTvKibKzLL1bVRVV1UZK3J3lFkhcmefVs34PurZuOsePJ9ms45ZCLwnG0K9fNjq+n/8L1xiQfGGNck+QDs88PsnfkK5/Dt1ujV+Rvn/ePZuPPgoPqHdn6z74D85wm5BbEGOMPxxgP73T/qro8yXPGGPeOjRc6vjPJK/dswAV1lnW7IcmdY4wnxxifTHIiyUtmbyfGGI+MMf4myZ2zfflK260hjqPz4YYkd8w+viMH8PlrszHGB5N8/hmbt1ujG5K8c2z4cJLnzv5MOHC2WbftXJDPaUKuh6ur6g+q6veq6rtn265IcnrTPqdn29hwRZJTmz5/en22237QvWF2ieb2TZe4rNX2rM18RpLfrar7q+robNvhMcZjSTJ7//zJpltc262R4+/cDsxz2qGpBzhIqur9SV6wxUNvHmO8e5sveyzJN44xPldVL07yrqp6UZKtXg93Qf4I8i7Xbbv12eovLxfkum12tjXMxmWZt2RjHd6S5OeS/FAO0DG2C9ZmPv9wjPFoVT0/yfuq6v9OPVBzjr+zO1DPaUJuH40xXraLr3kyyZOzj++vqpNJvjkbf5O4ctOuVyZ59HzMuWh2s27ZWJ+lTZ9vXp/ttl+wdrqGVfUrSX5r9unZ1vCgszZzGGM8Onv/eFXdnY3LWZ+pqsvHGI/NLgs+PumQi2m7NXL8ncUY4zNPf3wQntNcWl1wVXXZ0y/GrKpvysaLWx+ZnWb/y6p66eynVV+TZLuzUwfRPUlurqqLq+rqbKzbR5J8NMk1VXV1VX1tNl74es+Ec07uGa+tuTEbP0CSbL+GOI52rKqeVVXPfvrjJN+fjWPsniSvne322nj+2sp2a3RPktfMfnr1pUn+4ulLsBy85zRn5BZEVd2Y5BeSXJbkPVX1wBjj5Um+J8l/rKqnkqwnuWWM8fQLO38kGz+x83VJfnv2dqBst25jjIeq6q4kn0jyVJLXjzHWZ1/zhiTvTXJRktvHGA9NNP6i+NmqujYblxg+leR1SXK2NTzoxhhPOY527HCSuzf+vplDSX59jPE7VfXRJHdV1Q8n+dMkN0044+Sq6jeSrCS5tKpOJ/mJJD+TrdfoeJIj2Xix/l8l+cF9H3hBbLNuKwfpOc2dHQAAmnJpFQCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBzALlXVr1bVv5x6DuDg8g8CA8ypqm5I8tYkz07ypSSfTvLP3d0B2G9CDmAOVfX3k/yvbNwW6F8nuS/JF5LcmuRbL4Rb/gB9uLQKMJ/vT/KuMcYfPb1hjPE/k3w5GzfhBtg3Qg5gfltdyqjZG8C+EXIA83l/khtnl1iT/P/XzB1K8kfbfhXAHvAaOYA5VdWrkvx8kq9L8jdJPp/kn44xPj7pYMCBI+QAdqGqKsntSf73GOMXpp4HOJiEHABAU14jBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoKn/B7zOfmTFSShFAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "plt.hist2d(phi_psi.phi, phi_psi.psi, bins=(40, 40), cmap=plt.cm.Greys);\n",
    "plt.plot(np.zeros_like(phi_psi.phi), np.linspace(-180, 180, len(phi_psi.phi)))\n",
    "plt.plot(np.linspace(-180, 180, len(phi_psi.phi)), np.zeros_like(phi_psi.phi))\n",
    "plt.xlabel('$\\Phi$')\n",
    "plt.ylabel('$\\Psi$')\n",
    "plt.xlim([-180, 180])\n",
    "plt.ylim([-180, 180])\n",
    "plt.grid()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
