{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment II\n",
    "* Leoni Mota Loris"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* At first, Load the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import namedtuple\n",
    "\n",
    "Atom = namedtuple('Atom', ['name', 'element', 'aminoacid', 'x', 'y', 'z', 'h', 'residue_seq_number'])\n",
    "\n",
    "Atom.parse_name = lambda string: string[12:16].strip()\n",
    "Atom.parse_element = lambda string: string[12] if len(string.strip()) < 78 \\\n",
    "                     else string[76:78].strip()\n",
    "Atom.parse_aminoacid = lambda string: string[17:20].strip()\n",
    "Atom.parse_x = lambda string: float(string[30:38].strip())\n",
    "Atom.parse_y = lambda string: float(string[38:46].strip())\n",
    "Atom.parse_z = lambda string: float(string[46:54].strip())\n",
    "Atom.parse_residue_seq_number = lambda string: int(string[23:30].strip())\n",
    "\n",
    "Atom.parse = lambda string: Atom(\n",
    "    Atom.parse_name(string), Atom.parse_element(string), Atom.parse_aminoacid(string),\n",
    "    Atom.parse_x(string), Atom.parse_y(string), Atom.parse_z(string), 1,Atom.parse_residue_seq_number(string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_molecule(path):\n",
    "    atoms = []\n",
    "    with open(path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('ATOM'):\n",
    "                try:\n",
    "                    atoms.append(Atom.parse(line))\n",
    "                except Exception as e:\n",
    "                    print(e)\n",
    "    return pd.DataFrame(atoms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>17.047</td>\n",
       "      <td>14.099</td>\n",
       "      <td>3.625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>16.967</td>\n",
       "      <td>12.784</td>\n",
       "      <td>4.338</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.685</td>\n",
       "      <td>12.755</td>\n",
       "      <td>5.133</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.268</td>\n",
       "      <td>13.825</td>\n",
       "      <td>5.594</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.170</td>\n",
       "      <td>12.703</td>\n",
       "      <td>5.337</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>19.334</td>\n",
       "      <td>12.829</td>\n",
       "      <td>4.463</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.150</td>\n",
       "      <td>11.546</td>\n",
       "      <td>6.304</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.115</td>\n",
       "      <td>11.555</td>\n",
       "      <td>5.265</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>13.856</td>\n",
       "      <td>11.469</td>\n",
       "      <td>6.066</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>14.164</td>\n",
       "      <td>10.785</td>\n",
       "      <td>7.379</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name element aminoacid       x       y      z  h  residue_seq_number\n",
       "0    N       N       THR  17.047  14.099  3.625  1                   1\n",
       "1   CA       C       THR  16.967  12.784  4.338  1                   1\n",
       "2    C       C       THR  15.685  12.755  5.133  1                   1\n",
       "3    O       O       THR  15.268  13.825  5.594  1                   1\n",
       "4   CB       C       THR  18.170  12.703  5.337  1                   1\n",
       "5  OG1       O       THR  19.334  12.829  4.463  1                   1\n",
       "6  CG2       C       THR  18.150  11.546  6.304  1                   1\n",
       "7    N       N       THR  15.115  11.555  5.265  1                   2\n",
       "8   CA       C       THR  13.856  11.469  6.066  1                   2\n",
       "9    C       C       THR  14.164  10.785  7.379  1                   2"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crambin = read_molecule('assets/1crn.pdb')\n",
    "crambin[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* I) First we need to map the sequence of characters as in:\n",
    "\n",
    "**TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN**\n",
    "to a sequence of amino acids \n",
    "\n",
    "* II) And then, infer the position of each atom of the amino acid from a standard reference ($\\alpha$ carbon) \n",
    "\n",
    "\n",
    "* III) Infer the position of multiple amino acids based on the peptide bond."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### I) map the sequence of characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_mapping = {'T': 'THR', 'C': 'CYS', 'P': 'PRO', 'S': 'SER', 'I': 'ILE',\n",
    "                     'V': 'VAL', 'A': 'ALA', 'R': 'ARG', 'N': 'ASN', 'F': 'PHE',\n",
    "                     'L': 'LEU', 'G': 'GLY', 'E': 'GLU', 'D': 'ASP', 'Y': 'TYR'}\n",
    "    \n",
    "def map_to_aminoacids(sequence):\n",
    "    return list(map(lambda char: aminoacid_mapping[char], sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_sequence = 'TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids_names = map_to_aminoacids(aminoacid_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### II) infer the position of each atom of the amino acid. For this part, it is not well defined on how to infer the positions.\n",
    "\n",
    "\n",
    "* We could map this sequence to known sequences of the PDB and hope it is a known protein. This is called \n",
    "**Homology Modelling** (**SWISS-MODEL** is an example).\n",
    "\n",
    "\n",
    "* *Or*, we could predict the protein structure using an any number of softwares such as **I-TASSER (Iterative Threading ASSEmbly Refinement)**\n",
    "\n",
    "\n",
    "We're are going to use both and compare the results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I-TASSER:\n",
    "![assets/TASSER.png](assets/TASSER.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* But, if we're not using something already pre-defined, we should get the pdb file for each of those amino acids and infer the position of each amino acid by the peptide bond.\n",
    "\n",
    "*Obs.:* PDB's for each amino acid were downloaded here:\n",
    "https://www.nyu.edu/pages/mathmol/library/life/life1.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_aminoacids():\n",
    "    return list(map(lambda aminoacid_name: read_molecule('assets/' + aminoacid_name + '.pdb'), aminoacids_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids = load_aminoacids()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>0.080</td>\n",
       "      <td>-0.083</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.539</td>\n",
       "      <td>-0.033</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.032</td>\n",
       "      <td>1.394</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.268</td>\n",
       "      <td>2.339</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.116</td>\n",
       "      <td>-0.818</td>\n",
       "      <td>-1.226</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.743</td>\n",
       "      <td>-2.188</td>\n",
       "      <td>-1.162</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.654</td>\n",
       "      <td>-0.842</td>\n",
       "      <td>-1.361</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1H</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.523</td>\n",
       "      <td>0.813</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2H</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.461</td>\n",
       "      <td>-1.018</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.904</td>\n",
       "      <td>-0.508</td>\n",
       "      <td>0.929</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.691</td>\n",
       "      <td>-0.378</td>\n",
       "      <td>-2.154</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.066</td>\n",
       "      <td>-2.592</td>\n",
       "      <td>-1.973</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1HG2</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>3.986</td>\n",
       "      <td>-1.433</td>\n",
       "      <td>-2.235</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2HG2</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.083</td>\n",
       "      <td>0.168</td>\n",
       "      <td>-1.502</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3HG2</td>\n",
       "      <td>3</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.138</td>\n",
       "      <td>-1.279</td>\n",
       "      <td>-0.467</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.452</td>\n",
       "      <td>1.567</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.911</td>\n",
       "      <td>0.645</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name element aminoacid      x      y      z  h  residue_seq_number\n",
       "0      N               THR  0.080 -0.083  0.000  1                   1\n",
       "1     CA               THR  1.539 -0.033  0.000  1                   1\n",
       "2      C               THR  2.032  1.394  0.000  1                   1\n",
       "3      O               THR  1.268  2.339  0.000  1                   1\n",
       "4     CB               THR  2.116 -0.818 -1.226  1                   1\n",
       "5    OG1               THR  1.743 -2.188 -1.162  1                   1\n",
       "6    CG2               THR  3.654 -0.842 -1.361  1                   1\n",
       "7     1H       1       THR -0.523  0.813  0.000  1                   1\n",
       "8     2H       2       THR -0.461 -1.018  0.000  1                   1\n",
       "9     HA               THR  1.904 -0.508  0.929  1                   1\n",
       "10    HB               THR  1.691 -0.378 -2.154  1                   1\n",
       "11   HG1               THR  2.066 -2.592 -1.973  1                   1\n",
       "12  1HG2       1       THR  3.986 -1.433 -2.235  1                   1\n",
       "13  2HG2       2       THR  4.083  0.168 -1.502  1                   1\n",
       "14  3HG2       3       THR  4.138 -1.279 -0.467  1                   1\n",
       "15    OC               THR  3.452  1.567  0.000  1                   1\n",
       "16    HC               THR  3.911  0.645  0.000  1                   1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(aminoacids[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### III) Infer the position of multiple amino acids based on the peptide bond.\n",
    "\n",
    "\n",
    "* After the nucleotilic attack the resonance delocalization of electrons in the peptide bond forces the creation of a plane where the atoms will be located (amide plan)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* First, We define how translations and rotations are made.\n",
    "\n",
    "\n",
    "* Homogeneous coordinates were used throughout this work.\n",
    "![assets/rotation.png](assets/rotation.png)\n",
    "![assets/translation.png](assets/translation.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def translation_matrix(delta_x, delta_y, delta_z, ref_x=.0, ref_y=.0, ref_z=.0):\n",
    "    T = np.eye(4)\n",
    "    T[0, -1], T[1, -1], T[2, -1] = delta_x - ref_x, delta_y - ref_y, delta_z - ref_z\n",
    "    return T\n",
    "\n",
    "def rotation_matrix(x_rotation_rad, y_rotation_rad, z_rotation_rad):\n",
    "    Rx, Ry, Rz = np.eye(4), np.eye(4), np.eye(4)\n",
    "\n",
    "    Rx[1, 1], Rx[1, 2], Rx[2, 1], Rx[2, 2] =\\\n",
    "        np.cos(x_rotation_rad), -np.sin(x_rotation_rad), np.sin(x_rotation_rad), np.cos(x_rotation_rad)\n",
    "\n",
    "    Ry[0, 0], Ry[0, 2], Ry[2, 0], Rx[2, 2] =\\\n",
    "        np.cos(y_rotation_rad), np.sin(y_rotation_rad), -np.sin(y_rotation_rad), np.cos(y_rotation_rad)\n",
    "\n",
    "    Rz[0, 0], Rx[0, 1], Rx[1, 0], Rx[1, 1] =\\\n",
    "        np.cos(z_rotation_rad), -np.sin(z_rotation_rad), np.sin(z_rotation_rad), np.cos(z_rotation_rad)\n",
    "\n",
    "    return np.dot(np.dot(Rx, Ry), Rz)\n",
    "\n",
    "def apply_transformation(transformation, df):\n",
    "    df[['x', 'y', 'z', 'h']] = transformation.dot(df[['x', 'y', 'z', 'h']].T).T\n",
    "\n",
    "def apply_rotation(transformation, df, atom_reference='N'):\n",
    "    position = df[df.name == atom_reference]\n",
    "    apply_transformation(translation_matrix(-position.x, -position.y, -position.z), df)\n",
    "    apply_transformation(transformation, df)\n",
    "    apply_transformation(translation_matrix(position.x, position.y, position.z), df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# rotations_z = np.arange(0, 190, 10) * np.pi / 180\n",
    "\n",
    "# for rotation_z in rotations_z:\n",
    "#     a = aminoacids[0].copy()\n",
    "#     apply_rotation(rotation_matrix(rotation_z, 0, 0), a)    \n",
    "#     plt.scatter(a[:10].x, a[:10].z, c=np.arange(10), s=70)\n",
    "#     plt.xlabel('X!!!!')\n",
    "#     print(rotation_z *180 / np.pi)\n",
    "#     plt.xlim(-5, 5)\n",
    "#     plt.ylim(-5, 5)\n",
    "#     plt.grid()\n",
    "#     plt.show()\n",
    "#     plt.pause(0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dehydration(aminoacid1, aminoacid2):\n",
    "    aminoacid1 = aminoacid1[~aminoacid1.name.isin(['OC', 'HC'])]    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some utility functions for calculating the euclidean distance and the angle between two vectors:\n",
    "\n",
    "\n",
    "$$distance = \\sqrt{|(A - B)|}$$\n",
    "\n",
    "\n",
    "$$cos(\\alpha) = \\frac{A \\cdot B}{|A||B|}$$ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def euclidean_distance(p1, p2):\n",
    "    return np.sqrt(np.sum((p1 - p2) ** 2))\n",
    "\n",
    "def l2_norm(p):\n",
    "    return np.sqrt(np.sum(p ** 2))\n",
    "\n",
    "def get_angle_rad(a, b):\n",
    "    angle = np.arccos(((a).dot(b))/(l2_norm(a) * l2_norm(b))) \n",
    "    return angle if not np.isnan(angle) else .0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will consider the omega torsion angle of `Proline` to be close to zero degrees for the `cis` configuration, or most often and 180 degrees for the trans configuration.\n",
    "\n",
    "\n",
    "IMAGEM AQUI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_omega(current_aminoacid):\n",
    "    return 0 if 'PRO' in current_aminoacid.aminoacid else 180\n",
    "\n",
    "def get_desired_OCN_angle(current_aminoacid):\n",
    "    return 123 if 'PRO' in current_aminoacid.aminoacid else 123.5\n",
    "\n",
    "def get_desired_CNCa_angle(current_aminoacid):\n",
    "    return 126 if 'PRO' in current_aminoacid.aminoacid else 122\n",
    "\n",
    "def get_desired_bond_length(current_aminoacid):\n",
    "    return 1.32 if 'PRO' in current_aminoacid.aminoacid else 1.33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_nitrogen_location(aminoacid):\n",
    "    alpha_carbon = aminoacid[aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    carboxyl_carbon = aminoacid[aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    oxygen_double_bonded = aminoacid[aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "\n",
    "    CO = oxygen_double_bonded - carboxyl_carbon\n",
    "    CA = alpha_carbon - carboxyl_carbon\n",
    "\n",
    "    plane_normal = np.cross(CO, CA)\n",
    "    plane_vector_basis = np.array([\n",
    "        np.cross(CO, plane_normal) / l2_norm(np.cross(CO, plane_normal)), \n",
    "        CO / l2_norm(CO), \n",
    "        plane_normal / l2_norm(plane_normal)]).T\n",
    "\n",
    "    CO_new_norm = CO * get_desired_bond_length(aminoacid) / l2_norm(CO)\n",
    "    CO_new_norm_plane = np.linalg.solve(plane_vector_basis, CO_new_norm)[:2]\n",
    "    \n",
    "    rotation_angle = get_desired_OCN_angle(aminoacid) * np.pi / 180\n",
    "    rotation_angle_matrix = np.array(\n",
    "        [[np.cos(rotation_angle), np.sin(rotation_angle)], \n",
    "         [np.sin(rotation_angle), np.cos(rotation_angle)]])\n",
    "\n",
    "    CN_plane = CO_new_norm_plane @ rotation_angle_matrix\n",
    "    CN = np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array([CN_plane[0], CN_plane[1], 0]))\n",
    "    return CN + carboxyl_carbon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C'):\n",
    "    if terminal == 'C':\n",
    "        alpha_carbon = previous_aminoacid[previous_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        oxygen_double_bonded = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C1 = oxygen_double_bonded - carboxyl_carbon\n",
    "        C2 = alpha_carbon - carboxyl_carbon\n",
    "    if terminal == 'N':\n",
    "        last_carbon = current_aminoacid[current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        nitrogen_amide = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C2 = last_carbon - nitrogen_amide\n",
    "        C1 = carboxyl_carbon - nitrogen_amide\n",
    "    return np.cross(C1, C2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    desired_nitrogen_location = get_desired_nitrogen_location(previous_aminoacid)\n",
    "    current_nitrogen_location = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        \n",
    "    required_tranlation = desired_nitrogen_location - current_nitrogen_location\n",
    "    print('required_tranlation', required_tranlation)\n",
    "    required_tranlation_matrix = translation_matrix(required_tranlation[0], required_tranlation[1], required_tranlation[2])\n",
    "    apply_transformation(required_tranlation_matrix, current_aminoacid)\n",
    "    \n",
    "def rotate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    omega = get_desired_omega(current_aminoacid)\n",
    "    previous_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C')\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_x = get_angle_rad(current_normal[[1, 2]], previous_normal[[1, 2]])\n",
    "    apply_rotation(rotation_matrix(rotation_angle_x, 0, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_y = get_angle_rad(current_normal[[0, 2]], previous_normal[[0, 2]])\n",
    "    apply_rotation(rotation_matrix(0, rotation_angle_y, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_z = get_angle_rad(current_normal[[0, 1]], previous_normal[[0, 1]])\n",
    "    apply_rotation(rotation_matrix(0, 0, rotation_angle_z), current_aminoacid)\n",
    "\n",
    "def rotate_for_CNCa_angle(previous_aminoacid, current_aminoacid):\n",
    "    carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    nitrogen_amide = current_aminoacid[current_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    alpha_carbon = current_aminoacid[current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "\n",
    "    NC = carboxyl_carbon - nitrogen_amide\n",
    "    NCa = alpha_carbon - nitrogen_amide\n",
    "    \n",
    "    plane_normal = np.cross(NCa, NC)\n",
    "    plane_vector_basis = np.array([\n",
    "        NC / l2_norm(NC),\n",
    "        np.cross(NC, plane_normal) / l2_norm(np.cross(NC, plane_normal)), \n",
    "        plane_normal / l2_norm(plane_normal)]).T\n",
    "\n",
    "    new_NC = np.linalg.solve(plane_vector_basis, CO_new_norm)[:2]\n",
    "    new_NCa = np.linalg.solve(plane_vector_basis, CO_new_norm)[:2]\n",
    "    \n",
    "    CO_new_norm = CO * get_desired_bond_length(aminoacid) / l2_norm(CO)\n",
    "    CO_new_norm_plane = np.linalg.solve(plane_vector_basis, CO_new_norm)[:2]\n",
    "    \n",
    "    rotation_angle = get_desired_OCN_angle(aminoacid) * np.pi / 180\n",
    "    rotation_angle_matrix = np.array(\n",
    "        [[np.cos(rotation_angle), np.sin(rotation_angle)], \n",
    "         [np.sin(rotation_angle), np.cos(rotation_angle)]])\n",
    "\n",
    "    CN_plane = CO_new_norm_plane @ rotation_angle_matrix\n",
    "    CN = np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array([CN_plane[0], CN_plane[1], 0]))\n",
    "    return CN + carboxyl_carbon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_peptide_bond_length(previous_aminoacid, current_aminoacid):\n",
    "    a = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    b = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('euclidean_distance:', euclidean_distance(a, b))\n",
    "    \n",
    "def check_OCN_angle(previous_aminoacid, current_aminoacid):\n",
    "    c = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    n = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    o = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('angle ', 180*get_angle_rad(o-c, n-c)/np.pi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_tranlation [3.27597809 1.60341997 0.        ]\n",
      "required_tranlation [6.52995619 3.26683993 0.        ]\n",
      "required_tranlation [9.73111425 5.01321629 0.        ]\n",
      "required_tranlation [12.92227231  5.93259264 -0.124     ]\n",
      "required_tranlation [16.32312732  7.94976373 -1.35328062]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:8: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_tranlation [17.89360012  8.22475197 -2.99257939]\n",
      "required_tranlation [17.31595172 10.09411035 -1.90952138]\n",
      "required_tranlation [15.09805706  9.64464837 -1.24089123]\n",
      "required_tranlation [17.57025095  9.15413563 -3.05199295]\n",
      "required_tranlation [18.15341985 12.94823496 -3.55867366]\n",
      "required_tranlation [19.23194688 15.85334758 -4.23767733]\n",
      "required_tranlation [22.39419957 17.61776004 -4.88223142]\n",
      "required_tranlation [22.30698043 15.71905078 -3.83242688]\n",
      "required_tranlation [22.18352125 18.0162012  -4.98266061]\n",
      "required_tranlation [23.86806185 20.33106929 -6.39390939]\n",
      "required_tranlation [23.19182125 19.99526706 -8.96297437]\n",
      "required_tranlation [23.70761443 23.83605428 -9.48418243]\n",
      "required_tranlation [26.69535213 24.84228831 -9.10662031]\n",
      "required_tranlation [ 24.72382294  24.48818948 -11.46815849]\n",
      "required_tranlation [ 27.17106272  27.10822491 -11.56509476]\n",
      "required_tranlation [ 30.32036125  28.13999634 -11.75066971]\n",
      "required_tranlation [ 33.65725652  30.24229904 -12.88497225]\n",
      "required_tranlation [ 36.48190893  28.42165848 -14.31565615]\n",
      "required_tranlation [ 32.95559235  27.10010502 -13.93427299]\n",
      "required_tranlation [ 36.03466211  28.98932666 -13.52250639]\n",
      "required_tranlation [ 39.2085737   30.77624474 -13.40629843]\n",
      "required_tranlation [ 42.50969499  32.29375858 -13.36875355]\n",
      "required_tranlation [ 45.37555249  34.58291844 -12.70482525]\n",
      "required_tranlation [ 48.89291555  36.11341331 -12.57125501]\n",
      "required_tranlation [ 45.24891211  36.41633036 -13.30066308]\n",
      "required_tranlation [ 45.15003076  38.57751815 -13.09858828]\n",
      "required_tranlation [ 48.29281757  40.36829816 -13.74992548]\n",
      "required_tranlation [ 47.12710369  38.62518095 -12.48477971]\n",
      "required_tranlation [ 48.50800232  41.71124484 -12.24444873]\n",
      "required_tranlation [ 51.61613256  42.7290519  -12.81822354]\n",
      "required_tranlation [ 54.65013705  45.67569656 -13.89201859]\n",
      "required_tranlation [ 56.27467486  44.27147345 -14.7429431 ]\n",
      "required_tranlation [ 52.7737227   42.83580958 -14.41374376]\n",
      "required_tranlation [ 55.68456492  44.95175236 -13.37192085]\n",
      "required_tranlation [ 59.05905521  46.51347966 -13.03876036]\n",
      "required_tranlation [ 55.49347869  46.18873577 -14.03548062]\n",
      "required_tranlation [ 52.40556861  45.4763498  -15.682064  ]\n",
      "required_tranlation [ 48.69604721  44.70073286 -16.90299579]\n",
      "required_tranlation [ 53.57460012  46.89513072 -15.97760182]\n",
      "required_tranlation [ 54.42584628  49.51955408 -15.72074061]\n",
      "required_tranlation [ 57.3155818   51.36938957 -16.80969557]\n"
     ]
    }
   ],
   "source": [
    "aminoacids_df = load_aminoacids()\n",
    "for i in range(1, len(aminoacids_df)):\n",
    "    current_aminoacid = aminoacids_df[i]\n",
    "    previous_aminoacid = aminoacids_df[i - 1]\n",
    "    \n",
    "    dehydration(previous_aminoacid, current_aminoacid)\n",
    "    translate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "    rotate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "    rotate_for_CNCa_angle(previous_aminoacid, current_aminoacid)\n",
    "    \n",
    "#     check_peptide_bond_length(previous_aminoacid, current_aminoacid)\n",
    "#     check_OCN_angle(previous_aminoacid, current_aminoacid)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We have all the atoms aligned accordingly. So, let's create string rows for the pdb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_atom_to_file(atom_row, aminoacid_seq_number, fd):\n",
    "    global atom_seq_number\n",
    "    atom_seq_number += 1\n",
    "    if 'H' not in atom_row[0]:\n",
    "        atom_row.element = ''.join(filter(str.isalpha, atom_row[0]))[0]\n",
    "        template_row =\\\n",
    "        'ATOM    '+\\\n",
    "        ('%d'%atom_seq_number).rjust(3)+\\\n",
    "        '  '+\\\n",
    "        ('%s'%atom_row[0]).rjust(3)+\\\n",
    "        ' '+\\\n",
    "        ('%s'%atom_row.aminoacid).rjust(3)+\\\n",
    "        ' A'+\\\n",
    "        ('%d'%aminoacid_seq_number).rjust(4)+\\\n",
    "        '    '+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.y).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        '  1.00  5.99           '+\\\n",
    "        ('%s\\n'%atom_row.element)\n",
    "        f.write(template_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "atom_seq_number = 0\n",
    "\n",
    "with open('assets/final_inferred_molecule.pdb', 'w') as f:\n",
    "    for idx, aminoacid in enumerate(aminoacids_df):\n",
    "        for atom_row in aminoacid.iterrows():\n",
    "            write_atom_to_file(atom_row[1], idx, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## b) Receive PDB file and list $\\phi$ and $\\psi$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>17.047</td>\n",
       "      <td>14.099</td>\n",
       "      <td>3.625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>16.967</td>\n",
       "      <td>12.784</td>\n",
       "      <td>4.338</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.685</td>\n",
       "      <td>12.755</td>\n",
       "      <td>5.133</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name element aminoacid       x       y      z  h  residue_seq_number\n",
       "0    N       N       THR  17.047  14.099  3.625  1                   1\n",
       "1   CA       C       THR  16.967  12.784  4.338  1                   1\n",
       "2    C       C       THR  15.685  12.755  5.133  1                   1"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crambin = read_molecule('assets/1crn.pdb')\n",
    "crambin[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$ ψ  = Normal(N_{i}, C_{\\alpha i}, C_{i}) \\^ Normal(C_{\\alpha i}, C_{i}, N_{i+1})  $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$φ = Normal(C_{i-1}, N_{i}, C_{\\alpha i}) \\^ Normal(N_{i}, C_{\\alpha i}, C_{i}) $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![assets/phi_psi.png](assets/phi_psi.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_psi(ni, cai, ci, ni1):\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    normal_ci = np.cross((cai - ci), (ni1 - ci))\n",
    "    return np.degrees(get_angle_rad(normal_cai, normal_ci))\n",
    "\n",
    "def compute_phi(ci_prev, ni, cai, ci):\n",
    "    normal_ni = np.cross((ci_prev - ni), (cai - ni))\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    return np.degrees(get_angle_rad(normal_ni, normal_cai))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_phy_psi(molecule):\n",
    "    molecule['phi'] = None\n",
    "    molecule['psi'] = None\n",
    "\n",
    "    data = []\n",
    "\n",
    "    for i in range(1, len(molecule.residue_seq_number.unique())):\n",
    "        aminoacid_i = molecule[molecule.residue_seq_number == (i)]\n",
    "        ni = aminoacid_i[aminoacid_i.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        cai = aminoacid_i[aminoacid_i.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        ci = aminoacid_i[aminoacid_i.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        angle_data = {'residue_name': aminoacid_i.aminoacid.values[0], 'phi': None, 'psi': None}\n",
    "        try:\n",
    "            aminoacid_prev = molecule[molecule.residue_seq_number == (i-1)]\n",
    "            ci_prev = aminoacid_prev[aminoacid_prev.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['phi'] = compute_phi(ci_prev, ni, cai, ci)\n",
    "        except: pass\n",
    "\n",
    "        try:\n",
    "            aminoacid_i1 = molecule[molecule.residue_seq_number == (i+1)]\n",
    "            ni1 = aminoacid_i1[aminoacid_i1.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['psi'] = compute_psi(ni, cai, ci, ni1)\n",
    "        except: pass\n",
    "\n",
    "        data.append(angle_data)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:8: RuntimeWarning: invalid value encountered in arccos\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>LEU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ASP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>8.537736e-07</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             phi         psi residue_name\n",
       "0   1.800000e+02  180.000000          THR\n",
       "1   1.800000e+02  179.999999          CYS\n",
       "2   0.000000e+00    0.000000          CYS\n",
       "3   1.800000e+02    0.000000          PRO\n",
       "4   1.800000e+02    0.000000          SER\n",
       "5   1.800000e+02  179.999999          ILE\n",
       "6   1.800000e+02  180.000000          VAL\n",
       "7   0.000000e+00  179.999999          ALA\n",
       "8   0.000000e+00  180.000000          ARG\n",
       "9   0.000000e+00  180.000000          SER\n",
       "10  1.800000e+02  179.999999          ASN\n",
       "11  1.800000e+02  180.000000          PHE\n",
       "12  0.000000e+00    0.000000          ASN\n",
       "13  0.000000e+00  180.000000          VAL\n",
       "14  0.000000e+00    0.000000          CYS\n",
       "15  1.800000e+02  180.000000          ARG\n",
       "16  1.800000e+02  180.000000          LEU\n",
       "17  0.000000e+00    0.000000          PRO\n",
       "18  0.000000e+00  180.000000          GLY\n",
       "19  1.800000e+02  180.000000          THR\n",
       "20  1.800000e+02  179.999999          PRO\n",
       "21  0.000000e+00  180.000000          GLU\n",
       "22  1.800000e+02  180.000000          ALA\n",
       "23  1.800000e+02  180.000000          ILE\n",
       "24  1.800000e+02  180.000000          CYS\n",
       "25  1.800000e+02  180.000000          ALA\n",
       "26  0.000000e+00    0.000000          THR\n",
       "27  1.800000e+02  180.000000          TYR\n",
       "28  0.000000e+00  179.999999          THR\n",
       "29  1.800000e+02  180.000000          GLY\n",
       "30  1.800000e+02    0.000000          CYS\n",
       "31  0.000000e+00  180.000000          ILE\n",
       "32  0.000000e+00  180.000000          ILE\n",
       "33  1.800000e+02  179.999999          ILE\n",
       "34  0.000000e+00    0.000000          PRO\n",
       "35  1.800000e+02  180.000000          GLY\n",
       "36  0.000000e+00    0.000000          ALA\n",
       "37  1.800000e+02  180.000000          THR\n",
       "38  1.800000e+02  180.000000          THR\n",
       "39  0.000000e+00  180.000000          CYS\n",
       "40  1.800000e+02  180.000000          PRO\n",
       "41  1.800000e+02  180.000000          GLY\n",
       "42  0.000000e+00    0.000000          ASP\n",
       "43  8.537736e-07  180.000000          TYR\n",
       "44  1.800000e+02  180.000000          ALA\n",
       "45  0.000000e+00         NaN          ASN"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/final_inferred_molecule.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## c) Ramachandran "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>147.660026</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107.832573</td>\n",
       "      <td>144.346241</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>131.170389</td>\n",
       "      <td>133.291143</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>118.896523</td>\n",
       "      <td>151.202717</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>76.181506</td>\n",
       "      <td>18.979386</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>157.924288</td>\n",
       "      <td>166.027161</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>63.579524</td>\n",
       "      <td>42.100632</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>55.930329</td>\n",
       "      <td>44.561332</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>61.364845</td>\n",
       "      <td>43.779138</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>63.193412</td>\n",
       "      <td>43.303486</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>61.208800</td>\n",
       "      <td>42.430336</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>64.920400</td>\n",
       "      <td>39.489513</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>59.102393</td>\n",
       "      <td>47.211965</td>\n",
       "      <td>PHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>62.823308</td>\n",
       "      <td>35.231846</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>69.239667</td>\n",
       "      <td>41.218779</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>56.671614</td>\n",
       "      <td>36.000832</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>77.068519</td>\n",
       "      <td>16.126707</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>53.208796</td>\n",
       "      <td>46.167676</td>\n",
       "      <td>LEU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>77.189679</td>\n",
       "      <td>7.597741</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>106.257207</td>\n",
       "      <td>7.311793</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>52.674731</td>\n",
       "      <td>136.344219</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>56.977153</td>\n",
       "      <td>146.617514</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>56.412840</td>\n",
       "      <td>36.191815</td>\n",
       "      <td>GLU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>63.430344</td>\n",
       "      <td>34.860660</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>74.766620</td>\n",
       "      <td>37.894498</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>64.948033</td>\n",
       "      <td>31.686198</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>62.036226</td>\n",
       "      <td>54.031875</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>68.778289</td>\n",
       "      <td>25.486562</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>67.592836</td>\n",
       "      <td>36.300194</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>108.956767</td>\n",
       "      <td>18.473152</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>91.815951</td>\n",
       "      <td>3.067675</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>69.515392</td>\n",
       "      <td>164.375577</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>129.761553</td>\n",
       "      <td>157.032715</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>111.563161</td>\n",
       "      <td>129.588363</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>124.130558</td>\n",
       "      <td>158.127320</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>78.150912</td>\n",
       "      <td>24.351838</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>89.739760</td>\n",
       "      <td>161.711226</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>120.806412</td>\n",
       "      <td>1.238816</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>114.625917</td>\n",
       "      <td>104.511918</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>75.525053</td>\n",
       "      <td>145.356696</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>71.200027</td>\n",
       "      <td>162.749132</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>61.887021</td>\n",
       "      <td>23.112126</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>89.704786</td>\n",
       "      <td>2.008555</td>\n",
       "      <td>ASP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>130.332598</td>\n",
       "      <td>60.007908</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>88.769547</td>\n",
       "      <td>2.529881</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           phi         psi residue_name\n",
       "0          NaN  147.660026          THR\n",
       "1   107.832573  144.346241          THR\n",
       "2   131.170389  133.291143          CYS\n",
       "3   118.896523  151.202717          CYS\n",
       "4    76.181506   18.979386          PRO\n",
       "5   157.924288  166.027161          SER\n",
       "6    63.579524   42.100632          ILE\n",
       "7    55.930329   44.561332          VAL\n",
       "8    61.364845   43.779138          ALA\n",
       "9    63.193412   43.303486          ARG\n",
       "10   61.208800   42.430336          SER\n",
       "11   64.920400   39.489513          ASN\n",
       "12   59.102393   47.211965          PHE\n",
       "13   62.823308   35.231846          ASN\n",
       "14   69.239667   41.218779          VAL\n",
       "15   56.671614   36.000832          CYS\n",
       "16   77.068519   16.126707          ARG\n",
       "17   53.208796   46.167676          LEU\n",
       "18   77.189679    7.597741          PRO\n",
       "19  106.257207    7.311793          GLY\n",
       "20   52.674731  136.344219          THR\n",
       "21   56.977153  146.617514          PRO\n",
       "22   56.412840   36.191815          GLU\n",
       "23   63.430344   34.860660          ALA\n",
       "24   74.766620   37.894498          ILE\n",
       "25   64.948033   31.686198          CYS\n",
       "26   62.036226   54.031875          ALA\n",
       "27   68.778289   25.486562          THR\n",
       "28   67.592836   36.300194          TYR\n",
       "29  108.956767   18.473152          THR\n",
       "30   91.815951    3.067675          GLY\n",
       "31   69.515392  164.375577          CYS\n",
       "32  129.761553  157.032715          ILE\n",
       "33  111.563161  129.588363          ILE\n",
       "34  124.130558  158.127320          ILE\n",
       "35   78.150912   24.351838          PRO\n",
       "36   89.739760  161.711226          GLY\n",
       "37  120.806412    1.238816          ALA\n",
       "38  114.625917  104.511918          THR\n",
       "39   75.525053  145.356696          CYS\n",
       "40   71.200027  162.749132          PRO\n",
       "41   61.887021   23.112126          GLY\n",
       "42   89.704786    2.008555          ASP\n",
       "43  130.332598   60.007908          TYR\n",
       "44   88.769547    2.529881          ALA"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/1crn.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "phi_psi = phi_psi.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnIAAAJSCAYAAABHgiijAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHQZJREFUeJzt3W+MZfdd3/HPF68JaTYoQXYmJl5qUxlFSSQsPDKpKtCsCNhxHziO5MqpSgIBLaYJrdQ+aCI/IG20CKFCpNCQdGmsOEKwtaBOrGAIicWQVhAldnGDncSw64TsxlacfwU2gFknvz6Yu+10mdnde+OZc787r5c0mrnnnjvzm+9ezb73nDt7aowRAAD6+ZapFwAAwGKEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKb2Tb2A3XLZZZeNq666auplzO1rX/tanvOc50y9jHbMbX5mthhzm5+ZLcbc5td5Zg8++OCXxhiXn2+/PRNyV111VR544IGplzG39fX1rK2tTb2Mdsxtfma2GHObn5ktxtzm13lmVfXnF7KfU6sAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBT+6ZeAAB8M06cOLGjn//AgQNzP2beNS3yNXbaxfA9nD59+qL4Ps7FETkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0NS+qRcAAN+MZbzI+TKuaact48XpL7300ov+z8IROQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaWoqQq6o7q+rJqnp407a3VNXnq+qh2dtNm+57c1Udq6pHq+qGaVYNADCtpQi5JO9JcuMW2982xrh29nZfklTVS5LcluSls8f8SlVdsmsrBQBYEksRcmOMjyT5ygXufnOSo2OMp8YYn0lyLMn1O7Y4AIAlVWOMqdeQJKmqq5J8YIzxstnttyT5sSR/meSBJP92jPHVqvpPST46xvi12X7vTvI7Y4zf3OJzHkpyKElWVlauO3r06M5/I8+wU6dOZf/+/VMvox1zm5+ZLcbc5mdmi9ntuZ0+fXpHP/+ll166o58/6f1cO3jw4INjjNXz7bfM11p9Z5K3Jhmz97+Y5PVJaot9t6zRMcaRJEeSZHV1daytre3IQnfS+vp6Oq57auY2PzNbjLnNz8wWs9tzm/faqfPajWug7oXn2tKG3BjjC2c+rqpfTfKB2c2TSTb/6V+Z5PFdXBpAko2/6E6fPn3Bf+Fd7Bfvfqbs9MXXl/Hi7vPaje9hGb9v/r6leI3cVqrqik03b0ly5jda701yW1U9q6quTnJNko/t9voAAKa2FEfkquo3kqwluayqTib52SRrVXVtNk6bfjbJTyXJGOORqro7ySeTPJ3kDWOMr0+xbgCAKS1FyI0xXrPF5nefY//DSQ7v3IoAAJbf0p5aBQDg3IQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0txX8IDNDRgQMHcvz4cdekPI+zrwt6vuvT7vQ8L4Y/r2X8Hi6Ga9h25IgcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhq39QLAODidvbF0Y8fP+6C6bts3gvaJ/Nf1N6f6TQckQMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJpyrVUAuMi5DurFyxE5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATe2begEAwM46ceLE3I85cODADqyEZ5ojcgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJpaipCrqjur6smqenjTtu+oqg9V1Z/N3j9/tr2q6u1VdayqPlFV3zfdygEAprMUIZfkPUluPGvbm5LcP8a4Jsn9s9tJ8sok18zeDiV55y6tEQBgqSxFyI0xPpLkK2dtvjnJXbOP70ryqk3b3zs2fDTJ86rqit1ZKQDA8qgxxtRrSJJU1VVJPjDGeNns9v8eYzxv0/1fHWM8v6o+kOTnxxj/Y7b9/iT/bozxwBaf81A2jtplZWXluqNHj+78N/IMO3XqVPbv3z/1Mtoxt/mZ2WLMbX5mthhzm1/nmR08ePDBMcbq+fbbtxuLeYbVFtu2rNExxpEkR5JkdXV1rK2t7eCydsb6+no6rntq5jY/M1uMuc3PzBZjbvPbCzNbilOr2/jCmVOms/dPzrafTHJg035XJnl8l9cGADC5ZQ65e5O8bvbx65K8f9P2185+e/XlSf5ijPHEFAsEAJjSUpxararfSLKW5LKqOpnkZ5P8fJK7q+onknwuya2z3e9LclOSY0n+OsmP7/qCAQCWwFKE3BjjNdvc9UNb7DuSvGFnVwQAsPyW+dQqAADnIOQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICm9k29AACgvxMnTsy1/4EDB3ZoJXuLI3IAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoKl9Uy8AADZz8fWe/DlMwxE5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATS39fwhcVZ9N8ldJvp7k6THGalV9R5L/muSqJJ9N8s/GGF+dao0AAFPockTu4Bjj2jHG6uz2m5LcP8a4Jsn9s9sAAHtKl5A7281J7pp9fFeSV024FgCASSz9qdUkI8nvVdVI8p/HGEeSrIwxnkiSMcYTVfWCSVcIwLbOvnbq6dOnz3k91Xe9611zff7bb799rv1dE5SLSY0xpl7DOVXVd44xHp/F2oeS/EySe8cYz9u0z1fHGM/f4rGHkhxKkpWVleuOHj26W8t+xpw6dSr79++fehntmNv8zGwx5nZ+p0+f/v9u/+3f/m2+7du+bdv9v/jFL871+S+//PK59r/00kvn2n9ZeK7Nr/PMDh48+OCml5Rta+mPyI0xHp+9f7Kq7klyfZIvVNUVs6NxVyR5cpvHHklyJElWV1fH2traLq36mbO+vp6O656auc3PzBZjbud39tG3T3/603nxi1+87f7r6+tzff69ckTOc21+e2FmS/0auap6TlU998zHSX4kycNJ7k3yutlur0vy/mlWCAAwnWU/IreS5J6qSjbW+utjjN+tqo8nubuqfiLJ55LcOuEaAQAmsdQhN8Z4LMn3brH9y0l+aPdXBACwPJb61CoAANsTcgAATQk5AICmhBwAQFNCDgCgKSEHANDUUv/3IwD0d/aVFI4fP37Oqyv84R/+4Vyf/+d+7ufm2v9zn/vcXPt3vRIEe4MjcgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgqX1TLwAANvv93//9ufY/ceLEXPsfOHBgrv1hmTkiBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANOVaqwAslXmvnTqv++67b679b7rpph1ayeJcX5YzHJEDAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE3tm3oBALCZC7yfnxlxhiNyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU661CgDPsDvuuGOu/Q8fPjzX/idOnJhr/0WuzbobX4NvniNyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNtQ25qrqxqh6tqmNV9aap1wMAsNvOe4muqnrt7MNTSR4bYzy0s0s6v6q6JMk7kvxwkpNJPl5V944xPjntygAAds+FXGv14Oz9P0jy4qoaSX5yjPHAzi3rvK5PcmyM8ViSVNXRJDcnEXIAwJ5x3pAbY/z45ttV9U+TvK+qvneM8eUdW9m5vSjJ5qv5nkzy/ROtBYBGduPi7ocPH97Rz78b38NufA2+eTXGOPcOVf9mi82vT/LRJJ8cY/zSTizsXKrq1iQ3jDF+cnb7R5NcP8b4mbP2O5TkUJKsrKxcd/To0d1e6jft1KlT2b9//9TLaMfc5mdmizG3+ZnZYsxtfp1ndvDgwQfHGKvn2+9CTq0+d4tt35KNU61TTedkks3/VLgyyeNn7zTGOJLkSJKsrq6OtbW1XVncM2l9fT0d1z01c5ufmS3G3OZnZosxt/nthZldyKnVf7/5dlX9cDaOcv3rMcYXd2ph5/HxJNdU1dVJPp/ktiT/fKK1AABM4kJ+a/XO2YfPTvLi2WNunTDiMsZ4uqremOSDSS5JcucY45Gp1gMAMIULObX6B7P3X0vyWJI/Hud7Yd0uGGPcl+S+qdcBADCVCzm1etduLAQAgPm0vbIDAMBeJ+QAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJpa2pCrqrdU1eer6qHZ202b7ntzVR2rqker6oYp1wkAMJV9Uy/gPN42xviPmzdU1UuS3JbkpUm+M8mHq+p7xhhfn2KBAABTWdojcudwc5KjY4ynxhifSXIsyfUTrwkAYNcte8i9sao+UVV3VtXzZ9telOTEpn1OzrYBAOwpNcaY7otXfTjJC7e4644kH03ypSQjyVuTXDHGeH1VvSPJH40xfm32Od6d5L4xxm9t8fkPJTmUJCsrK9cdPXp0Z76RHXTq1Kns379/6mW0Y27zM7PFmNv8zGwx5ja/zjM7ePDgg2OM1fPtN+lr5MYYr7iQ/arqV5N8YHbzZJIDm+6+Msnj23z+I0mOJMnq6upYW1tbeK1TWV9fT8d1T83c5mdmizG3+ZnZYsxtfnthZkt7arWqrth085YkD88+vjfJbVX1rKq6Osk1ST622+sDAJjaMv/W6i9U1bXZOLX62SQ/lSRjjEeq6u4kn0zydJI3+I1VAGAvWtqQG2P86DnuO5zk8C4uBwBg6SztqVUAAM5NyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACamjTkqurWqnqkqr5RVatn3ffmqjpWVY9W1Q2btt8423asqt60+6sGAFgOUx+RezjJq5N8ZPPGqnpJktuSvDTJjUl+paouqapLkrwjySuTvCTJa2b7AgDsOfum/OJjjE8lSVWdfdfNSY6OMZ5K8pmqOpbk+tl9x8YYj80ed3S27yd3Z8UAAMtj0pA7hxcl+eim2ydn25LkxFnbv3+7T1JVh5IcSpKVlZWsr68/s6vcBadOnWq57qmZ2/zMbDHmNj8zW4y5zW8vzGzHQ66qPpzkhVvcdccY4/3bPWyLbSNbnwoe233tMcaRJEeSZHV1daytrZ17sUtofX09Hdc9NXObn5ktxtzmZ2aLMbf57YWZ7XjIjTFescDDTiY5sOn2lUken3283XYAgD1l6l922M69SW6rqmdV1dVJrknysSQfT3JNVV1dVd+ajV+IuHfCdQIATGbS18hV1S1JfjnJ5Ul+u6oeGmPcMMZ4pKruzsYvMTyd5A1jjK/PHvPGJB9MckmSO8cYj0y0fACASU39W6v3JLlnm/sOJzm8xfb7kty3w0sDAFh6y3pqFQCA8xByAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNTRpyVXVrVT1SVd+oqtVN26+qqr+pqodmb+/adN91VfUnVXWsqt5eVTXN6gEApjX1EbmHk7w6yUe2uO/4GOPa2dvtm7a/M8mhJNfM3m7c+WUCACyfSUNujPGpMcajF7p/VV2R5NvHGH80xhhJ3pvkVTu2QACAJTb1Eblzubqq/riq/qCqfmC27UVJTm7a5+RsGwDAnrNvp79AVX04yQu3uOuOMcb7t3nYE0m+a4zx5aq6Lsn7quqlSbZ6Pdw4x9c+lI3TsFlZWcn6+vpca18Gp06darnuqZnb/MxsMeY2PzNbjLnNby/MbMdDbozxigUe81SSp2YfP1hVx5N8TzaOwF25adcrkzx+js9zJMmRJFldXR1ra2vzLmVy6+vr6bjuqZnb/MxsMeY2PzNbjLnNby/MbClPrVbV5VV1yezj787GLzU8NsZ4IslfVdXLZ7+t+tok2x3VAwC4qE3934/cUlUnk/zjJL9dVR+c3fWDST5RVf8ryW8muX2M8ZXZfT+d5L8kOZbkeJLf2eVlAwAshR0/tXouY4x7ktyzxfbfSvJb2zzmgSQv2+GlAQAsvaU8tQoAwPkJOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaqjHG1GvYFVX1xSR/PvU6FnBZki9NvYiGzG1+ZrYYc5ufmS3G3ObXeWb/cIxx+fl22jMh11VVPTDGWJ16Hd2Y2/zMbDHmNj8zW4y5zW8vzMypVQCApoQcAEBTQm75HZl6AU2Z2/zMbDHmNj8zW4y5ze+in5nXyAEANOWIHABAU0JuSVTVrVX1SFV9o6pWN22/qqr+pqoemr29a9N911XVn1TVsap6e1XVNKufznZzm9335tlsHq2qGzZtv3G27VhVvWn3V71cquotVfX5Tc+xmzbdt+UM8TyaR1V9dvaz6qGqemC27Tuq6kNV9Wez98+fep1Tqqo7q+rJqnp407YtZ1Qb3j577n2iqr5vupVPa5u57amfaUJueTyc5NVJPrLFfcfHGNfO3m7ftP2dSQ4luWb2duPOL3PpbDm3qnpJktuSvDQbc/mVqrqkqi5J8o4kr0zykiSvme27171t03PsvmT7GU65yGXhebSQg7Pn15l/cL0pyf1jjGuS3D+7vZe9J3//Z/h2M3pl/t/P/UPZ+Ltgr3pPtv67b8/8TBNyS2KM8akxxqMXun9VXZHk28cYfzQ2Xuj43iSv2rEFLqlzzO3mJEfHGE+NMT6T5FiS62dvx8YYj40x/i7J0dm+/H3bzRDPo2fCzUnumn18V/bgz6/NxhgfSfKVszZvN6Obk7x3bPhokufN/k7Yc7aZ23Yuyp9pQq6Hq6vqj6vqD6rqB2bbXpTk5KZ9Ts62seFFSU5sun1mPttt3+veODtFc+emU1xmtT2zmc9I8ntV9WBVHZptWxljPJEks/cvmGx1y2u7GXn+nd+e+Zm2b+oF7CVV9eEkL9zirjvGGO/f5mFPJPmuMcaXq+q6JO+rqpcm2er1cBflryAvOLft5rPVP14uyrltdq4ZZuO0zFuzMYe3JvnFJK/PHnqOLcBs5vNPxhiPV9ULknyoqj499YKa8/w7tz31M03I7aIxxisWeMxTSZ6affxgVR1P8j3Z+JfElZt2vTLJ48/EOpfNInPLxnwObLq9eT7bbb9oXegMq+pXk3xgdvNcM9zrzGYOY4zHZ++frKp7snE66wtVdcUY44nZacEnJ13kctpuRp5/5zDG+MKZj/fCzzSnVpdcVV1+5sWYVfXd2Xhx62Ozw+x/VVUvn/226muTbHd0ai+6N8ltVfWsqro6G3P7WJKPJ7mmqq6uqm/Nxgtf751wnZM767U1t2TjF0iS7WeI59EFq6rnVNVzz3yc5Eey8Ry7N8nrZru9Ln5+bWW7Gd2b5LWz3159eZK/OHMKlr33M80RuSVRVbck+eUklyf57ap6aIxxQ5IfTPIfqurpJF9PcvsY48wLO386G7+x8+wkvzN721O2m9sY45GqujvJJ5M8neQNY4yvzx7zxiQfTHJJkjvHGI9MtPxl8QtVdW02TjF8NslPJcm5ZrjXjTGe9jy6YCtJ7tn492b2Jfn1McbvVtXHk9xdVT+R5HNJbp1wjZOrqt9Ispbksqo6meRnk/x8tp7RfUluysaL9f86yY/v+oKXxDZzW9tLP9Nc2QEAoCmnVgEAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBLKiq3l1V/3LqdQB7l/8QGGBOVXVzkrcleW6S00k+n+THXN0B2G1CDmAOVfWPkvz3bFwW6F8leSDJXyY5nORlF8Mlf4A+nFoFmM+PJHnfGONPz2wYY/y3JN/IxkW4AXaNkAOY31anMmr2BrBrhBzAfD6c5JbZKdYk//c1c/uS/Om2jwLYAV4jBzCnqnp1kl9K8uwkf5fkK0n+xRjjTyZdGLDnCDmABVRVJbkzyf8cY/zy1OsB9iYhBwDQlNfIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaOr/ANzZMHqux+yAAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "plt.hist2d(phi_psi.phi, phi_psi.psi, bins=(20, 20), cmap=plt.cm.Greys);\n",
    "plt.xlabel('$\\Phi$')\n",
    "plt.ylabel('$\\Psi$')\n",
    "plt.xlim([-180, 180])\n",
    "plt.ylim([-180, 180])\n",
    "plt.grid()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
