{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment II\n",
    "* Leoni Mota Loris"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* At first, Load the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import namedtuple\n",
    "\n",
    "Atom = namedtuple('Atom', ['name', 'element', 'aminoacid', 'x', 'y', 'z', 'h', 'residue_seq_number'])\n",
    "\n",
    "Atom.parse_name = lambda string: string[12:16].strip()\n",
    "Atom.parse_element = lambda string: string[12] if len(string.strip()) < 78 \\\n",
    "                     else string[76:78].strip()\n",
    "Atom.parse_aminoacid = lambda string: string[17:20].strip()\n",
    "Atom.parse_x = lambda string: float(string[30:38].strip())\n",
    "Atom.parse_y = lambda string: float(string[38:46].strip())\n",
    "Atom.parse_z = lambda string: float(string[46:54].strip())\n",
    "Atom.parse_residue_seq_number = lambda string: int(string[23:30].strip())\n",
    "\n",
    "Atom.parse = lambda string: Atom(\n",
    "    Atom.parse_name(string), Atom.parse_element(string), Atom.parse_aminoacid(string),\n",
    "    Atom.parse_x(string), Atom.parse_y(string), Atom.parse_z(string), 1,Atom.parse_residue_seq_number(string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_molecule(path):\n",
    "    atoms = []\n",
    "    with open(path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('ATOM'):\n",
    "                try:\n",
    "                    atoms.append(Atom.parse(line))\n",
    "                except Exception as e:\n",
    "                    print(e)\n",
    "    return pd.DataFrame(atoms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>17.047</td>\n",
       "      <td>14.099</td>\n",
       "      <td>3.625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>16.967</td>\n",
       "      <td>12.784</td>\n",
       "      <td>4.338</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.685</td>\n",
       "      <td>12.755</td>\n",
       "      <td>5.133</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.268</td>\n",
       "      <td>13.825</td>\n",
       "      <td>5.594</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.170</td>\n",
       "      <td>12.703</td>\n",
       "      <td>5.337</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>19.334</td>\n",
       "      <td>12.829</td>\n",
       "      <td>4.463</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.150</td>\n",
       "      <td>11.546</td>\n",
       "      <td>6.304</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.115</td>\n",
       "      <td>11.555</td>\n",
       "      <td>5.265</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>13.856</td>\n",
       "      <td>11.469</td>\n",
       "      <td>6.066</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>14.164</td>\n",
       "      <td>10.785</td>\n",
       "      <td>7.379</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name element aminoacid       x       y      z  h  residue_seq_number\n",
       "0    N       N       THR  17.047  14.099  3.625  1                   1\n",
       "1   CA       C       THR  16.967  12.784  4.338  1                   1\n",
       "2    C       C       THR  15.685  12.755  5.133  1                   1\n",
       "3    O       O       THR  15.268  13.825  5.594  1                   1\n",
       "4   CB       C       THR  18.170  12.703  5.337  1                   1\n",
       "5  OG1       O       THR  19.334  12.829  4.463  1                   1\n",
       "6  CG2       C       THR  18.150  11.546  6.304  1                   1\n",
       "7    N       N       THR  15.115  11.555  5.265  1                   2\n",
       "8   CA       C       THR  13.856  11.469  6.066  1                   2\n",
       "9    C       C       THR  14.164  10.785  7.379  1                   2"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crambin = read_molecule('assets/1crn.pdb')\n",
    "crambin[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* I) First we need to map the sequence of characters as in:\n",
    "\n",
    "**TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN**\n",
    "to a sequence of amino acids \n",
    "\n",
    "* II) And then, infer the position of each atom of the amino acid from a standard reference ($\\alpha$ carbon) \n",
    "\n",
    "\n",
    "* III) Infer the position of multiple amino acids based on the peptide bond."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### I) map the sequence of characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_mapping = {'T': 'THR', 'C': 'CYS', 'P': 'PRO', 'S': 'SER', 'I': 'ILE',\n",
    "                     'V': 'VAL', 'A': 'ALA', 'R': 'ARG', 'N': 'ASN', 'F': 'PHE',\n",
    "                     'L': 'LEU', 'G': 'GLY', 'E': 'GLU', 'D': 'ASP', 'Y': 'TYR'}\n",
    "    \n",
    "def map_to_aminoacids(sequence):\n",
    "    return list(map(lambda char: aminoacid_mapping[char], sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_sequence = 'TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids_names = map_to_aminoacids(aminoacid_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### II) infer the position of each atom of the amino acid. For this part, it is not well defined on how to infer the positions.\n",
    "\n",
    "\n",
    "* We could map this sequence to known sequences of the PDB and hope it is a known protein. This is called \n",
    "**Homology Modelling** (**SWISS-MODEL** is an example).\n",
    "\n",
    "\n",
    "* *Or*, we could predict the protein structure using an any number of softwares such as **I-TASSER (Iterative Threading ASSEmbly Refinement)**\n",
    "\n",
    "\n",
    "We're are going to use both and compare the results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I-TASSER:\n",
    "![assets/TASSER.png](assets/TASSER.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* But, if we're not using something already pre-defined, we should get the pdb file for each of those amino acids and infer the position of each amino acid by the peptide bond.\n",
    "\n",
    "*Obs.:* PDB's for each amino acid were downloaded here:\n",
    "https://www.nyu.edu/pages/mathmol/library/life/life1.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_aminoacids():\n",
    "    return list(map(lambda aminoacid_name: read_molecule('assets/' + aminoacid_name + '.pdb'), aminoacids_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids = load_aminoacids()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>0.080</td>\n",
       "      <td>-0.083</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.539</td>\n",
       "      <td>-0.033</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.032</td>\n",
       "      <td>1.394</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.268</td>\n",
       "      <td>2.339</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.116</td>\n",
       "      <td>-0.818</td>\n",
       "      <td>-1.226</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.743</td>\n",
       "      <td>-2.188</td>\n",
       "      <td>-1.162</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.654</td>\n",
       "      <td>-0.842</td>\n",
       "      <td>-1.361</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1H</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.523</td>\n",
       "      <td>0.813</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2H</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.461</td>\n",
       "      <td>-1.018</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.904</td>\n",
       "      <td>-0.508</td>\n",
       "      <td>0.929</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.691</td>\n",
       "      <td>-0.378</td>\n",
       "      <td>-2.154</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.066</td>\n",
       "      <td>-2.592</td>\n",
       "      <td>-1.973</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1HG2</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>3.986</td>\n",
       "      <td>-1.433</td>\n",
       "      <td>-2.235</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2HG2</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.083</td>\n",
       "      <td>0.168</td>\n",
       "      <td>-1.502</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3HG2</td>\n",
       "      <td>3</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.138</td>\n",
       "      <td>-1.279</td>\n",
       "      <td>-0.467</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.452</td>\n",
       "      <td>1.567</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.911</td>\n",
       "      <td>0.645</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name element aminoacid      x      y      z  h  residue_seq_number\n",
       "0      N               THR  0.080 -0.083  0.000  1                   1\n",
       "1     CA               THR  1.539 -0.033  0.000  1                   1\n",
       "2      C               THR  2.032  1.394  0.000  1                   1\n",
       "3      O               THR  1.268  2.339  0.000  1                   1\n",
       "4     CB               THR  2.116 -0.818 -1.226  1                   1\n",
       "5    OG1               THR  1.743 -2.188 -1.162  1                   1\n",
       "6    CG2               THR  3.654 -0.842 -1.361  1                   1\n",
       "7     1H       1       THR -0.523  0.813  0.000  1                   1\n",
       "8     2H       2       THR -0.461 -1.018  0.000  1                   1\n",
       "9     HA               THR  1.904 -0.508  0.929  1                   1\n",
       "10    HB               THR  1.691 -0.378 -2.154  1                   1\n",
       "11   HG1               THR  2.066 -2.592 -1.973  1                   1\n",
       "12  1HG2       1       THR  3.986 -1.433 -2.235  1                   1\n",
       "13  2HG2       2       THR  4.083  0.168 -1.502  1                   1\n",
       "14  3HG2       3       THR  4.138 -1.279 -0.467  1                   1\n",
       "15    OC               THR  3.452  1.567  0.000  1                   1\n",
       "16    HC               THR  3.911  0.645  0.000  1                   1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(aminoacids[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### III) Infer the position of multiple amino acids based on the peptide bond.\n",
    "\n",
    "\n",
    "* After the nucleotilic attack the resonance delocalization of electrons in the peptide bond forces the creation of a plane where the atoms will be located (amide plan)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* First, We define how translations and rotations are made.\n",
    "\n",
    "\n",
    "* Homogeneous coordinates were used throughout this work.\n",
    "![assets/rotation.png](assets/rotation.png)\n",
    "![assets/translation.png](assets/translation.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def translation_matrix(delta_x, delta_y, delta_z, ref_x=.0, ref_y=.0, ref_z=.0):\n",
    "    T = np.eye(4)\n",
    "    T[0, -1], T[1, -1], T[2, -1] = delta_x - ref_x, delta_y - ref_y, delta_z - ref_z\n",
    "    return T\n",
    "\n",
    "def rotation_matrix(x_rotation_rad, y_rotation_rad, z_rotation_rad):\n",
    "    Rx, Ry, Rz = np.eye(4), np.eye(4), np.eye(4)\n",
    "\n",
    "    Rx[1, 1], Rx[1, 2], Rx[2, 1], Rx[2, 2] =\\\n",
    "        np.cos(x_rotation_rad), -np.sin(x_rotation_rad), np.sin(x_rotation_rad), np.cos(x_rotation_rad)\n",
    "\n",
    "    Ry[0, 0], Ry[0, 2], Ry[2, 0], Rx[2, 2] =\\\n",
    "        np.cos(y_rotation_rad), np.sin(y_rotation_rad), -np.sin(y_rotation_rad), np.cos(y_rotation_rad)\n",
    "\n",
    "    Rz[0, 0], Rx[0, 1], Rx[1, 0], Rx[1, 1] =\\\n",
    "        np.cos(z_rotation_rad), -np.sin(z_rotation_rad), np.sin(z_rotation_rad), np.cos(z_rotation_rad)\n",
    "\n",
    "    return np.dot(np.dot(Rx, Ry), Rz)\n",
    "\n",
    "def apply_transformation(transformation, df):\n",
    "    df[['x', 'y', 'z', 'h']] = transformation.dot(df[['x', 'y', 'z', 'h']].T).T\n",
    "\n",
    "def apply_rotation(transformation, df, atom_reference='N'):\n",
    "    position = df[df.name == atom_reference]\n",
    "    apply_transformation(translation_matrix(-position.x, -position.y, -position.z), df)\n",
    "    apply_transformation(transformation, df)\n",
    "    apply_transformation(translation_matrix(position.x, position.y, position.z), df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Some utility functions for calculating the euclidean distance and the angle between two vectors:\n",
    "\n",
    "\n",
    "$$distance = \\sqrt{|(A - B)|}$$\n",
    "\n",
    "\n",
    "$$cos(\\alpha) = \\frac{A \\cdot B}{|A||B|}$$ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 582,
   "metadata": {},
   "outputs": [],
   "source": [
    "def euclidean_distance(p1, p2):\n",
    "    return np.sqrt(np.sum((p1 - p2) ** 2))\n",
    "\n",
    "def l2_norm(p):\n",
    "    return np.sqrt(np.sum(p ** 2))\n",
    "\n",
    "def get_angle_rad(a, b, normal_reference):\n",
    "    negative = np.all(np.dot(np.cross(a, b), normal_reference) < 0)\n",
    "    angle = np.arccos(((a).dot(b))/(l2_norm(a) * l2_norm(b))) \n",
    "    angle = angle if not np.isnan(angle) else .0\n",
    "    return angle if not negative else -angle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We will consider the omega torsion angle of `Proline` to be close to zero degrees for the `cis` configuration, and, most often and 180 degrees for the trans configuration.\n",
    "\n",
    "\n",
    "![assets/rotation.png](assets/ct.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_omega(current_aminoacid):\n",
    "    return 0 if 'PRO' in current_aminoacid.aminoacid.values else 180\n",
    "\n",
    "def get_desired_OCN_angle(current_aminoacid):\n",
    "    return 123 if 'PRO' in current_aminoacid.aminoacid.values else 123.5\n",
    "\n",
    "def get_desired_CNCa_angle(current_aminoacid):\n",
    "    return 126 if 'PRO' in current_aminoacid.aminoacid.values else 122\n",
    "\n",
    "def get_desired_bond_length(current_aminoacid):\n",
    "    return 1.32 if 'PRO' in current_aminoacid.aminoacid.values else 1.33"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "________________________________\n",
    "* This function infers the position at which the amide nitrogen should be, so we can find a translation of the \"next\" aminoacid.\n",
    "\n",
    "\n",
    "* It return a position that is equals to the rotation of the oxygen inside the plane C$\\alpha$-C-N, and uses the length of the peptide bond (1.33Å or 1.32Å) and the Angle (123.5 or 123)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_nitrogen_location(aminoacid):\n",
    "    alpha_carbon = aminoacid[aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    carboxyl_carbon = aminoacid[aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    oxygen_double_bonded = aminoacid[aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "\n",
    "    CO = oxygen_double_bonded - carboxyl_carbon\n",
    "    CA = alpha_carbon - carboxyl_carbon\n",
    "\n",
    "    plane_normal = np.cross(CO, CA)\n",
    "    plane_vector_basis = np.array([\n",
    "        np.cross(CO, plane_normal) / l2_norm(np.cross(CO, plane_normal)), \n",
    "        CO / l2_norm(CO), \n",
    "        plane_normal / l2_norm(plane_normal)]).T\n",
    "\n",
    "    CO_new_norm = CO * get_desired_bond_length(aminoacid) / l2_norm(CO)\n",
    "    CO_new_norm_plane = np.linalg.solve(plane_vector_basis, CO_new_norm)[:2]\n",
    "    \n",
    "    rotation_angle = np.radians(get_desired_OCN_angle(aminoacid))\n",
    "    rotation_angle_matrix = np.array(\n",
    "        [[np.cos(rotation_angle), -np.sin(rotation_angle)], \n",
    "         [np.sin(rotation_angle), np.cos(rotation_angle)]])\n",
    "\n",
    "    CN_plane = CO_new_norm_plane @ rotation_angle_matrix\n",
    "    CN = np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array([CN_plane[0], CN_plane[1], 0]))\n",
    "    return CN + carboxyl_carbon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C'):\n",
    "    if terminal == 'C':\n",
    "        alpha_carbon = previous_aminoacid[previous_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        oxygen_double_bonded = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C1 = oxygen_double_bonded - carboxyl_carbon\n",
    "        C2 = alpha_carbon - carboxyl_carbon\n",
    "    if terminal == 'N':\n",
    "        last_carbon = current_aminoacid[current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        nitrogen_amide = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C2 = last_carbon - nitrogen_amide\n",
    "        C1 = carboxyl_carbon - nitrogen_amide\n",
    "    return np.cross(C1, C2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    desired_nitrogen_location = get_desired_nitrogen_location(previous_aminoacid)\n",
    "    current_nitrogen_location = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        \n",
    "    required_tranlation = desired_nitrogen_location - current_nitrogen_location\n",
    "    \n",
    "    required_tranlation_matrix = translation_matrix(required_tranlation[0], required_tranlation[1], required_tranlation[2])\n",
    "    apply_transformation(required_tranlation_matrix, current_aminoacid)\n",
    "    \n",
    "def rotate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    omega = get_desired_omega(current_aminoacid)\n",
    "    previous_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C')\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_x = get_angle_rad(current_normal[[1, 2]], previous_normal[[1, 2]], \n",
    "                                     -previous_normal[[1, 2]])\n",
    "    apply_rotation(rotation_matrix(rotation_angle_x, 0, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_y = get_angle_rad(current_normal[[0, 2]], previous_normal[[0, 2]], \n",
    "                                     -previous_normal[[0, 2]])\n",
    "    apply_rotation(rotation_matrix(0, rotation_angle_y, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_z = get_angle_rad(current_normal[[0, 1]], previous_normal[[0, 1]],\n",
    "                                     -previous_normal[[0, 1]])\n",
    "    apply_rotation(rotation_matrix(0, 0, rotation_angle_z), current_aminoacid)\n",
    "\n",
    "def rotate_for_CNCa_angle(previous_aminoacid, current_aminoacid):\n",
    "    carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    nitrogen_amide = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    alpha_carbon = current_aminoacid[current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "\n",
    "    NC = carboxyl_carbon - nitrogen_amide\n",
    "    NCa = alpha_carbon - nitrogen_amide\n",
    "\n",
    "    plane_normal = np.cross(NCa, NC)\n",
    "    plane_vector_basis = np.array([\n",
    "        np.cross(NC, plane_normal) / l2_norm(np.cross(NC, plane_normal)), \n",
    "        NC / l2_norm(NC),\n",
    "        plane_normal / l2_norm(plane_normal)]).T\n",
    "\n",
    "    current_CNCa_angle_rad = get_angle_rad(NC, NCa, -plane_normal)\n",
    "    desired_CNCa_angle_rad = np.radians(get_desired_CNCa_angle(current_aminoacid))\n",
    "    \n",
    "    needed_rotation_rad = desired_CNCa_angle_rad - current_CNCa_angle_rad\n",
    "    print('needed_rotation', np.degrees(needed_rotation_rad))\n",
    "    \n",
    "    new_current_aminoacid = current_aminoacid.copy()\n",
    "    for i, atom in new_current_aminoacid.iterrows():\n",
    "        new_current_aminoacid.loc[i, ['x', 'y', 'z']] =\\\n",
    "            np.linalg.solve(plane_vector_basis, np.array(atom.loc[['x', 'y', 'z']].values.reshape(-1,), dtype='float'))\n",
    "    \n",
    "    apply_rotation(rotation_matrix(0, 0, needed_rotation_rad), new_current_aminoacid)\n",
    "\n",
    "    for i, atom in new_current_aminoacid.iterrows():\n",
    "        new_current_aminoacid.loc[i, ['x', 'y', 'z']] =\\\n",
    "            np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array(atom.loc[['x', 'y', 'z']].values.reshape(-1,), dtype='float'))\n",
    "    \n",
    "    new_alpha_carbon = new_current_aminoacid[new_current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    NC = carboxyl_carbon - nitrogen_amide\n",
    "    NCa = new_alpha_carbon - nitrogen_amide\n",
    "    print('NCa', NCa)\n",
    "    print(np.degrees(get_angle_rad(NC, NCa, -plane_normal)))\n",
    "#     CN = np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array([CN_plane[0], CN_plane[1], 0]))\n",
    "#     return CN + carboxyl_carbon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 581,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dehydration(aminoacid1, aminoacid2):\n",
    "    aminoacid1 = aminoacid1[~aminoacid1.name.isin(['OC', 'HC'])]    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__________________________\n",
    "* Functions to verify the constrains regarding angles and distances in the amide plane."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_peptide_bond_length(previous_aminoacid, current_aminoacid):\n",
    "    a = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    b = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('euclidean_distance for the peptide bond between %s and %s: %.2f' % \n",
    "          (previous_aminoacid.aminoacid[0], current_aminoacid.aminoacid[0], euclidean_distance(a, b)))\n",
    "    \n",
    "def check_OCN_angle(previous_aminoacid, current_aminoacid):\n",
    "    c = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    n = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    o = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('O-C-N Angle for the peptide bond between %s and %s:        %.2f' % \n",
    "          (previous_aminoacid.aminoacid[0], current_aminoacid.aminoacid[0], abs(180*get_angle_rad(o-c, n-c, (o-c))/np.pi)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* The actual translations and rotations happens below:\n",
    "    * We first translate the \"next\" aminoacid and then we do the rotations so the angles and distances can be as depicted below:\n",
    "   \n",
    "![assets/rotation.png](assets/pep.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCa [ 0.271 -2.372  0.   ]\n",
      "needed_rotation 70.89484381311769\n",
      "NCa [-0.75771099 -1.95440275  0.        ]\n",
      "23.396365497856767\n",
      "euclidean_distance for the peptide bond between THR and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and THR:        123.50\n",
      "NCa [ 0.393 -2.377  0.   ]\n",
      "needed_rotation 70.24844712336981\n",
      "NCa [-0.67824914 -1.98937919  0.        ]\n",
      "23.537470128605243\n",
      "euclidean_distance for the peptide bond between THR and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and CYS:        123.50\n",
      "NCa [ 0.393 -2.377  0.   ]\n",
      "needed_rotation 70.9539567695964\n",
      "NCa [-0.69562587 -2.00467564  0.        ]\n",
      "22.521142529575304\n",
      "euclidean_distance for the peptide bond between CYS and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and CYS:        123.50\n",
      "NCa [ 0.20119612 -2.45571957  0.76193793]\n",
      "needed_rotation 78.27389356391811\n",
      "NCa [-0.89058777 -2.09183837  0.76193793]\n",
      "21.317163583109995\n",
      "euclidean_distance for the peptide bond between CYS and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and PRO:        123.50\n",
      "NCa [-1.13719168  1.64556719 -2.35831239]\n",
      "needed_rotation 91.44554494173673\n",
      "NCa [-1.58734388  2.37064883 -2.35831239]\n",
      "24.318144872393617\n",
      "euclidean_distance for the peptide bond between PRO and SER: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and SER:        123.00\n",
      "NCa [1.97458844 2.39818074 1.53708978]\n",
      "needed_rotation 114.1773562887383\n",
      "NCa [4.08123098 3.72879003 1.53708978]\n",
      "17.381656103773715\n",
      "euclidean_distance for the peptide bond between SER and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between SER and ILE:        123.50\n",
      "NCa [ 1.79532996 -1.67965816 -0.81001449]\n",
      "needed_rotation 96.41041814219884\n",
      "NCa [-0.21510786 -2.69751444 -0.81001449]\n",
      "-35.49712246316003\n",
      "euclidean_distance for the peptide bond between ILE and VAL: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and VAL:        123.50\n",
      "NCa [ 0.1491071  -2.39709645 -0.33990856]\n",
      "needed_rotation 82.59457080880041\n",
      "NCa [-1.10459717 -2.20273336 -0.33990856]\n",
      "9.694088005065764\n",
      "euclidean_distance for the peptide bond between VAL and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between VAL and ALA:        123.50\n",
      "NCa [ 2.95076501 -1.10887997 -0.84951471]\n",
      "needed_rotation 118.09479585987121\n",
      "NCa [ 4.19541024 -0.30224936 -0.84951471]\n",
      "16.71692241661521\n",
      "euclidean_distance for the peptide bond between ALA and ARG: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and ARG:        123.50\n",
      "NCa [1.45891302 1.60692423 0.05413389]\n",
      "needed_rotation 113.60206150316911\n",
      "NCa [3.60496577 3.01987527 0.05413389]\n",
      "15.694146840625015\n",
      "euclidean_distance for the peptide bond between ARG and SER: 1.33\n",
      "O-C-N Angle for the peptide bond between ARG and SER:        123.50\n",
      "NCa [ 0.47602372 -2.33693459 -0.79033241]\n",
      "needed_rotation 110.95546837403663\n",
      "NCa [-1.37806426 -3.25092807 -0.79033241]\n",
      "-22.981657009224975\n",
      "euclidean_distance for the peptide bond between SER and ASN: 1.33\n",
      "O-C-N Angle for the peptide bond between SER and ASN:        123.50\n",
      "NCa [-1.06675327 -2.77373407 -1.35893746]\n",
      "needed_rotation 92.88864504122671\n",
      "NCa [-2.2607225  -3.21789499 -1.35893746]\n",
      "17.980531169084504\n",
      "euclidean_distance for the peptide bond between ASN and PHE: 1.33\n",
      "O-C-N Angle for the peptide bond between ASN and PHE:        123.50\n",
      "NCa [ 2.26097432 -1.06602599  1.02532928]\n",
      "needed_rotation 90.90559040544665\n",
      "NCa [ 3.54253105 -0.52425938  1.02532928]\n",
      "35.13404668712779\n",
      "euclidean_distance for the peptide bond between PHE and ASN: 1.33\n",
      "O-C-N Angle for the peptide bond between PHE and ASN:        123.50\n",
      "NCa [ 2.19470345  0.57473947 -0.34011953]\n",
      "needed_rotation 89.78143102695225\n",
      "NCa [ 3.02219261  0.92674426 -0.34011953]\n",
      "28.8684803512882\n",
      "euclidean_distance for the peptide bond between ASN and VAL: 1.33\n",
      "O-C-N Angle for the peptide bond between ASN and VAL:        123.50\n",
      "NCa [ 1.15321012 -2.11651892  0.14036144]\n",
      "needed_rotation 88.28189397434991\n",
      "NCa [-0.46648594 -2.57998543  0.14036144]\n",
      "-17.895376533948173\n",
      "euclidean_distance for the peptide bond between VAL and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between VAL and CYS:        123.50\n",
      "NCa [-2.51224834 -0.21327703  0.80372632]\n",
      "needed_rotation 115.38806195810288\n",
      "NCa [-2.76111151 -0.21256861  0.80372632]\n",
      "6.31777046827768\n",
      "euclidean_distance for the peptide bond between CYS and ARG: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and ARG:        123.50\n",
      "NCa [ 0.44668927 -2.30736112 -0.20882323]\n",
      "needed_rotation 71.11140006868466\n",
      "NCa [ 1.30900481 -2.33992292 -0.20882323]\n",
      "40.23699520064242\n",
      "euclidean_distance for the peptide bond between ARG and LEU: 1.33\n",
      "O-C-N Angle for the peptide bond between ARG and LEU:        123.50\n",
      "NCa [-3.02416245  1.0106382   0.21986946]\n",
      "needed_rotation 112.99113120309914\n",
      "NCa [-3.25063155  1.42230939  0.21986946]\n",
      "8.19737285756967\n",
      "euclidean_distance for the peptide bond between LEU and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between LEU and PRO:        123.50\n",
      "NCa [-1.38152159 -1.88406646 -0.01380112]\n",
      "needed_rotation 92.86214791128552\n",
      "NCa [-1.60188347 -2.06650973 -0.01380112]\n",
      "29.3243658141338\n",
      "euclidean_distance for the peptide bond between PRO and GLY: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLY:        123.00\n",
      "NCa [0.39789066 0.56020108 0.55541286]\n",
      "needed_rotation 100.92314233591621\n",
      "NCa [0.99403344 0.89657898 0.55541286]\n",
      "24.373917790094296\n",
      "euclidean_distance for the peptide bond between GLY and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and THR:        123.50\n",
      "NCa [-1.13982453 -0.46955147  1.01861112]\n",
      "needed_rotation 89.3019798297642\n",
      "NCa [-0.80062541 -0.38144573  1.01861112]\n",
      "28.846593215530557\n",
      "euclidean_distance for the peptide bond between THR and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and PRO:        123.50\n",
      "NCa [ 1.03270732 -2.11943829 -0.06427132]\n",
      "needed_rotation 90.077990096785\n",
      "NCa [ 1.87205415 -2.05904929 -0.06427132]\n",
      "16.01193202467813\n",
      "euclidean_distance for the peptide bond between PRO and GLU: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLU:        123.00\n",
      "NCa [-2.17139556 -0.60999882 -1.04619154]\n",
      "needed_rotation 84.23825801329205\n",
      "NCa [-2.13886037 -0.40260939 -1.04619154]\n",
      "35.69643240175248\n",
      "euclidean_distance for the peptide bond between GLU and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between GLU and ALA:        123.50\n",
      "NCa [ 2.27124112 -0.55840266 -0.10333696]\n",
      "needed_rotation 90.24533593095053\n",
      "NCa [ 2.9204364  -0.34348862 -0.10333696]\n",
      "28.57790303063272\n",
      "euclidean_distance for the peptide bond between ALA and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and ILE:        123.50\n",
      "NCa [-1.27828707 -1.98588363  0.2501202 ]\n",
      "needed_rotation 70.67977394588783\n",
      "NCa [-1.48246513 -1.38566983  0.2501202 ]\n",
      "44.741910138605256\n",
      "euclidean_distance for the peptide bond between ILE and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and CYS:        123.50\n",
      "NCa [ 1.75199739 -1.45673367  1.08502109]\n",
      "needed_rotation 76.88539045381515\n",
      "NCa [ 2.75336124 -1.16643086  1.08502109]\n",
      "32.91194946158623\n",
      "euclidean_distance for the peptide bond between CYS and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and ALA:        123.50\n",
      "NCa [-0.57021395 -2.31648355 -0.18882591]\n",
      "needed_rotation 77.28890312305052\n",
      "NCa [-1.18930455 -1.68712258 -0.18882591]\n",
      "35.117597987737135\n",
      "euclidean_distance for the peptide bond between ALA and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and THR:        123.50\n",
      "NCa [2.45004891 0.65299015 1.48959094]\n",
      "needed_rotation 102.02072752425354\n",
      "NCa [1.84515197 0.24175863 1.48959094]\n",
      "22.882383077436995\n",
      "euclidean_distance for the peptide bond between THR and TYR: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and TYR:        123.50\n",
      "NCa [-2.35066301  0.56820258 -0.2362688 ]\n",
      "needed_rotation 86.20397140693875\n",
      "NCa [-2.39618114  0.30045342 -0.2362688 ]\n",
      "29.39486892094275\n",
      "euclidean_distance for the peptide bond between TYR and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between TYR and THR:        123.50\n",
      "NCa [-0.0029079  -2.38697261  0.32670862]\n",
      "needed_rotation 97.63784195345279\n",
      "NCa [ 1.17743695 -2.24037964  0.32670862]\n",
      "-8.284098310246703\n",
      "euclidean_distance for the peptide bond between THR and GLY: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and GLY:        123.50\n",
      "NCa [2.35965977 0.66240589 2.34517463]\n",
      "needed_rotation 105.58370383000711\n",
      "NCa [1.71127719 0.17654183 2.34517463]\n",
      "24.536743310316396\n",
      "euclidean_distance for the peptide bond between GLY and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and CYS:        123.50\n",
      "NCa [ 1.7779593   2.40255106 -1.22532292]\n",
      "needed_rotation 103.73128020315224\n",
      "NCa [ 3.59989554  3.36530272 -1.22532292]\n",
      "10.011385717461389\n",
      "euclidean_distance for the peptide bond between CYS and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and ILE:        123.50\n",
      "NCa [ 2.03687311 -1.49099405  0.72766579]\n",
      "needed_rotation 94.18768011269027\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCa [ 0.19715986 -2.4282526   0.72766579]\n",
      "-34.98333251851884\n",
      "euclidean_distance for the peptide bond between ILE and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and ILE:        123.50\n",
      "NCa [ 0.44176759 -2.37188529  0.17607086]\n",
      "needed_rotation 80.8285354015082\n",
      "NCa [-0.90386484 -2.25569456  0.17607086]\n",
      "11.337851929385849\n",
      "euclidean_distance for the peptide bond between ILE and ILE: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and ILE:        123.50\n",
      "NCa [-0.33750704 -1.89946756  1.4155066 ]\n",
      "needed_rotation 80.8538725417639\n",
      "NCa [-1.08068173 -1.27477117  1.4155066 ]\n",
      "22.969845227547435\n",
      "euclidean_distance for the peptide bond between ILE and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between ILE and PRO:        123.50\n",
      "NCa [ 0.19985805  2.03833716 -2.0139528 ]\n",
      "needed_rotation 79.49577958000829\n",
      "NCa [-0.41579662  2.71635661 -2.0139528 ]\n",
      "29.841917893309294\n",
      "euclidean_distance for the peptide bond between PRO and GLY: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLY:        123.00\n",
      "NCa [1.52924246 2.5524415  2.20701704]\n",
      "needed_rotation 112.27688941682263\n",
      "NCa [3.56148152 3.45475921 2.20701704]\n",
      "18.92737143314218\n",
      "euclidean_distance for the peptide bond between GLY and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and ALA:        123.50\n",
      "NCa [ 0.69255377 -2.27478701 -1.129924  ]\n",
      "needed_rotation 114.71776342912952\n",
      "NCa [-1.48217752 -3.46935683 -1.129924  ]\n",
      "-31.46375307572901\n",
      "euclidean_distance for the peptide bond between ALA and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and THR:        123.50\n",
      "NCa [-0.62938381 -2.50245346 -1.90559049]\n",
      "needed_rotation 84.93935797788173\n",
      "NCa [-1.96297289 -2.61769955 -1.90559049]\n",
      "21.2141730022938\n",
      "euclidean_distance for the peptide bond between THR and THR: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and THR:        123.50\n",
      "NCa [ 1.48585112 -1.67761707  1.60895151]\n",
      "needed_rotation 110.721101462538\n",
      "NCa [ 2.8626721  -0.81913647  1.60895151]\n",
      "32.05292464050154\n",
      "euclidean_distance for the peptide bond between THR and CYS: 1.33\n",
      "O-C-N Angle for the peptide bond between THR and CYS:        123.50\n",
      "NCa [ 0.65288114 -1.62528492  1.63422449]\n",
      "needed_rotation 112.7902403902347\n",
      "NCa [-1.56978476 -2.39311854  1.63422449]\n",
      "-32.61191786187281\n",
      "euclidean_distance for the peptide bond between CYS and PRO: 1.33\n",
      "O-C-N Angle for the peptide bond between CYS and PRO:        123.50\n",
      "NCa [-1.56073535  2.50001429  0.95355575]\n",
      "needed_rotation 103.69580077910673\n",
      "NCa [-0.28295146  2.5185092   0.95355575]\n",
      "31.46941796934875\n",
      "euclidean_distance for the peptide bond between PRO and GLY: 1.32\n",
      "O-C-N Angle for the peptide bond between PRO and GLY:        123.00\n",
      "NCa [ 1.83589081  1.12297968 -2.42368724]\n",
      "needed_rotation 96.40269012782497\n",
      "NCa [ 3.90972022  2.19936038 -2.42368724]\n",
      "5.9091310293411485\n",
      "euclidean_distance for the peptide bond between GLY and ASP: 1.33\n",
      "O-C-N Angle for the peptide bond between GLY and ASP:        123.50\n",
      "NCa [ 2.45494558 -0.12872767 -0.13282228]\n",
      "needed_rotation 104.399962174507\n",
      "NCa [ 2.46194781 -0.16154536 -0.13282228]\n",
      "18.077968127186686\n",
      "euclidean_distance for the peptide bond between ASP and TYR: 1.33\n",
      "O-C-N Angle for the peptide bond between ASP and TYR:        123.50\n",
      "NCa [ 0.01283459 -2.40338117 -0.11010437]\n",
      "needed_rotation 80.51636896190043\n",
      "NCa [-1.04624495 -2.10142617 -0.11010437]\n",
      "18.02936820152998\n",
      "euclidean_distance for the peptide bond between TYR and ALA: 1.33\n",
      "O-C-N Angle for the peptide bond between TYR and ALA:        123.50\n",
      "NCa [ 2.34657034 -0.22148579  2.29763128]\n",
      "needed_rotation 100.72585705055228\n",
      "NCa [ 2.70628101 -0.04723952  2.29763128]\n",
      "25.282527452910813\n",
      "euclidean_distance for the peptide bond between ALA and ASN: 1.33\n",
      "O-C-N Angle for the peptide bond between ALA and ASN:        123.50\n",
      "last aminoacid:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>21.596354</td>\n",
       "      <td>14.948196</td>\n",
       "      <td>-7.701691</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>21.388192</td>\n",
       "      <td>14.833382</td>\n",
       "      <td>-7.833907</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>19.911009</td>\n",
       "      <td>14.912137</td>\n",
       "      <td>-9.247973</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>19.041622</td>\n",
       "      <td>15.054867</td>\n",
       "      <td>-10.131538</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>20.902537</td>\n",
       "      <td>14.828958</td>\n",
       "      <td>-5.881166</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CG</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>22.271507</td>\n",
       "      <td>14.753246</td>\n",
       "      <td>-4.289677</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>OD1</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>21.986287</td>\n",
       "      <td>14.860724</td>\n",
       "      <td>-3.140273</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ND2</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>23.808440</td>\n",
       "      <td>14.567608</td>\n",
       "      <td>-4.178466</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1H</td>\n",
       "      <td>1</td>\n",
       "      <td>ASN</td>\n",
       "      <td>20.769576</td>\n",
       "      <td>15.072002</td>\n",
       "      <td>-8.535599</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2H</td>\n",
       "      <td>2</td>\n",
       "      <td>ASN</td>\n",
       "      <td>22.586966</td>\n",
       "      <td>14.914417</td>\n",
       "      <td>-6.763543</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HA</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>22.717123</td>\n",
       "      <td>14.685621</td>\n",
       "      <td>-8.326296</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HO</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>18.107965</td>\n",
       "      <td>15.084017</td>\n",
       "      <td>-11.014317</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1HB</td>\n",
       "      <td>1</td>\n",
       "      <td>ASN</td>\n",
       "      <td>19.528680</td>\n",
       "      <td>14.976194</td>\n",
       "      <td>-5.436053</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2HB</td>\n",
       "      <td>2</td>\n",
       "      <td>ASN</td>\n",
       "      <td>20.707109</td>\n",
       "      <td>14.745371</td>\n",
       "      <td>-5.963823</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1HD2</td>\n",
       "      <td>1</td>\n",
       "      <td>ASN</td>\n",
       "      <td>24.712962</td>\n",
       "      <td>14.533488</td>\n",
       "      <td>-3.205561</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2HD2</td>\n",
       "      <td>2</td>\n",
       "      <td>ASN</td>\n",
       "      <td>24.063233</td>\n",
       "      <td>14.482555</td>\n",
       "      <td>-5.228295</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>OC</td>\n",
       "      <td></td>\n",
       "      <td>ASN</td>\n",
       "      <td>19.583469</td>\n",
       "      <td>14.810824</td>\n",
       "      <td>-9.498266</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name element aminoacid          x          y          z    h  \\\n",
       "0      N               ASN  21.596354  14.948196  -7.701691  1.0   \n",
       "1     CA               ASN  21.388192  14.833382  -7.833907  1.0   \n",
       "2      C               ASN  19.911009  14.912137  -9.247973  1.0   \n",
       "3      O               ASN  19.041622  15.054867 -10.131538  1.0   \n",
       "4     CB               ASN  20.902537  14.828958  -5.881166  1.0   \n",
       "5     CG               ASN  22.271507  14.753246  -4.289677  1.0   \n",
       "6    OD1               ASN  21.986287  14.860724  -3.140273  1.0   \n",
       "7    ND2               ASN  23.808440  14.567608  -4.178466  1.0   \n",
       "8     1H       1       ASN  20.769576  15.072002  -8.535599  1.0   \n",
       "9     2H       2       ASN  22.586966  14.914417  -6.763543  1.0   \n",
       "10    HA               ASN  22.717123  14.685621  -8.326296  1.0   \n",
       "11    HO               ASN  18.107965  15.084017 -11.014317  1.0   \n",
       "12   1HB       1       ASN  19.528680  14.976194  -5.436053  1.0   \n",
       "13   2HB       2       ASN  20.707109  14.745371  -5.963823  1.0   \n",
       "14  1HD2       1       ASN  24.712962  14.533488  -3.205561  1.0   \n",
       "15  2HD2       2       ASN  24.063233  14.482555  -5.228295  1.0   \n",
       "16    OC               ASN  19.583469  14.810824  -9.498266  1.0   \n",
       "\n",
       "    residue_seq_number  \n",
       "0                    1  \n",
       "1                    1  \n",
       "2                    1  \n",
       "3                    1  \n",
       "4                    1  \n",
       "5                    1  \n",
       "6                    1  \n",
       "7                    1  \n",
       "8                    1  \n",
       "9                    1  \n",
       "10                   1  \n",
       "11                   1  \n",
       "12                   1  \n",
       "13                   1  \n",
       "14                   1  \n",
       "15                   1  \n",
       "16                   1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aminoacids_df = load_aminoacids()\n",
    "for i in range(1, len(aminoacids_df)):\n",
    "    current_aminoacid = aminoacids_df[i]\n",
    "    previous_aminoacid = aminoacids_df[i - 1]\n",
    "    \n",
    "    dehydration(previous_aminoacid, current_aminoacid)\n",
    "    translate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "    rotate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "    rotate_for_CNCa_angle(previous_aminoacid, current_aminoacid)\n",
    "    \n",
    "    # check_peptide_bond_length(previous_aminoacid, current_aminoacid)\n",
    "    # check_OCN_angle(previous_aminoacid, current_aminoacid)\n",
    "\n",
    "print('last aminoacid:')\n",
    "display(current_aminoacid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We have all the atoms aligned accordingly. So, let's create string rows for the pdb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 585,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_atom_to_file(atom_row, aminoacid_seq_number, fd):\n",
    "    global atom_seq_number\n",
    "    atom_seq_number += 1\n",
    "    if 'H' not in atom_row[0]:\n",
    "        atom_row.element = ''.join(filter(str.isalpha, atom_row[0]))[0]\n",
    "        template_row =\\\n",
    "        'ATOM    '+\\\n",
    "        ('%d'%atom_seq_number).rjust(3)+\\\n",
    "        '  '+\\\n",
    "        ('%s'%atom_row[0]).rjust(3)+\\\n",
    "        ' '+\\\n",
    "        ('%s'%atom_row.aminoacid).rjust(3)+\\\n",
    "        ' A'+\\\n",
    "        ('%d'%aminoacid_seq_number).rjust(4)+\\\n",
    "        '    '+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.y).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        '  1.00  5.99           '+\\\n",
    "        ('%s\\n'%atom_row.element)\n",
    "        f.write(template_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATOM      1    N THR A   0       0.080  -0.083   0.080  1.00  5.99           N\n",
      "\n",
      "ATOM      2   CA THR A   0       1.539  -0.033   1.539  1.00  5.99           C\n",
      "\n",
      "ATOM      3    C THR A   0       2.032   1.394   2.032  1.00  5.99           C\n",
      "\n",
      "ATOM      4    O THR A   0       1.268   2.339   1.268  1.00  5.99           O\n",
      "\n",
      "ATOM      5   CB THR A   0       2.116  -0.818   2.116  1.00  5.99           C\n",
      "\n",
      "...\n"
     ]
    }
   ],
   "source": [
    "atom_seq_number = 0\n",
    "\n",
    "with open('assets/final_inferred_molecule.pdb', 'w') as f:\n",
    "    for idx, aminoacid in enumerate(aminoacids_df):\n",
    "        for atom_row in aminoacid.iterrows():\n",
    "            write_atom_to_file(atom_row[1], idx, f)\n",
    "\n",
    "with open(\"assets/final_inferred_molecule.pdb\") as f:\n",
    "    head = [next(f) for x in range(5)]\n",
    "print('\\n'.join(head))\n",
    "print('...')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n",
    "_________________________________________________________________________________\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## b) Receive PDB file and list $\\phi$ and $\\psi$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$ ψ  = Angle( Normal(N_{i}, C_{\\alpha i}, C_{i}); Normal(C_{\\alpha i}, C_{i}, N_{i+1}) ) $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$φ = Angle( Normal(C_{i-1}, N_{i}, C_{\\alpha i}); Normal(N_{i}, C_{\\alpha i}, C_{i}) )$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![assets/phi_psi.png](assets/phi_psi.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_psi(ni, cai, ci, ni1):\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    normal_ci = np.cross((cai - ci), (ni1 - ci))\n",
    "    return np.degrees(get_angle_rad(normal_cai, normal_ci, (ci-cai)))\n",
    "\n",
    "def compute_phi(ci_prev, ni, cai, ci):\n",
    "    normal_ni = np.cross((ci_prev - ni), (cai - ni))\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    return np.degrees(get_angle_rad(normal_ni, normal_cai, -(ni-cai)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_phy_psi(molecule):\n",
    "    molecule['phi'] = None\n",
    "    molecule['psi'] = None\n",
    "\n",
    "    data = []\n",
    "\n",
    "    for i in range(1, len(molecule.residue_seq_number.unique())):\n",
    "        aminoacid_i = molecule[molecule.residue_seq_number == (i)]\n",
    "        ni = aminoacid_i[aminoacid_i.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        cai = aminoacid_i[aminoacid_i.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        ci = aminoacid_i[aminoacid_i.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        angle_data = {'residue_name': aminoacid_i.aminoacid.values[0], 'phi': None, 'psi': None}\n",
    "        try:\n",
    "            aminoacid_prev = molecule[molecule.residue_seq_number == (i-1)]\n",
    "            ci_prev = aminoacid_prev[aminoacid_prev.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['phi'] = compute_phi(ci_prev, ni, cai, ci)\n",
    "        except: pass\n",
    "\n",
    "        try:\n",
    "            aminoacid_i1 = molecule[molecule.residue_seq_number == (i+1)]\n",
    "            ni1 = aminoacid_i1[aminoacid_i1.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['psi'] = compute_psi(ni, cai, ci, ni1)\n",
    "        except: pass\n",
    "\n",
    "        data.append(angle_data)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 592,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:9: RuntimeWarning: invalid value encountered in arccos\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>179.999999</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>179.999999</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>180.000000</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>180.000000</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>180.000000</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>179.999999</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>180.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          phi         psi residue_name\n",
       "0  179.999999  180.000000          THR\n",
       "1  179.999999  179.999999          CYS\n",
       "2    0.000000    0.000000          CYS\n",
       "3  180.000000  180.000000          PRO\n",
       "4    0.000000  180.000000          SER\n",
       "5  180.000000  179.999999          ILE\n",
       "6  180.000000  180.000000          VAL\n",
       "7  179.999999  180.000000          ALA\n",
       "8    0.000000  179.999999          ARG\n",
       "9  180.000000    0.000000          SER"
      ]
     },
     "execution_count": 592,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/final_inferred_molecule.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## c) Ramachandran "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>147.660026</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-107.832573</td>\n",
       "      <td>144.346241</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-131.170389</td>\n",
       "      <td>133.291143</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-118.896523</td>\n",
       "      <td>151.202717</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-76.181506</td>\n",
       "      <td>-18.979386</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-157.924288</td>\n",
       "      <td>166.027161</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-63.579524</td>\n",
       "      <td>-42.100632</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-55.930329</td>\n",
       "      <td>-44.561332</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-61.364845</td>\n",
       "      <td>-43.779138</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-63.193412</td>\n",
       "      <td>-43.303486</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          phi         psi residue_name\n",
       "0         NaN  147.660026          THR\n",
       "1 -107.832573  144.346241          THR\n",
       "2 -131.170389  133.291143          CYS\n",
       "3 -118.896523  151.202717          CYS\n",
       "4  -76.181506  -18.979386          PRO\n",
       "5 -157.924288  166.027161          SER\n",
       "6  -63.579524  -42.100632          ILE\n",
       "7  -55.930329  -44.561332          VAL\n",
       "8  -61.364845  -43.779138          ALA\n",
       "9  -63.193412  -43.303486          ARG"
      ]
     },
     "execution_count": 593,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/1crn.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 594,
   "metadata": {},
   "outputs": [],
   "source": [
    "phi_psi = phi_psi.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 595,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnIAAAJSCAYAAABHgiijAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHdtJREFUeJzt3WGMZfdZ3/Hfg1e4NE6UIDsbYw+1qYxQAsIiq5ASgcYiEGd54TiSW6dqkwLVJjRpK+iLJgoqtJGBokKk0OB0KSaOKnCtUicWWQhJpCFV5ZDYxU3sUMPaCezGVkwShrDBmHjy74u52w72zO7cYWfOfWY+H2k0M+eeO/vs38d3v3vOvXtrjBEAAPr5mqkHAABgZ4QcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApg5NPcBeufTSS8dVV1019Rhz+/KXv5xnPetZU4/RjnWbnzWb3yN/8uWsra3lmhc8Z+pRWnGs7Yx1m1/nNbvvvvs+P8a47Hz7HZiQu+qqq3LvvfdOPcbcVlZWsry8PPUY7Vi3+Vmz+f2D/3RPVldX84F//cqpR2nFsbYz1m1+ndesqv5oO/u5tAoA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADR1aOoBDpJTp05te9+lpaVdnORvbj/9XqZ04sSJbe979OjRXZwEgI6ckQMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATR2aeoCDpMObx586dWrqEQ6Uo0ePTj0CAI05IwcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAUwsRclV1W1U9XlUPbNj2k1X12aq6f/ZxdMNtb6mqk1X1UFW9YpqpAQCmtRAhl+TdSa7fZPvbxxjXzj5OJElVvTDJzUleNLvPL1bVRXs2KQDAgliIkBtjfCTJF7e5+w1J7hhjPDnG+HSSk0lesmvDAQAsqEV/i643VdVrk9yb5F+NMf40yRVJPrphn9Ozbc9QVceSHEuSw4cPZ2VlZXen3QVnzpzZ07m/8pWvXPCf+fDDD1/wn3k+e71u+4E1m9/q6hNZW1uzbnNyrO2MdZvfQVizRQ65W5O8LcmYff65JD+UpDbZd2z2A8YYx5McT5IjR46M5eXlXRl0N62srGQv596N91qd4j1m93rd9gNrNr9bH7onq6ur1m1OjrWdsW7zOwhrtrAhN8b43Nmvq+qXkvzG7NvTSTaWwZVJHt3D0fa1KaLrrO1G5JQzbteUv5f9tI5T2s46Pvnkk3swCcDWFuI5cpupqss3fHtjkrOvaL07yc1VdXFVXZ3kmiQf2+v5AACmthBn5Krq15IsJ7m0qk4n+Ykky1V1bdYvm34myeuTZIzxYFXdmeRTSZ5K8sYxxtoUcwMATGkhQm6M8ZpNNv/yOfa/JcktuzcRAMDiW9hLqwAAnJuQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKCphfgHgVkcU75P535678/99Hs5qLbz3/Dii0/niSee2INpADbnjBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCU91rlr5nyPUKnfJ/X/cT6ABwczsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKYOTT0AnOXN3gFgPs7IAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANDUoakHYP87derUtvZbWlra5UkAYH9xRg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoKmFCLmquq2qHq+qBzZs+/qq+mBV/eHs8/Nm26uq3lFVJ6vqE1X1HdNNDgAwnYUIuSTvTnL907a9OcmHxxjXJPnw7PskeWWSa2Yfx5LcukczAgAslIUIuTHGR5J88Wmbb0hy++zr25O8asP294x1H03y3Kq6fG8mBQBYHIemHuAcDo8xHkuSMcZjVfX82fYrkmx8F/bTs22PPf0HVNWxrJ+1y+HDh7OysrKrA++GM2fOtJx7Jx5++OEL9rMO0rpdKNZsfqurT2Rtbc26zcmxtjPWbX4HYc0WOeS2UptsG5vtOMY4nuR4khw5cmQsLy/v4li7Y2VlJR3nnpp1m581m9+tD92T1dVV6zYnx9rOWLf5HYQ1W4hLq1v43NlLprPPj8+2n06ytGG/K5M8usezAQBMbpFD7u4kr5t9/bok79uw/bWzV6++NMmfnb0ECwBwkCzEpdWq+rUky0kurarTSX4iyc8kubOqfjjJHye5abb7iSRHk5xM8hdJfnDPBwYAWAALEXJjjNdscdP3brLvSPLG3Z0IAGDxLfKlVQAAzkHIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJo6NPUALJZTp05ta7+lpaVdngQAOB9n5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU4emHoDFsrS0NPUIAMA2OSMHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANDUwr+zQ1V9JsmfJ1lL8tQY40hVfX2S/5rkqiSfSfL3xxh/OtWMAABT6HJG7roxxrVjjCOz79+c5MNjjGuSfHj2PQDAgdIl5J7uhiS3z76+PcmrJpwFAGASHUJuJPntqrqvqo7Nth0eYzyWJLPPz59sOgCAidQYY+oZzqmqvmGM8WhVPT/JB5P88yR3jzGeu2GfPx1jPG+T+x5LcixJDh8+/OI77rhjr8a+YM6cOZNLLrlk6jHasW7zs2bz++nffSJra2v58e+ybvNwrO2MdZtf5zW77rrr7tvwlLItLfyLHcYYj84+P15VdyV5SZLPVdXlY4zHquryJI9vcd/jSY4nyZEjR8by8vIeTX3hrKyspOPcU7Nu87Nm87v1oXuyurpq3ebkWNsZ6za/g7BmC31ptaqeVVXPPvt1ku9P8kCSu5O8brbb65K8b5oJAQCms+hn5A4nuauqkvVZf3WM8VtV9fEkd1bVDyf54yQ3TTgjAMAkFjrkxhiPJPn2TbZ/Icn37v1EAACLY6EvrQIAsDUhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApg5NPQAAcG6nTp3KV77ylZw6deq8+y4tLe3BRCwKZ+QAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATbUNuaq6vqoeqqqTVfXmqecBANhr532v1ap67ezLM0keGWPcv7sjnV9VXZTknUm+L8npJB+vqrvHGJ+adrKDYzvv93dWh/f9O3HixLb2O3r06C5PAnS33cfHeR4bl5aW8vDDD7d4PGVvnTfkklw3+/y3k3xLVY0k/3SMce/ujXVeL0lycozxSJJU1R1Jbkgi5ACAA+O8ITfG+MGN31fVDyR5b1V9+xjjC7s22bldkWTjX3lOJ/nOc97j83+Y/MoP7OZMu+La1dXk08+deoxnuOzJv9z+zhf/rd0bZAvzrtt3fnGbh/Ln3rnDiRbfoh5ri+zffOHP8tRTTyW/8h+nHqWV/X6sbfvxcc7Hxv2+brvhIKzZdi6t/tgmm1eT/Puq+tQY4+cv/FjnVZtsG8/YqepYkmNJ8q2XX5zV1dXdnuuCW1tbW8i5vzq+uu19//KJOaLvApl33cZ4xuGzqUX8b3GhLOqxtsieemotGfv7uNgN+/1Y2+7j47yPjft93XbDQViz7VxaffYm274m65daL7mw42zb6SQbnyhwZZJHn77TGON4kuNJcuTIkfHcH/2fezPdBbSyspLl5eWpx3iGRX+O3Lzr5jlyi3usLbLX/6d7srq6mg/86CunHqWV/X6s7cZz5JL9v267ofWa/dhm56yeaTuXVv/txu+r6vuyfpbrX44x/mRHw/3NfTzJNVV1dZLPJrk5yT+caBYAgEls59LqbbMvvy7Jt8zuc9OEEZcxxlNV9aYkH0hyUZLbxhgPTjUPAMAUtnNp9Xdmn7+c5JEkvze2+4SiXTTGOJFke9fDAAD2oe1cWr19LwYBAGA+bd/ZAQDgoBNyAABNCTkAgKaEHABAU9t51So8w3574+YL/Q/97tY/CAosPv9fs5eckQMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJryXqssjP30/qSf/OQnt7Vfh98LAIvLGTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKAp77XKwthP7zv6bd/2bdvab7vvLzuP/bSOAJybM3IAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoKlDUw8AbM+73vWube33spe9bFv7HT169G8yDgALwBk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKe+1CrtgaWlpW/udOnVq2z/zp37qp7a13/vf//5t/0wAenNGDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaMp7rcKEtvuerMn230P16NGjOx0HgGYW9oxcVf1kVX22qu6ffRzdcNtbqupkVT1UVa+Yck4AgKks+hm5t48x/sPGDVX1wiQ3J3lRkm9I8qGq+uYxxtoUAwIATGVhz8idww1J7hhjPDnG+HSSk0leMvFMAAB7btFD7k1V9Ymquq2qnjfbdkWSUxv2OT3bBgBwoNQYY7pfvOpDSV6wyU1vTfLRJJ9PMpK8LcnlY4wfqqp3JrlnjPFfZj/jl5OcGGP8+iY//1iSY0ly+PDhF99xxx278xvZRWfOnMkll1wy9Rjt7Md1+9KXvrSt/Z7znOfs6OfvxzXbbT/9u09kbW0tP/5d1m0ejrWdsW7z67xm11133X1jjCPn22/S58iNMV6+nf2q6peS/Mbs29NJNr7U78okj27x848nOZ4kR44cGcvLyzuedSorKyvpOPfU9uO6nThxYlv77fT3vR/XbLfd+tA9WV1dtW5zcqztjHWb30FYs4W9tFpVl2/49sYkD8y+vjvJzVV1cVVdneSaJB/b6/kAAKa2yK9a/dmqujbrl1Y/k+T1STLGeLCq7kzyqSRPJXmjV6wCAAfRwobcGOMfn+O2W5LcsofjAAAsnIW9tAoAwLkJOQCApoQcAEBTC/scOeCvO3r06Pl3AuBAcUYOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKYmDbmquqmqHqyqr1bVkafd9paqOllVD1XVKzZsv3627WRVvXnvpwYAWAxTn5F7IMmrk3xk48aqemGSm5O8KMn1SX6xqi6qqouSvDPJK5O8MMlrZvsCABw4h6b8xccYv58kVfX0m25IcscY48kkn66qk0leMrvt5Bjjkdn97pjt+6m9mRgAYHFMGnLncEWSj274/vRsW5Kcetr279zqh1TVsSTHkuTw4cNZWVm5sFPugTNnzrSce2rWbX7WbH6rq09kbW3Nus3JsbYz1m1+B2HNdj3kqupDSV6wyU1vHWO8b6u7bbJtZPNLwWOrX3uMcTzJ8SQ5cuTIWF5ePvewC2hlZSUd556adZufNZvfrQ/dk9XVVes2J8fazli3+R2ENdv1kBtjvHwHdzudZGnD91cmeXT29VbbAQAOlKlf7LCVu5PcXFUXV9XVSa5J8rEkH09yTVVdXVVfm/UXRNw94ZwAAJOZ9DlyVXVjkl9IclmS91fV/WOMV4wxHqyqO7P+IoankrxxjLE2u8+bknwgyUVJbhtjPDjR+AAAk5r6Vat3Jblri9tuSXLLJttPJDmxy6MBACy8Rb20CgDAeQg5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmJg25qrqpqh6sqq9W1ZEN26+qqieq6v7Zx7s23PbiqvpkVZ2sqndUVU0zPQDAtKY+I/dAklcn+cgmtz08xrh29vGGDdtvTXIsyTWzj+t3f0wAgMUzaciNMX5/jPHQdvevqsuTPGeMcc8YYyR5T5JX7dqAAAALbOozcudydVX9XlX9TlV992zbFUlOb9jn9GwbAMCBc2i3f4Gq+lCSF2xy01vHGO/b4m6PJfnGMcYXqurFSd5bVS9Kstnz4cY5fu1jWb8Mm8OHD2dlZWWu2RfBmTNnWs49Nes2P2s2v9XVJ7K2tmbd5uRY2xnrNr+DsGa7HnJjjJfv4D5PJnly9vV9VfVwkm/O+hm4KzfsemWSR8/xc44nOZ4kR44cGcvLy/OOMrmVlZV0nHtq1m1+1mx+tz50T1ZXV63bnBxrO2Pd5ncQ1mwhL61W1WVVddHs62/K+osaHhljPJbkz6vqpbNXq742yVZn9QAA9rWp//mRG6vqdJK/l+T9VfWB2U3fk+QTVfW/k/y3JG8YY3xxdtuPJPnPSU4meTjJb+7x2AAAC2HXL62eyxjjriR3bbL915P8+hb3uTfJt+7yaAAAC2/SkINFcerUqW3tt7S0tMuTAMD2LeRz5AAAOD8hBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANHVo6gFgESwtLU09AgDMzRk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFM1xph6hj1RVX+S5I+mnmMHLk3y+amHaMi6zc+a7Yx1m5812xnrNr/Oa/Z3xhiXnW+nAxNyXVXVvWOMI1PP0Y11m5812xnrNj9rtjPWbX4HYc1cWgUAaErIAQA0JeQW3/GpB2jKus3Pmu2MdZufNdsZ6za/fb9mniMHANCUM3IAAE0JuQVRVTdV1YNV9dWqOrJh+1VV9URV3T/7eNeG215cVZ+sqpNV9Y6qqmmmn85W6za77S2ztXmoql6xYfv1s20nq+rNez/1Yqmqn6yqz244xo5uuG3TNcRxNI+q+szsser+qrp3tu3rq+qDVfWHs8/Pm3rOKVXVbVX1eFU9sGHbpmtU694xO/Y+UVXfMd3k09pi3Q7UY5qQWxwPJHl1ko9sctvDY4xrZx9v2LD91iTHklwz+7h+98dcOJuuW1W9MMnNSV6U9XX5xaq6qKouSvLOJK9M8sIkr5nte9C9fcMxdiLZeg2nHHJROI525LrZ8XX2L1xvTvLhMcY1ST48+/4ge3ee+Ri+1Rq9Mv//cf9Y1v8sOKjenc3/7Dswj2lCbkGMMX5/jPHQdvevqsuTPGeMcc9Yf6Lje5K8atcGXFDnWLcbktwxxnhyjPHpJCeTvGT2cXKM8cgY46+S3DHbl2faag1xHF0INyS5ffb17TmAj18bjTE+kuSLT9u81RrdkOQ9Y91Hkzx39mfCgbPFum1lXz6mCbkerq6q36uq36mq755tuyLJ6Q37nJ5tY90VSU5t+P7s+my1/aB70+wSzW0bLnFZq61Zm/mMJL9dVfdV1bHZtsNjjMeSZPb5+ZNNt7i2WiPH3/kdmMe0Q1MPcJBU1YeSvGCTm946xnjfFnd7LMk3jjG+UFUvTvLeqnpRks2eD7cvX4K8w3Xban02+8vLvly3jc61hlm/LPO2rK/D25L8XJIfygE6xnbA2sznZWOMR6vq+Uk+WFX/Z+qBmnP8nduBekwTcntojPHyHdznySRPzr6+r6oeTvLNWf+bxJUbdr0yyaMXYs5Fs5N1y/r6LG34fuP6bLV939ruGlbVLyX5jdm351rDg87azGGM8ejs8+NVdVfWL2d9rqouH2M8Nrss+PikQy6mrdbI8XcOY4zPnf36IDymubS64KrqsrNPxqyqb8r6k1sfmZ1m//Oqeuns1aqvTbLV2amD6O4kN1fVxVV1ddbX7WNJPp7kmqq6uqq+NutPfL17wjkn97Tn1tyY9ReQJFuvIY6jbauqZ1XVs89+neT7s36M3Z3kdbPdXhePX5vZao3uTvLa2atXX5rkz85eguXgPaY5I7cgqurGJL+Q5LIk76+q+8cYr0jyPUn+XVU9lWQtyRvGGGef2PkjWX/Fztcl+c3Zx4Gy1bqNMR6sqjuTfCrJU0neOMZYm93nTUk+kOSiJLeNMR6caPxF8bNVdW3WLzF8Jsnrk+Rca3jQjTGechxt2+Ekd63/fTOHkvzqGOO3qurjSe6sqh9O8sdJbppwxslV1a8lWU5yaVWdTvITSX4mm6/RiSRHs/5k/b9I8oN7PvCC2GLdlg/SY5p3dgAAaMqlVQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAO1RVv1xV/2zqOYCDyz8IDDCnqrohyduTPDvJV5J8Nsk/8e4OwF4TcgBzqKq/m+R/ZP1tgf5FknuTfCnJLUm+dT+85Q/Qh0urAPP5/iTvHWP8wdkNY4z/nuSrWX8TboA9I+QA5rfZpYyafQDsGSEHMJ8PJblxdok1yf97ztyhJH+w5b0AdoHnyAHMqapeneTnk3xdkr9K8sUk/2iM8clJBwMOHCEHsANVVUluS/K/xhi/MPU8wMEk5AAAmvIcOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE39XxoIf5bM7Q88AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "plt.hist2d(phi_psi.phi, phi_psi.psi, bins=(40, 40), cmap=plt.cm.Greys);\n",
    "plt.plot(np.zeros_like(phi_psi.phi), np.linspace(-180, 180, len(phi_psi.phi)))\n",
    "plt.plot(np.linspace(-180, 180, len(phi_psi.phi)), np.zeros_like(phi_psi.phi))\n",
    "plt.xlabel('$\\Phi$')\n",
    "plt.ylabel('$\\Psi$')\n",
    "plt.xlim([-180, 180])\n",
    "plt.ylim([-180, 180])\n",
    "plt.grid()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__________________\n",
    "__________________\n",
    "__________________\n",
    "__________________\n",
    "__________________\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## d) Secondary structure:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*When five or more consecutive amino acids have torsion angles in the $\\alpha$-helical region of the Ramachandran plot, the\n",
    "hydrogen bonds typical of the $\\alpha$-helix will form and the structure will snap into a very highly populated and finely\n",
    "focused area near '$\\phi$ = 63.8°, $\\psi$ = 41.1° (..) as 39% of all 237 384 amino acids were found here, in an area\n",
    "just 2% of the Ramachandran plot -89° < $\\phi$ < -39°, -66° < $\\psi$ < -16°*\n",
    "___________________________\n",
    "(...)\n",
    "\n",
    "\n",
    "*The conformations of the amino acids in five of the six groups were surprisingly similar with respect to both the average and values (all within -130° < $\\phi$ < -105° and 128° < $\\psi$ < 147°) for the first five groups and the standard deviations (15...26°).*\n",
    "___________________________\n",
    "\n",
    "    - Source: HOVMOLLER S1, ZHOU T, OHLSON T. Conformations of amino acids in proteins. Acta Crystallogr D Biol Crystallogr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 596,
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha_helix_phi = (-89, -39)\n",
    "alpha_helix_psi = (-66, -16)\n",
    "beta_sheet_phi = (-130, -105)\n",
    "beta_sheet_psi = (128, 147)\n",
    "N_AMINOACIDS = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 597,
   "metadata": {},
   "outputs": [],
   "source": [
    "secondary_structure = []\n",
    "for i, row in phi_psi.iterrows():\n",
    "    if alpha_helix_phi[0] < row.loc['phi'] < alpha_helix_phi[1] and\\\n",
    "       alpha_helix_psi[0] < row.loc['psi'] < alpha_helix_psi[1]:\n",
    "        secondary_structure.append('H')\n",
    "    elif beta_sheet_phi[0] < row.loc['phi'] < beta_sheet_phi[1] and\\\n",
    "         beta_sheet_psi[0] < row.loc['psi'] < beta_sheet_psi[1]:\n",
    "        secondary_structure.append('B')\n",
    "    else:\n",
    "        secondary_structure.append('-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 598,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Primary Structure:\n",
      "TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATCPGDYAN\n",
      "\n",
      "Secondary Structure:\n",
      "-B--H-HHHHHHHHHHHH----HHHHHHH----B-H-----H---\n"
     ]
    }
   ],
   "source": [
    "print('Primary Structure:')\n",
    "print('TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATCPGDYAN\\n')\n",
    "print('Secondary Structure:')\n",
    "print(''.join(secondary_structure))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
