{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment II\n",
    "* Leoni Mota Loris"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* At first, Load the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import namedtuple\n",
    "\n",
    "Atom = namedtuple('Atom', ['name', 'element', 'aminoacid', 'x', 'y', 'z', 'h', 'residue_seq_number'])\n",
    "\n",
    "Atom.parse_name = lambda string: string[12:16].strip()\n",
    "Atom.parse_element = lambda string: string[12] if len(string.strip()) < 78 \\\n",
    "                     else string[76:78].strip()\n",
    "Atom.parse_aminoacid = lambda string: string[17:20].strip()\n",
    "Atom.parse_x = lambda string: float(string[30:38].strip())\n",
    "Atom.parse_y = lambda string: float(string[38:46].strip())\n",
    "Atom.parse_z = lambda string: float(string[46:54].strip())\n",
    "Atom.parse_residue_seq_number = lambda string: int(string[23:30].strip())\n",
    "\n",
    "Atom.parse = lambda string: Atom(\n",
    "    Atom.parse_name(string), Atom.parse_element(string), Atom.parse_aminoacid(string),\n",
    "    Atom.parse_x(string), Atom.parse_y(string), Atom.parse_z(string), 1,Atom.parse_residue_seq_number(string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_molecule(path):\n",
    "    atoms = []\n",
    "    with open(path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('ATOM'):\n",
    "                try:\n",
    "                    atoms.append(Atom.parse(line))\n",
    "                except Exception as e:\n",
    "                    print(e)\n",
    "    return pd.DataFrame(atoms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>17.047</td>\n",
       "      <td>14.099</td>\n",
       "      <td>3.625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>16.967</td>\n",
       "      <td>12.784</td>\n",
       "      <td>4.338</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.685</td>\n",
       "      <td>12.755</td>\n",
       "      <td>5.133</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.268</td>\n",
       "      <td>13.825</td>\n",
       "      <td>5.594</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.170</td>\n",
       "      <td>12.703</td>\n",
       "      <td>5.337</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td>O</td>\n",
       "      <td>THR</td>\n",
       "      <td>19.334</td>\n",
       "      <td>12.829</td>\n",
       "      <td>4.463</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>18.150</td>\n",
       "      <td>11.546</td>\n",
       "      <td>6.304</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.115</td>\n",
       "      <td>11.555</td>\n",
       "      <td>5.265</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>13.856</td>\n",
       "      <td>11.469</td>\n",
       "      <td>6.066</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>14.164</td>\n",
       "      <td>10.785</td>\n",
       "      <td>7.379</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name element aminoacid       x       y      z  h  residue_seq_number\n",
       "0    N       N       THR  17.047  14.099  3.625  1                   1\n",
       "1   CA       C       THR  16.967  12.784  4.338  1                   1\n",
       "2    C       C       THR  15.685  12.755  5.133  1                   1\n",
       "3    O       O       THR  15.268  13.825  5.594  1                   1\n",
       "4   CB       C       THR  18.170  12.703  5.337  1                   1\n",
       "5  OG1       O       THR  19.334  12.829  4.463  1                   1\n",
       "6  CG2       C       THR  18.150  11.546  6.304  1                   1\n",
       "7    N       N       THR  15.115  11.555  5.265  1                   2\n",
       "8   CA       C       THR  13.856  11.469  6.066  1                   2\n",
       "9    C       C       THR  14.164  10.785  7.379  1                   2"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crambin = read_molecule('assets/1crn.pdb')\n",
    "crambin[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* I) First we need to map the sequence of characters as in:\n",
    "\n",
    "**TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN**\n",
    "to a sequence of amino acids \n",
    "\n",
    "* II) And then, infer the position of each atom of the amino acid from a standard reference ($\\alpha$ carbon) \n",
    "\n",
    "\n",
    "* III) Infer the position of multiple amino acids based on the peptide bond."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### I) map the sequence of characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_mapping = {'T': 'THR', 'C': 'CYS', 'P': 'PRO', 'S': 'SER', 'I': 'ILE',\n",
    "                     'V': 'VAL', 'A': 'ALA', 'R': 'ARG', 'N': 'ASN', 'F': 'PHE',\n",
    "                     'L': 'LEU', 'G': 'GLY', 'E': 'GLU', 'D': 'ASP', 'Y': 'TYR'}\n",
    "    \n",
    "def map_to_aminoacids(sequence):\n",
    "    return list(map(lambda char: aminoacid_mapping[char], sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacid_sequence = 'TTCCPSIVARSNFNVCRLPGTPEAICATYTGCIIIPGATTCPGDYAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids_names = map_to_aminoacids(aminoacid_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### II) infer the position of each atom of the amino acid. For this part, it is not well defined on how to infer the positions.\n",
    "\n",
    "\n",
    "* We could map this sequence to known sequences of the PDB and hope it is a known protein. This is called \n",
    "**Homology Modelling** (**SWISS-MODEL** is an example).\n",
    "\n",
    "\n",
    "* *Or*, we could predict the protein structure using an any number of softwares such as **I-TASSER (Iterative Threading ASSEmbly Refinement)**\n",
    "\n",
    "\n",
    "We're are going to use both and compare the results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I-TASSER:\n",
    "![assets/TASSER.png](assets/TASSER.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* But, if we're not using something already pre-defined, we should get the pdb file for each of those amino acids and infer the position of each amino acid by the peptide bond.\n",
    "\n",
    "*Obs.:* PDB's for each amino acid were downloaded here:\n",
    "https://www.nyu.edu/pages/mathmol/library/life/life1.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_aminoacids():\n",
    "    return list(map(lambda aminoacid_name: read_molecule('assets/' + aminoacid_name + '.pdb'), aminoacids_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids = load_aminoacids()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>0.080</td>\n",
       "      <td>-0.083</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.539</td>\n",
       "      <td>-0.033</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.032</td>\n",
       "      <td>1.394</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.268</td>\n",
       "      <td>2.339</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.116</td>\n",
       "      <td>-0.818</td>\n",
       "      <td>-1.226</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.743</td>\n",
       "      <td>-2.188</td>\n",
       "      <td>-1.162</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG2</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.654</td>\n",
       "      <td>-0.842</td>\n",
       "      <td>-1.361</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1H</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.523</td>\n",
       "      <td>0.813</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2H</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.461</td>\n",
       "      <td>-1.018</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HA</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.904</td>\n",
       "      <td>-0.508</td>\n",
       "      <td>0.929</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HB</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>1.691</td>\n",
       "      <td>-0.378</td>\n",
       "      <td>-2.154</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HG1</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>2.066</td>\n",
       "      <td>-2.592</td>\n",
       "      <td>-1.973</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1HG2</td>\n",
       "      <td>1</td>\n",
       "      <td>THR</td>\n",
       "      <td>3.986</td>\n",
       "      <td>-1.433</td>\n",
       "      <td>-2.235</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2HG2</td>\n",
       "      <td>2</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.083</td>\n",
       "      <td>0.168</td>\n",
       "      <td>-1.502</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3HG2</td>\n",
       "      <td>3</td>\n",
       "      <td>THR</td>\n",
       "      <td>4.138</td>\n",
       "      <td>-1.279</td>\n",
       "      <td>-0.467</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.452</td>\n",
       "      <td>1.567</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HC</td>\n",
       "      <td></td>\n",
       "      <td>THR</td>\n",
       "      <td>3.911</td>\n",
       "      <td>0.645</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name element aminoacid      x      y      z  h  residue_seq_number\n",
       "0      N               THR  0.080 -0.083  0.000  1                   1\n",
       "1     CA               THR  1.539 -0.033  0.000  1                   1\n",
       "2      C               THR  2.032  1.394  0.000  1                   1\n",
       "3      O               THR  1.268  2.339  0.000  1                   1\n",
       "4     CB               THR  2.116 -0.818 -1.226  1                   1\n",
       "5    OG1               THR  1.743 -2.188 -1.162  1                   1\n",
       "6    CG2               THR  3.654 -0.842 -1.361  1                   1\n",
       "7     1H       1       THR -0.523  0.813  0.000  1                   1\n",
       "8     2H       2       THR -0.461 -1.018  0.000  1                   1\n",
       "9     HA               THR  1.904 -0.508  0.929  1                   1\n",
       "10    HB               THR  1.691 -0.378 -2.154  1                   1\n",
       "11   HG1               THR  2.066 -2.592 -1.973  1                   1\n",
       "12  1HG2       1       THR  3.986 -1.433 -2.235  1                   1\n",
       "13  2HG2       2       THR  4.083  0.168 -1.502  1                   1\n",
       "14  3HG2       3       THR  4.138 -1.279 -0.467  1                   1\n",
       "15    OC               THR  3.452  1.567  0.000  1                   1\n",
       "16    HC               THR  3.911  0.645  0.000  1                   1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(aminoacids[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "______________________________________________________________________________________________________________\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### III) Infer the position of multiple amino acids based on the peptide bond.\n",
    "\n",
    "\n",
    "* After the nucleotilic attack the resonance delocalization of electrons in the peptide bond forces the creation of a plane where the atoms will be located (amide plan)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* First, We define how translations and rotations are made.\n",
    "\n",
    "\n",
    "* Homogeneous coordinates were used throughout this work.\n",
    "![assets/rotation.png](assets/rotation.png)\n",
    "![assets/translation.png](assets/translation.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def translation_matrix(delta_x, delta_y, delta_z, ref_x=.0, ref_y=.0, ref_z=.0):\n",
    "    T = np.eye(4)\n",
    "    T[0, -1], T[1, -1], T[2, -1] = delta_x - ref_x, delta_y - ref_y, delta_z - ref_z\n",
    "    return T\n",
    "\n",
    "def rotation_matrix(x_rotation_rad, y_rotation_rad, z_rotation_rad):\n",
    "    Rx, Ry, Rz = np.eye(4), np.eye(4), np.eye(4)\n",
    "\n",
    "    Rx[1, 1], Rx[1, 2], Rx[2, 1], Rx[2, 2] =\\\n",
    "        np.cos(x_rotation_rad), -np.sin(x_rotation_rad), np.sin(x_rotation_rad), np.cos(x_rotation_rad)\n",
    "\n",
    "    Ry[0, 0], Ry[0, 2], Ry[2, 0], Rx[2, 2] =\\\n",
    "        np.cos(y_rotation_rad), np.sin(y_rotation_rad), -np.sin(y_rotation_rad), np.cos(y_rotation_rad)\n",
    "\n",
    "    Rz[0, 0], Rx[0, 1], Rx[1, 0], Rx[1, 1] =\\\n",
    "        np.cos(z_rotation_rad), -np.sin(z_rotation_rad), np.sin(z_rotation_rad), np.cos(z_rotation_rad)\n",
    "\n",
    "    return np.dot(np.dot(Rx, Ry), Rz)\n",
    "\n",
    "def apply_transformation(transformation, df):\n",
    "    df[['x', 'y', 'z', 'h']] = transformation.dot(df[['x', 'y', 'z', 'h']].T).T\n",
    "\n",
    "def apply_rotation(transformation, df, atom_reference='N'):\n",
    "    position = df[df.name == atom_reference]\n",
    "    apply_transformation(translation_matrix(-position.x, -position.y, -position.z), df)\n",
    "    apply_transformation(transformation, df)\n",
    "    apply_transformation(translation_matrix(position.x, position.y, position.z), df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# rotations_z = np.arange(0, 190, 10) * np.pi / 180\n",
    "\n",
    "# for rotation_z in rotations_z:\n",
    "#     a = aminoacids[0].copy()\n",
    "#     apply_rotation(rotation_matrix(rotation_z, 0, 0), a)    \n",
    "#     plt.scatter(a[:10].x, a[:10].z, c=np.arange(10), s=70)\n",
    "#     plt.xlabel('X!!!!')\n",
    "#     print(rotation_z *180 / np.pi)\n",
    "#     plt.xlim(-5, 5)\n",
    "#     plt.ylim(-5, 5)\n",
    "#     plt.grid()\n",
    "#     plt.show()\n",
    "#     plt.pause(0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dehydration(aminoacid1, aminoacid2):\n",
    "    aminoacid1 = aminoacid1[~aminoacid1.name.isin(['OC', 'HC'])]    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some utility functions for calculating the euclidean distance and the angle between two vectors:\n",
    "\n",
    "\n",
    "$$distance = \\sqrt{|(A - B)|}$$\n",
    "\n",
    "\n",
    "$$cos(\\alpha) = \\frac{A \\cdot B}{|A||B|}$$ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def euclidean_distance(p1, p2):\n",
    "    return np.sqrt(np.sum((p1 - p2) ** 2))\n",
    "\n",
    "def l2_norm(p):\n",
    "    return np.sqrt(np.sum(p ** 2))\n",
    "\n",
    "def get_angle_rad(a, b):\n",
    "    angle = np.arccos(((a).dot(b))/(l2_norm(a) * l2_norm(b))) \n",
    "    return angle if not np.isnan(angle) else .0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will consider the omega torsion angle of `Proline` to be close to zero degrees for the `cis` configuration, or most often and 180 degrees for the trans configuration.\n",
    "\n",
    "\n",
    "IMAGEM AQUI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_omega(current_aminoacid):\n",
    "    return 0 if 'PRO' in current_aminoacid.aminoacid else 180\n",
    "\n",
    "def get_desired_OCN_angle(current_aminoacid):\n",
    "    return 123 if 'PRO' in current_aminoacid.aminoacid else 123.5\n",
    "\n",
    "def get_desired_bond_length(current_aminoacid):\n",
    "    return 1.32 if 'PRO' in current_aminoacid.aminoacid else 1.33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_desired_nitrogen_location(aminoacid):\n",
    "    alpha_carbon = aminoacid[aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    carboxyl_carbon = aminoacid[aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    oxygen_double_bonded = aminoacid[aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "\n",
    "    CO = oxygen_double_bonded - carboxyl_carbon\n",
    "    CA = alpha_carbon - carboxyl_carbon\n",
    "\n",
    "    plane_normal = np.cross(CO, CA)\n",
    "    plane_vector_basis = np.array([\n",
    "        np.cross(CO, plane_normal) / l2_norm(np.cross(CO, plane_normal)), \n",
    "        CO / l2_norm(CO), \n",
    "        plane_normal / l2_norm(plane_normal)]).T\n",
    "\n",
    "    CO_new_norm = CO * get_desired_bond_length(aminoacid) / l2_norm(CO)\n",
    "    CO_new_norm_plane = np.linalg.solve(plane_vector_basis, CO_new_norm)[:2]\n",
    "    \n",
    "    rotation_angle = get_desired_OCN_angle(aminoacid) * np.pi / 180\n",
    "    rotation_angle_matrix = np.array(\n",
    "        [[np.cos(rotation_angle), np.sin(rotation_angle)], \n",
    "         [np.sin(rotation_angle), np.cos(rotation_angle)]])\n",
    "\n",
    "    CN_plane = CO_new_norm_plane @ rotation_angle_matrix\n",
    "    CN = np.linalg.solve(np.linalg.inv(plane_vector_basis), np.array([CN_plane[0], CN_plane[1], 0]))\n",
    "    return CN + carboxyl_carbon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C'):\n",
    "    if terminal == 'C':\n",
    "        alpha_carbon = previous_aminoacid[previous_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        oxygen_double_bonded = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C1 = oxygen_double_bonded - carboxyl_carbon\n",
    "        C2 = alpha_carbon - carboxyl_carbon\n",
    "    if terminal == 'N':\n",
    "        last_carbon = current_aminoacid[current_aminoacid.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        nitrogen_amide = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        carboxyl_carbon = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        C2 = last_carbon - nitrogen_amide\n",
    "        C1 = carboxyl_carbon - nitrogen_amide\n",
    "    return np.cross(C1, C2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    desired_nitrogen_location = get_desired_nitrogen_location(previous_aminoacid)\n",
    "    current_nitrogen_location = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        \n",
    "    required_tranlation = desired_nitrogen_location - current_nitrogen_location\n",
    "    print('required_tranlation', required_tranlation)\n",
    "    required_tranlation_matrix = translation_matrix(required_tranlation[0], required_tranlation[1], required_tranlation[2])\n",
    "    apply_transformation(required_tranlation_matrix, current_aminoacid)\n",
    "    \n",
    "def rotate_for_peptide_bond(previous_aminoacid, current_aminoacid):\n",
    "    omega = get_desired_omega(current_aminoacid)\n",
    "    previous_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='C')\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_x = get_angle_rad(current_normal[[1, 2]], previous_normal[[1, 2]])\n",
    "    apply_rotation(rotation_matrix(rotation_angle_x, 0, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_y = get_angle_rad(current_normal[[0, 2]], previous_normal[[0, 2]])\n",
    "    apply_rotation(rotation_matrix(0, rotation_angle_y, 0), current_aminoacid)\n",
    "    \n",
    "    current_normal = get_normal_from_amide_plan(previous_aminoacid, current_aminoacid, terminal='N')\n",
    "    rotation_angle_z = get_angle_rad(current_normal[[0, 1]], previous_normal[[0, 1]])\n",
    "    apply_rotation(rotation_matrix(0, 0, rotation_angle_z), current_aminoacid)\n",
    "#     print(np.degrees(rotation_angle_x), np.degrees(rotation_angle_y), np.degrees(rotation_angle_z))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_peptide_bond_length(previous_aminoacid, current_aminoacid):\n",
    "    a = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    b = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('euclidean_distance:', euclidean_distance(a, b))\n",
    "    \n",
    "def check_OCN_angle(previous_aminoacid, current_aminoacid):\n",
    "    c = previous_aminoacid[previous_aminoacid.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    n = current_aminoacid[current_aminoacid.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    o = previous_aminoacid[previous_aminoacid.name == 'O'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "    print('angle ', 180*get_angle_rad(o-c, n-c)/np.pi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_tranlation [3.27597809 1.60341997 0.        ]\n",
      "required_tranlation [6.52995619 3.26683993 0.        ]\n",
      "required_tranlation [9.73111425 5.01321629 0.        ]\n",
      "required_tranlation [12.92227231  5.93259264 -0.124     ]\n",
      "required_tranlation [16.32312732  7.94976373 -1.35328062]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:8: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_tranlation [17.89360012  8.22475197 -2.99257939]\n",
      "required_tranlation [17.31595172 10.09411035 -1.90952138]\n",
      "required_tranlation [15.09805706  9.64464837 -1.24089123]\n",
      "required_tranlation [17.57025095  9.15413563 -3.05199295]\n",
      "required_tranlation [18.15341985 12.94823496 -3.55867366]\n",
      "required_tranlation [19.23194688 15.85334758 -4.23767733]\n",
      "required_tranlation [22.39419957 17.61776004 -4.88223142]\n",
      "required_tranlation [22.30698043 15.71905078 -3.83242688]\n",
      "required_tranlation [22.18352125 18.0162012  -4.98266061]\n",
      "required_tranlation [23.86806185 20.33106929 -6.39390939]\n",
      "required_tranlation [23.19182125 19.99526706 -8.96297437]\n",
      "required_tranlation [23.70761443 23.83605428 -9.48418243]\n",
      "required_tranlation [26.69535213 24.84228831 -9.10662031]\n",
      "required_tranlation [ 24.72382294  24.48818948 -11.46815849]\n",
      "required_tranlation [ 27.17106272  27.10822491 -11.56509476]\n",
      "required_tranlation [ 30.32036125  28.13999634 -11.75066971]\n",
      "required_tranlation [ 33.65725652  30.24229904 -12.88497225]\n",
      "required_tranlation [ 36.48190893  28.42165848 -14.31565615]\n",
      "required_tranlation [ 32.95559235  27.10010502 -13.93427299]\n",
      "required_tranlation [ 36.03466211  28.98932666 -13.52250639]\n",
      "required_tranlation [ 39.2085737   30.77624474 -13.40629843]\n",
      "required_tranlation [ 42.50969499  32.29375858 -13.36875355]\n",
      "required_tranlation [ 45.37555249  34.58291844 -12.70482525]\n",
      "required_tranlation [ 48.89291555  36.11341331 -12.57125501]\n",
      "required_tranlation [ 45.24891211  36.41633036 -13.30066308]\n",
      "required_tranlation [ 45.15003076  38.57751815 -13.09858828]\n",
      "required_tranlation [ 48.29281757  40.36829816 -13.74992548]\n",
      "required_tranlation [ 47.12710369  38.62518095 -12.48477971]\n",
      "required_tranlation [ 48.50800232  41.71124484 -12.24444873]\n",
      "required_tranlation [ 51.61613256  42.7290519  -12.81822354]\n",
      "required_tranlation [ 54.65013705  45.67569656 -13.89201859]\n",
      "required_tranlation [ 56.27467486  44.27147345 -14.7429431 ]\n",
      "required_tranlation [ 52.7737227   42.83580958 -14.41374376]\n",
      "required_tranlation [ 55.68456492  44.95175236 -13.37192085]\n",
      "required_tranlation [ 59.05905521  46.51347966 -13.03876036]\n",
      "required_tranlation [ 55.49347869  46.18873577 -14.03548062]\n",
      "required_tranlation [ 52.40556861  45.4763498  -15.682064  ]\n",
      "required_tranlation [ 48.69604721  44.70073286 -16.90299579]\n",
      "required_tranlation [ 53.57460012  46.89513072 -15.97760182]\n",
      "required_tranlation [ 54.42584628  49.51955408 -15.72074061]\n",
      "required_tranlation [ 57.3155818   51.36938957 -16.80969557]\n"
     ]
    }
   ],
   "source": [
    "aminoacids_df = load_aminoacids()\n",
    "for i in range(1, len(aminoacids_df)):\n",
    "    current_aminoacid = aminoacids_df[i]\n",
    "    previous_aminoacid = aminoacids_df[i - 1]\n",
    "    \n",
    "    dehydration(previous_aminoacid, current_aminoacid)\n",
    "    translate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "    rotate_for_peptide_bond(previous_aminoacid, current_aminoacid)\n",
    "    \n",
    "#     check_peptide_bond_length(previous_aminoacid, current_aminoacid)\n",
    "#     check_OCN_angle(previous_aminoacid, current_aminoacid)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We have all the atoms aligned accordingly. So, let's create string rows for the pdb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_atom_to_file(atom_row, aminoacid_seq_number, fd):\n",
    "    global atom_seq_number\n",
    "    atom_seq_number += 1\n",
    "    if 'H' not in atom_row[0]:\n",
    "        atom_row.element = ''.join(filter(str.isalpha, atom_row[0]))[0]\n",
    "#         template_row = 'ATOM   ' + ('%d'%atom_seq_number).rjust(4) +\\\n",
    "#                        '  %s'% atom_row.element + '   %s'% atom_row.aminoacid +\\\n",
    "#         ' A'+('%d'%(idx+1)).rjust(3) +\\\n",
    "#         ('%.3f'%atom_row.x).rjust(6)+\\\n",
    "#         ('%.3f'%atom_row.y).rjust(8)+\\\n",
    "#         ('%.3f'%atom_row.z).rjust(8)+\\\n",
    "#         '  1.00  0.0 '+\\\n",
    "#         '           %s  \\n'%atom_row.element\n",
    "        \n",
    "        template_row =\\\n",
    "        'ATOM    '+\\\n",
    "        ('%d'%atom_seq_number).rjust(3)+\\\n",
    "        '  '+\\\n",
    "        ('%s'%atom_row[0]).rjust(3)+\\\n",
    "        ' '+\\\n",
    "        ('%s'%atom_row.aminoacid).rjust(3)+\\\n",
    "        ' A'+\\\n",
    "        ('%d'%aminoacid_seq_number).rjust(4)+\\\n",
    "        '    '+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.y).rjust(8)+\\\n",
    "        ('%.3f'%atom_row.x).rjust(8)+\\\n",
    "        '  1.00  5.99           '+\\\n",
    "        ('%s\\n'%atom_row.element)\n",
    "        print(template_row)\n",
    "        f.write(template_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATOM      1    N THR A   0       0.080  -0.083   0.080  1.00  5.99           N\n",
      "\n",
      "ATOM      2   CA THR A   0       1.539  -0.033   1.539  1.00  5.99           C\n",
      "\n",
      "ATOM      3    C THR A   0       2.032   1.394   2.032  1.00  5.99           C\n",
      "\n",
      "ATOM      4    O THR A   0       1.268   2.339   1.268  1.00  5.99           O\n",
      "\n",
      "ATOM      5   CB THR A   0       2.116  -0.818   2.116  1.00  5.99           C\n",
      "\n",
      "ATOM      6  OG1 THR A   0       1.743  -2.188   1.743  1.00  5.99           O\n",
      "\n",
      "ATOM      7  CG2 THR A   0       3.654  -0.842   3.654  1.00  5.99           C\n",
      "\n",
      "ATOM     16   OC THR A   0       3.452   1.567   3.452  1.00  5.99           O\n",
      "\n",
      "ATOM     18    N THR A   1       3.356   1.520   3.356  1.00  5.99           N\n",
      "\n",
      "ATOM     19   CA THR A   1       4.815   1.570   4.815  1.00  5.99           C\n",
      "\n",
      "ATOM     20    C THR A   1       5.308   2.997   5.308  1.00  5.99           C\n",
      "\n",
      "ATOM     21    O THR A   1       4.544   3.942   4.544  1.00  5.99           O\n",
      "\n",
      "ATOM     22   CB THR A   1       5.392   0.785   5.392  1.00  5.99           C\n",
      "\n",
      "ATOM     23  OG1 THR A   1       5.019  -0.585   5.019  1.00  5.99           O\n",
      "\n",
      "ATOM     24  CG2 THR A   1       6.930   0.761   6.930  1.00  5.99           C\n",
      "\n",
      "ATOM     33   OC THR A   1       6.728   3.170   6.728  1.00  5.99           O\n",
      "\n",
      "ATOM     35    N CYS A   2       6.632   3.124   6.632  1.00  5.99           N\n",
      "\n",
      "ATOM     36   CA CYS A   2       8.088   3.238   8.088  1.00  5.99           C\n",
      "\n",
      "ATOM     37    C CYS A   2       8.516   4.686   8.516  1.00  5.99           C\n",
      "\n",
      "ATOM     38    O CYS A   2       7.695   5.615   7.695  1.00  5.99           O\n",
      "\n",
      "ATOM     39   CB CYS A   2       8.641   2.460   8.641  1.00  5.99           C\n",
      "\n",
      "ATOM     40   SG CYS A   2      10.446   2.411  10.446  1.00  5.99           S\n",
      "\n",
      "ATOM     48   OC CYS A   2       9.926   4.922   9.926  1.00  5.99           O\n",
      "\n",
      "ATOM     49    N CYS A   3       9.833   4.870   9.833  1.00  5.99           N\n",
      "\n",
      "ATOM     50   CA CYS A   3      11.289   4.984  11.289  1.00  5.99           C\n",
      "\n",
      "ATOM     51    C CYS A   3      11.717   6.432  11.717  1.00  5.99           C\n",
      "\n",
      "ATOM     52    O CYS A   3      10.896   7.361  10.896  1.00  5.99           O\n",
      "\n",
      "ATOM     53   CB CYS A   3      11.842   4.206  11.842  1.00  5.99           C\n",
      "\n",
      "ATOM     54   SG CYS A   3      13.647   4.157  13.647  1.00  5.99           S\n",
      "\n",
      "ATOM     62   OC CYS A   3      13.127   6.668  13.127  1.00  5.99           O\n",
      "\n",
      "ATOM     63    N PRO A   4      13.034   6.617  13.034  1.00  5.99           N\n",
      "\n",
      "ATOM     64   CA PRO A   4      14.512   6.529  14.512  1.00  5.99           C\n",
      "\n",
      "ATOM     65    C PRO A   4      15.083   7.921  15.083  1.00  5.99           C\n",
      "\n",
      "ATOM     66    O PRO A   4      14.311   8.985  14.311  1.00  5.99           O\n",
      "\n",
      "ATOM     67   CB PRO A   4      14.621   5.356  14.621  1.00  5.99           C\n",
      "\n",
      "ATOM     68   CG PRO A   4      13.276   5.543  13.276  1.00  5.99           C\n",
      "\n",
      "ATOM     69   CD PRO A   4      12.277   5.808  12.277  1.00  5.99           C\n",
      "\n",
      "ATOM     78   OC PRO A   4      16.453   8.013  16.453  1.00  5.99           O\n",
      "\n",
      "ATOM     80    N SER A   5      16.366   7.894  16.366  1.00  5.99           N\n",
      "\n",
      "ATOM     81   CA SER A   5      17.474   8.566  17.474  1.00  5.99           C\n",
      "\n",
      "ATOM     82    C SER A   5      16.837   7.602  16.837  1.00  5.99           C\n",
      "\n",
      "ATOM     83    O SER A   5      15.542   6.413  15.542  1.00  5.99           O\n",
      "\n",
      "ATOM     84   CB SER A   5      17.902   8.684  17.902  1.00  5.99           C\n",
      "\n",
      "ATOM     85   OG SER A   5      16.618   7.138  16.618  1.00  5.99           O\n",
      "\n",
      "ATOM     92   OC SER A   5      17.826   8.150  17.826  1.00  5.99           O\n",
      "\n",
      "ATOM     94    N ILE A   6      18.010   8.009  18.010  1.00  5.99           N\n",
      "\n",
      "ATOM     95   CA ILE A   6      17.620   8.120  17.620  1.00  5.99           C\n",
      "\n",
      "ATOM     96    C ILE A   6      18.320   9.740  18.320  1.00  5.99           C\n",
      "\n",
      "ATOM     97    O ILE A   6      19.112  10.790  19.112  1.00  5.99           O\n",
      "\n",
      "ATOM     98   CB ILE A   6      15.668   7.962  15.668  1.00  5.99           C\n",
      "\n",
      "ATOM     99  CG1 ILE A   6      14.839   6.363  14.839  1.00  5.99           C\n",
      "\n",
      "ATOM    100  CG2 ILE A   6      15.214   7.960  15.214  1.00  5.99           C\n",
      "\n",
      "ATOM    101  CD1 ILE A   6      12.902   6.380  12.902  1.00  5.99           C\n",
      "\n",
      "ATOM    115   OC ILE A   6      18.015   9.987  18.015  1.00  5.99           O\n",
      "\n",
      "ATOM    116    N VAL A   7      17.411   9.913  17.411  1.00  5.99           N\n",
      "\n",
      "ATOM    117   CA VAL A   7      17.063   9.745  17.063  1.00  5.99           C\n",
      "\n",
      "ATOM    118    C VAL A   7      15.659  10.015  15.659  1.00  5.99           C\n",
      "\n",
      "ATOM    119    O VAL A   7      14.944  10.332  14.944  1.00  5.99           O\n",
      "\n",
      "ATOM    120   CB VAL A   7      16.507   9.575  16.507  1.00  5.99           C\n",
      "\n",
      "ATOM    121  CG1 VAL A   7      17.982   9.259  17.982  1.00  5.99           C\n",
      "\n",
      "ATOM    122  CG2 VAL A   7      15.854   9.429  15.854  1.00  5.99           C\n",
      "\n",
      "ATOM    133   OC VAL A   7      15.208   9.880  15.208  1.00  5.99           O\n",
      "\n",
      "ATOM    135    N ALA A   8      15.137   9.617  15.137  1.00  5.99           N\n",
      "\n",
      "ATOM    136   CA ALA A   8      16.580   9.735  16.580  1.00  5.99           C\n",
      "\n",
      "ATOM    137    C ALA A   8      17.002  11.184  17.002  1.00  5.99           C\n",
      "\n",
      "ATOM    138    O ALA A   8      16.189  12.108  16.189  1.00  5.99           O\n",
      "\n",
      "ATOM    139   CB ALA A   8      16.944   8.964  16.944  1.00  5.99           C\n",
      "\n",
      "ATOM    146   OC ALA A   8      18.399  11.423  18.399  1.00  5.99           O\n",
      "\n",
      "ATOM    148    N ARG A   9      18.305  11.373  18.305  1.00  5.99           N\n",
      "\n",
      "ATOM    149   CA ARG A   9      18.541  11.753  18.541  1.00  5.99           C\n",
      "\n",
      "ATOM    150    C ARG A   9      17.240  12.166  17.240  1.00  5.99           C\n",
      "\n",
      "ATOM    151    O ARG A   9      16.155  12.161  16.155  1.00  5.99           O\n",
      "\n",
      "ATOM    152   CB ARG A   9      20.059  11.475  20.059  1.00  5.99           C\n",
      "\n",
      "ATOM    153   CG ARG A   9      20.640  11.804  20.640  1.00  5.99           C\n",
      "\n",
      "ATOM    154   CD ARG A   9      22.121  11.505  22.121  1.00  5.99           C\n",
      "\n",
      "ATOM    155   NE ARG A   9      22.635  11.837  22.635  1.00  5.99           N\n",
      "\n",
      "ATOM    156   CZ ARG A   9      23.872  11.698  23.872  1.00  5.99           C\n",
      "\n",
      "ATOM    173   OC ARG A   9      17.333  12.566  17.333  1.00  5.99           O\n",
      "\n",
      "ATOM    175    N SER A  10      18.196  12.892  18.196  1.00  5.99           N\n",
      "\n",
      "ATOM    176   CA SER A  10      19.101  13.611  19.101  1.00  5.99           C\n",
      "\n",
      "ATOM    177    C SER A  10      18.420  14.841  18.420  1.00  5.99           C\n",
      "\n",
      "ATOM    178    O SER A  10      17.251  15.149  17.251  1.00  5.99           O\n",
      "\n",
      "ATOM    179   CB SER A  10      19.625  13.307  19.625  1.00  5.99           C\n",
      "\n",
      "ATOM    180   OG SER A  10      18.591  13.609  18.591  1.00  5.99           O\n",
      "\n",
      "ATOM    187   OC SER A  10      19.213  15.628  19.213  1.00  5.99           O\n",
      "\n",
      "ATOM    189    N ASN A  11      19.314  15.735  19.314  1.00  5.99           N\n",
      "\n",
      "ATOM    190   CA ASN A  11      20.665  15.845  20.665  1.00  5.99           C\n",
      "\n",
      "ATOM    191    C ASN A  11      21.259  17.296  21.259  1.00  5.99           C\n",
      "\n",
      "ATOM    192    O ASN A  11      20.639  18.222  20.639  1.00  5.99           O\n",
      "\n",
      "ATOM    193   CB ASN A  11      20.601  15.525  20.601  1.00  5.99           C\n",
      "\n",
      "ATOM    194   CG ASN A  11      20.012  14.088  20.012  1.00  5.99           C\n",
      "\n",
      "ATOM    195  OD1 ASN A  11      18.853  13.962  18.853  1.00  5.99           O\n",
      "\n",
      "ATOM    196  ND2 ASN A  11      20.770  12.957  20.770  1.00  5.99           N\n",
      "\n",
      "ATOM    205   OC ASN A  11      22.586  17.529  22.586  1.00  5.99           O\n",
      "\n",
      "ATOM    206    N PHE A  12      22.504  17.418  22.504  1.00  5.99           N\n",
      "\n",
      "ATOM    207   CA PHE A  12      23.102  17.584  23.102  1.00  5.99           C\n",
      "\n",
      "ATOM    208    C PHE A  12      21.681  15.943  21.681  1.00  5.99           C\n",
      "\n",
      "ATOM    209    O PHE A  12      20.193  14.638  20.193  1.00  5.99           O\n",
      "\n",
      "ATOM    210   CB PHE A  12      24.545  17.923  24.545  1.00  5.99           C\n",
      "\n",
      "ATOM    211   CG PHE A  12      26.110  19.547  26.110  1.00  5.99           C\n",
      "\n",
      "ATOM    212  CD1 PHE A  12      26.141  19.415  26.141  1.00  5.99           C\n",
      "\n",
      "ATOM    213  CD2 PHE A  12      27.472  21.176  27.472  1.00  5.99           C\n",
      "\n",
      "ATOM    214  CE1 PHE A  12      27.504  20.895  27.504  1.00  5.99           C\n",
      "\n",
      "ATOM    215  CE2 PHE A  12      28.843  22.654  28.843  1.00  5.99           C\n",
      "\n",
      "ATOM    216   CZ PHE A  12      28.857  22.515  28.857  1.00  5.99           C\n",
      "\n",
      "ATOM    227   OC PHE A  12      22.120  15.955  22.120  1.00  5.99           O\n",
      "\n",
      "ATOM    229    N ASN A  13      22.389  15.601  22.389  1.00  5.99           N\n",
      "\n",
      "ATOM    230   CA ASN A  13      22.785  16.070  22.785  1.00  5.99           C\n",
      "\n",
      "ATOM    231    C ASN A  13      21.838  16.992  21.838  1.00  5.99           C\n",
      "\n",
      "ATOM    232    O ASN A  13      20.875  17.277  20.875  1.00  5.99           O\n",
      "\n",
      "ATOM    233   CB ASN A  13      22.539  15.832  22.539  1.00  5.99           C\n",
      "\n",
      "ATOM    234   CG ASN A  13      23.420  14.912  23.420  1.00  5.99           C\n",
      "\n",
      "ATOM    235  OD1 ASN A  13      22.811  14.460  22.811  1.00  5.99           O\n",
      "\n",
      "ATOM    236  ND2 ASN A  13      24.879  14.597  24.879  1.00  5.99           N\n",
      "\n",
      "ATOM    245   OC ASN A  13      22.128  17.516  22.128  1.00  5.99           O\n",
      "\n",
      "ATOM    246    N VAL A  14      22.279  17.835  22.279  1.00  5.99           N\n",
      "\n",
      "ATOM    247   CA VAL A  14      23.018  18.298  23.018  1.00  5.99           C\n",
      "\n",
      "ATOM    248    C VAL A  14      23.010  19.867  23.010  1.00  5.99           C\n",
      "\n",
      "ATOM    249    O VAL A  14      22.437  20.610  22.437  1.00  5.99           O\n",
      "\n",
      "ATOM    250   CB VAL A  14      22.342  17.683  22.342  1.00  5.99           C\n",
      "\n",
      "ATOM    251  CG1 VAL A  14      22.492  16.078  22.492  1.00  5.99           C\n",
      "\n",
      "ATOM    252  CG2 VAL A  14      22.881  18.230  22.881  1.00  5.99           C\n",
      "\n",
      "ATOM    263   OC VAL A  14      23.706  20.441  23.706  1.00  5.99           O\n",
      "\n",
      "ATOM    265    N CYS A  15      23.970  20.188  23.970  1.00  5.99           N\n",
      "\n",
      "ATOM    266   CA CYS A  15      24.176  20.518  24.176  1.00  5.99           C\n",
      "\n",
      "ATOM    267    C CYS A  15      23.452  21.656  23.452  1.00  5.99           C\n",
      "\n",
      "ATOM    268    O CYS A  15      22.785  22.201  22.785  1.00  5.99           O\n",
      "\n",
      "ATOM    269   CB CYS A  15      23.190  20.119  23.190  1.00  5.99           C\n",
      "\n",
      "ATOM    270   SG CYS A  15      23.604  20.385  23.604  1.00  5.99           S\n",
      "\n",
      "ATOM    278   OC CYS A  15      23.583  22.068  23.583  1.00  5.99           O\n",
      "\n",
      "ATOM    279    N ARG A  16      23.927  22.214  23.927  1.00  5.99           N\n",
      "\n",
      "ATOM    280   CA ARG A  16      24.264  22.572  24.264  1.00  5.99           C\n",
      "\n",
      "ATOM    281    C ARG A  16      22.908  23.352  22.908  1.00  5.99           C\n",
      "\n",
      "ATOM    282    O ARG A  16      21.709  23.621  21.709  1.00  5.99           O\n",
      "\n",
      "ATOM    283   CB ARG A  16      24.843  21.218  24.843  1.00  5.99           C\n",
      "\n",
      "ATOM    284   CG ARG A  16      25.543  21.427  25.543  1.00  5.99           C\n",
      "\n",
      "ATOM    285   CD ARG A  16      26.048  20.041  26.048  1.00  5.99           C\n",
      "\n",
      "ATOM    286   NE ARG A  16      26.729  20.305  26.729  1.00  5.99           N\n",
      "\n",
      "ATOM    287   CZ ARG A  16      27.275  19.334  27.275  1.00  5.99           C\n",
      "\n",
      "ATOM    304   OC ARG A  16      23.115  23.783  23.115  1.00  5.99           O\n",
      "\n",
      "ATOM    306    N LEU A  17      23.710  23.855  23.710  1.00  5.99           N\n",
      "\n",
      "ATOM    307   CA LEU A  17      25.138  23.875  25.138  1.00  5.99           C\n",
      "\n",
      "ATOM    308    C LEU A  17      25.787  25.297  25.787  1.00  5.99           C\n",
      "\n",
      "ATOM    309    O LEU A  17      25.185  26.285  25.185  1.00  5.99           O\n",
      "\n",
      "ATOM    310   CB LEU A  17      25.267  23.315  25.267  1.00  5.99           C\n",
      "\n",
      "ATOM    311   CG LEU A  17      24.935  21.791  24.935  1.00  5.99           C\n",
      "\n",
      "ATOM    312  CD1 LEU A  17      24.885  21.481  24.885  1.00  5.99           C\n",
      "\n",
      "ATOM    313  CD2 LEU A  17      25.960  20.833  25.960  1.00  5.99           C\n",
      "\n",
      "ATOM    327   OC LEU A  17      26.754  25.520  26.754  1.00  5.99           O\n",
      "\n",
      "ATOM    328    N PRO A  18      26.807  25.526  26.807  1.00  5.99           N\n",
      "\n",
      "ATOM    329   CA PRO A  18      26.533  25.281  26.533  1.00  5.99           C\n",
      "\n",
      "ATOM    330    C PRO A  18      24.590  24.304  24.590  1.00  5.99           C\n",
      "\n",
      "ATOM    331    O PRO A  18      23.420  23.801  23.420  1.00  5.99           O\n",
      "\n",
      "ATOM    332   CB PRO A  18      27.909  26.189  27.909  1.00  5.99           C\n",
      "\n",
      "ATOM    333   CG PRO A  18      27.939  26.498  27.939  1.00  5.99           C\n",
      "\n",
      "ATOM    334   CD PRO A  18      27.933  26.399  27.933  1.00  5.99           C\n",
      "\n",
      "ATOM    343   OC PRO A  18      24.112  23.969  24.112  1.00  5.99           O\n",
      "\n",
      "ATOM    345    N GLY A  19      24.886  24.286  24.886  1.00  5.99           N\n",
      "\n",
      "ATOM    346   CA GLY A  19      26.198  24.892  26.198  1.00  5.99           C\n",
      "\n",
      "ATOM    347    C GLY A  19      26.084  26.393  26.084  1.00  5.99           C\n",
      "\n",
      "ATOM    348    O GLY A  19      25.002  26.974  25.002  1.00  5.99           O\n",
      "\n",
      "ATOM    353   OC GLY A  19      27.311  27.097  27.311  1.00  5.99           O\n",
      "\n",
      "ATOM    355    N THR A  20      27.251  27.025  27.251  1.00  5.99           N\n",
      "\n",
      "ATOM    356   CA THR A  20      28.695  27.169  28.695  1.00  5.99           C\n",
      "\n",
      "ATOM    357    C THR A  20      29.123  28.627  29.123  1.00  5.99           C\n",
      "\n",
      "ATOM    358    O THR A  20      28.326  29.522  28.326  1.00  5.99           O\n",
      "\n",
      "ATOM    359   CB THR A  20      29.146  26.652  29.146  1.00  5.99           C\n",
      "\n",
      "ATOM    360  OG1 THR A  20      28.843  25.247  28.843  1.00  5.99           O\n",
      "\n",
      "ATOM    361  CG2 THR A  20      30.655  26.752  30.655  1.00  5.99           C\n",
      "\n",
      "ATOM    370   OC THR A  20      30.524  28.891  30.524  1.00  5.99           O\n",
      "\n",
      "ATOM    372    N PRO A  21      30.432  28.824  30.432  1.00  5.99           N\n",
      "\n",
      "ATOM    373   CA PRO A  21      31.945  28.792  31.945  1.00  5.99           C\n",
      "\n",
      "ATOM    374    C PRO A  21      32.493  29.969  32.493  1.00  5.99           C\n",
      "\n",
      "ATOM    375    O PRO A  21      31.676  30.836  31.676  1.00  5.99           O\n",
      "\n",
      "ATOM    376   CB PRO A  21      32.249  27.491  32.249  1.00  5.99           C\n",
      "\n",
      "ATOM    377   CG PRO A  21      30.991  27.367  30.991  1.00  5.99           C\n",
      "\n",
      "ATOM    378   CD PRO A  21      29.853  27.780  29.853  1.00  5.99           C\n",
      "\n",
      "ATOM    387   OC PRO A  21      33.891  30.084  33.891  1.00  5.99           O\n",
      "\n",
      "ATOM    389    N GLU A  22      33.802  30.019  33.802  1.00  5.99           N\n",
      "\n",
      "ATOM    390   CA GLU A  22      35.186  29.948  35.186  1.00  5.99           C\n",
      "\n",
      "ATOM    391    C GLU A  22      35.195  28.496  35.195  1.00  5.99           C\n",
      "\n",
      "ATOM    392    O GLU A  22      34.148  27.543  34.148  1.00  5.99           O\n",
      "\n",
      "ATOM    393   CB GLU A  22      35.996  29.504  35.996  1.00  5.99           C\n",
      "\n",
      "ATOM    394   CG GLU A  22      35.554  27.523  35.554  1.00  5.99           C\n",
      "\n",
      "ATOM    395   CD GLU A  22      36.269  27.102  36.269  1.00  5.99           C\n",
      "\n",
      "ATOM    396  OE1 GLU A  22      35.969  25.611  35.969  1.00  5.99           O\n",
      "\n",
      "ATOM    397  OE2 GLU A  22      37.265  28.541  37.265  1.00  5.99           O\n",
      "\n",
      "ATOM    407   OC GLU A  22      36.500  28.299  36.500  1.00  5.99           O\n",
      "\n",
      "ATOM    408    N ALA A  23      36.521  28.394  36.521  1.00  5.99           N\n",
      "\n",
      "ATOM    409   CA ALA A  23      35.067  28.323  35.067  1.00  5.99           C\n",
      "\n",
      "ATOM    410    C ALA A  23      34.396  26.887  34.396  1.00  5.99           C\n",
      "\n",
      "ATOM    411    O ALA A  23      35.044  25.940  35.044  1.00  5.99           O\n",
      "\n",
      "ATOM    412   CB ALA A  23      34.715  28.584  34.715  1.00  5.99           C\n",
      "\n",
      "ATOM    419   OC ALA A  23      32.965  26.694  32.965  1.00  5.99           O\n",
      "\n",
      "ATOM    421    N ILE A  24      33.072  26.884  33.072  1.00  5.99           N\n",
      "\n",
      "ATOM    422   CA ILE A  24      34.515  27.107  34.515  1.00  5.99           C\n",
      "\n",
      "ATOM    423    C ILE A  24      34.834  28.583  34.834  1.00  5.99           C\n",
      "\n",
      "ATOM    424    O ILE A  24      33.945  29.447  33.945  1.00  5.99           O\n",
      "\n",
      "ATOM    425   CB ILE A  24      35.206  26.619  35.206  1.00  5.99           C\n",
      "\n",
      "ATOM    426  CG1 ILE A  24      34.859  25.122  34.859  1.00  5.99           C\n",
      "\n",
      "ATOM    427  CG2 ILE A  24      36.760  26.756  36.760  1.00  5.99           C\n",
      "\n",
      "ATOM    428  CD1 ILE A  24      35.243  24.724  35.243  1.00  5.99           C\n",
      "\n",
      "ATOM    442   OC ILE A  24      36.222  28.924  36.222  1.00  5.99           O\n",
      "\n",
      "ATOM    443    N CYS A  25      36.137  28.846  36.137  1.00  5.99           N\n",
      "\n",
      "ATOM    444   CA CYS A  25      37.585  29.033  37.585  1.00  5.99           C\n",
      "\n",
      "ATOM    445    C CYS A  25      37.941  30.501  37.941  1.00  5.99           C\n",
      "\n",
      "ATOM    446    O CYS A  25      37.075  31.387  37.075  1.00  5.99           O\n",
      "\n",
      "ATOM    447   CB CYS A  25      38.163  28.379  38.163  1.00  5.99           C\n",
      "\n",
      "ATOM    448   SG CYS A  25      39.968  28.418  39.968  1.00  5.99           S\n",
      "\n",
      "ATOM    456   OC CYS A  25      39.337  30.807  39.337  1.00  5.99           O\n",
      "\n",
      "ATOM    457    N ALA A  26      39.248  30.748  39.248  1.00  5.99           N\n",
      "\n",
      "ATOM    458   CA ALA A  26      40.708  30.736  40.708  1.00  5.99           C\n",
      "\n",
      "ATOM    459    C ALA A  26      41.262  32.141  41.262  1.00  5.99           C\n",
      "\n",
      "ATOM    460    O ALA A  26      40.527  33.134  40.527  1.00  5.99           O\n",
      "\n",
      "ATOM    461   CB ALA A  26      41.152  29.986  41.152  1.00  5.99           C\n",
      "\n",
      "ATOM    468   OC ALA A  26      42.688  32.253  42.688  1.00  5.99           O\n",
      "\n",
      "ATOM    470    N THR A  27      42.590  32.211  42.590  1.00  5.99           N\n",
      "\n",
      "ATOM    471   CA THR A  27      43.980  32.620  43.980  1.00  5.99           C\n",
      "\n",
      "ATOM    472    C THR A  27      44.212  34.152  44.212  1.00  5.99           C\n",
      "\n",
      "ATOM    473    O THR A  27      43.319  34.899  43.319  1.00  5.99           O\n",
      "\n",
      "ATOM    474   CB THR A  27      44.289  32.814  44.289  1.00  5.99           C\n",
      "\n",
      "ATOM    475  OG1 THR A  27      44.183  31.324  44.183  1.00  5.99           O\n",
      "\n",
      "ATOM    476  CG2 THR A  27      45.726  33.260  45.726  1.00  5.99           C\n",
      "\n",
      "ATOM    485   OC THR A  27      45.544  34.673  45.544  1.00  5.99           O\n",
      "\n",
      "ATOM    487    N TYR A  28      45.491  34.478  45.491  1.00  5.99           N\n",
      "\n",
      "ATOM    488   CA TYR A  28      46.871  34.583  46.871  1.00  5.99           C\n",
      "\n",
      "ATOM    489    C TYR A  28      47.736  36.043  47.736  1.00  5.99           C\n",
      "\n",
      "ATOM    490    O TYR A  28      47.282  36.982  47.282  1.00  5.99           O\n",
      "\n",
      "ATOM    491   CB TYR A  28      46.729  35.018  46.729  1.00  5.99           C\n",
      "\n",
      "ATOM    492   CG TYR A  28      45.958  33.635  45.958  1.00  5.99           C\n",
      "\n",
      "ATOM    493  CD1 TYR A  28      46.633  32.035  46.633  1.00  5.99           C\n",
      "\n",
      "ATOM    494  CD2 TYR A  28      44.562  33.969  44.562  1.00  5.99           C\n",
      "\n",
      "ATOM    495  CE1 TYR A  28      45.915  30.779  45.915  1.00  5.99           C\n",
      "\n",
      "ATOM    496  CE2 TYR A  28      43.847  32.714  43.847  1.00  5.99           C\n",
      "\n",
      "ATOM    497   CZ TYR A  28      44.525  31.120  44.525  1.00  5.99           C\n",
      "\n",
      "ATOM    509   OC TYR A  28      49.115  36.269  49.115  1.00  5.99           O\n",
      "\n",
      "ATOM    511    N THR A  29      48.973  36.030  48.973  1.00  5.99           N\n",
      "\n",
      "ATOM    512   CA THR A  29      47.628  35.532  47.628  1.00  5.99           C\n",
      "\n",
      "ATOM    513    C THR A  29      46.657  36.675  46.657  1.00  5.99           C\n",
      "\n",
      "ATOM    514    O THR A  29      47.005  37.839  47.005  1.00  5.99           O\n",
      "\n",
      "ATOM    515   CB THR A  29      46.983  35.270  46.983  1.00  5.99           C\n",
      "\n",
      "ATOM    516  OG1 THR A  29      47.846  34.100  47.846  1.00  5.99           O\n",
      "\n",
      "ATOM    517  CG2 THR A  29      45.549  34.749  45.549  1.00  5.99           C\n",
      "\n",
      "ATOM    526   OC THR A  29      45.302  36.305  45.302  1.00  5.99           O\n",
      "\n",
      "ATOM    528    N GLY A  30      45.411  36.214  45.411  1.00  5.99           N\n",
      "\n",
      "ATOM    529   CA GLY A  30      45.737  36.730  45.737  1.00  5.99           C\n",
      "\n",
      "ATOM    530    C GLY A  30      44.498  37.353  44.498  1.00  5.99           C\n",
      "\n",
      "ATOM    531    O GLY A  30      43.387  37.391  43.387  1.00  5.99           O\n",
      "\n",
      "ATOM    536   OC GLY A  30      44.696  37.894  44.696  1.00  5.99           O\n",
      "\n",
      "ATOM    538    N CYS A  31      45.252  38.435  45.252  1.00  5.99           N\n",
      "\n",
      "ATOM    539   CA CYS A  31      46.631  38.549  46.631  1.00  5.99           C\n",
      "\n",
      "ATOM    540    C CYS A  31      47.147  39.997  47.147  1.00  5.99           C\n",
      "\n",
      "ATOM    541    O CYS A  31      46.448  40.926  46.448  1.00  5.99           O\n",
      "\n",
      "ATOM    542   CB CYS A  31      46.679  38.051  46.679  1.00  5.99           C\n",
      "\n",
      "ATOM    543   SG CYS A  31      48.387  37.993  48.387  1.00  5.99           S\n",
      "\n",
      "ATOM    551   OC CYS A  31      48.492  40.233  48.492  1.00  5.99           O\n",
      "\n",
      "ATOM    552    N ILE A  32      48.409  40.152  48.409  1.00  5.99           N\n",
      "\n",
      "ATOM    553   CA ILE A  32      48.572  40.157  48.572  1.00  5.99           C\n",
      "\n",
      "ATOM    554    C ILE A  32      46.735  38.384  46.735  1.00  5.99           C\n",
      "\n",
      "ATOM    555    O ILE A  32      45.319  37.136  45.319  1.00  5.99           O\n",
      "\n",
      "ATOM    556   CB ILE A  32      50.041  40.806  50.041  1.00  5.99           C\n",
      "\n",
      "ATOM    557  CG1 ILE A  32      51.951  42.590  51.951  1.00  5.99           C\n",
      "\n",
      "ATOM    558  CG2 ILE A  32      50.343  40.940  50.343  1.00  5.99           C\n",
      "\n",
      "ATOM    559  CD1 ILE A  32      53.266  43.056  53.266  1.00  5.99           C\n",
      "\n",
      "ATOM    573   OC ILE A  32      46.726  38.232  46.726  1.00  5.99           O\n",
      "\n",
      "ATOM    574    N ILE A  33      47.243  38.409  47.243  1.00  5.99           N\n",
      "\n",
      "ATOM    575   CA ILE A  33      48.246  39.257  48.246  1.00  5.99           C\n",
      "\n",
      "ATOM    576    C ILE A  33      47.646  40.600  47.646  1.00  5.99           C\n",
      "\n",
      "ATOM    577    O ILE A  33      46.451  40.888  46.451  1.00  5.99           O\n",
      "\n",
      "ATOM    578   CB ILE A  33      48.840  39.223  48.840  1.00  5.99           C\n",
      "\n",
      "ATOM    579  CG1 ILE A  33      49.409  37.854  49.409  1.00  5.99           C\n",
      "\n",
      "ATOM    580  CG2 ILE A  33      49.983  40.052  49.983  1.00  5.99           C\n",
      "\n",
      "ATOM    581  CD1 ILE A  33      49.687  37.755  49.687  1.00  5.99           C\n",
      "\n",
      "ATOM    595   OC ILE A  33      48.539  41.517  48.539  1.00  5.99           O\n",
      "\n",
      "ATOM    596    N ILE A  34      48.624  41.495  48.624  1.00  5.99           N\n",
      "\n",
      "ATOM    597   CA ILE A  34      49.997  41.711  49.997  1.00  5.99           C\n",
      "\n",
      "ATOM    598    C ILE A  34      50.472  43.190  50.472  1.00  5.99           C\n",
      "\n",
      "ATOM    599    O ILE A  34      49.745  44.062  49.745  1.00  5.99           O\n",
      "\n",
      "ATOM    600   CB ILE A  34      50.128  41.346  50.128  1.00  5.99           C\n",
      "\n",
      "ATOM    601  CG1 ILE A  34      49.584  39.858  49.584  1.00  5.99           C\n",
      "\n",
      "ATOM    602  CG2 ILE A  34      51.601  41.473  51.601  1.00  5.99           C\n",
      "\n",
      "ATOM    603  CD1 ILE A  34      49.394  39.598  49.394  1.00  5.99           C\n",
      "\n",
      "ATOM    617   OC ILE A  34      51.808  43.524  51.808  1.00  5.99           O\n",
      "\n",
      "ATOM    618    N PRO A  35      51.728  43.413  51.728  1.00  5.99           N\n",
      "\n",
      "ATOM    619   CA PRO A  35      53.171  43.561  53.171  1.00  5.99           C\n",
      "\n",
      "ATOM    620    C PRO A  35      53.558  45.392  53.558  1.00  5.99           C\n",
      "\n",
      "ATOM    621    O PRO A  35      52.683  46.583  52.683  1.00  5.99           O\n",
      "\n",
      "ATOM    622   CB PRO A  35      53.261  42.610  53.261  1.00  5.99           C\n",
      "\n",
      "ATOM    623   CG PRO A  35      51.818  42.977  51.818  1.00  5.99           C\n",
      "\n",
      "ATOM    624   CD PRO A  35      50.923  42.800  50.923  1.00  5.99           C\n",
      "\n",
      "ATOM    633   OC PRO A  35      54.876  45.745  54.876  1.00  5.99           O\n",
      "\n",
      "ATOM    635    N GLY A  36      54.812  45.474  54.812  1.00  5.99           N\n",
      "\n",
      "ATOM    636   CA GLY A  36      55.864  45.825  55.864  1.00  5.99           C\n",
      "\n",
      "ATOM    637    C GLY A  36      54.996  44.421  54.996  1.00  5.99           C\n",
      "\n",
      "ATOM    638    O GLY A  36      53.588  43.175  53.588  1.00  5.99           O\n",
      "\n",
      "ATOM    643   OC GLY A  36      55.911  44.628  55.911  1.00  5.99           O\n",
      "\n",
      "ATOM    645    N ALA A  37      56.314  44.243  56.314  1.00  5.99           N\n",
      "\n",
      "ATOM    646   CA ALA A  37      54.863  44.182  54.863  1.00  5.99           C\n",
      "\n",
      "ATOM    647    C ALA A  37      54.172  42.747  54.172  1.00  5.99           C\n",
      "\n",
      "ATOM    648    O ALA A  37      54.803  41.795  54.803  1.00  5.99           O\n",
      "\n",
      "ATOM    649   CB ALA A  37      54.570  44.381  54.570  1.00  5.99           C\n",
      "\n",
      "ATOM    656   OC ALA A  37      52.743  42.563  52.743  1.00  5.99           O\n",
      "\n",
      "ATOM    658    N THR A  38      52.854  42.753  52.854  1.00  5.99           N\n",
      "\n",
      "ATOM    659   CA THR A  38      54.270  43.101  54.270  1.00  5.99           C\n",
      "\n",
      "ATOM    660    C THR A  38      54.462  44.600  54.462  1.00  5.99           C\n",
      "\n",
      "ATOM    661    O THR A  38      53.523  45.368  53.523  1.00  5.99           O\n",
      "\n",
      "ATOM    662   CB THR A  38      54.795  43.322  54.795  1.00  5.99           C\n",
      "\n",
      "ATOM    663  OG1 THR A  38      54.719  41.858  54.719  1.00  5.99           O\n",
      "\n",
      "ATOM    664  CG2 THR A  38      56.281  43.710  56.281  1.00  5.99           C\n",
      "\n",
      "ATOM    673   OC THR A  38      55.815  45.060  55.815  1.00  5.99           O\n",
      "\n",
      "ATOM    675    N THR A  39      55.765  44.869  55.765  1.00  5.99           N\n",
      "\n",
      "ATOM    676   CA THR A  39      57.204  44.923  57.204  1.00  5.99           C\n",
      "\n",
      "ATOM    677    C THR A  39      57.852  46.352  57.852  1.00  5.99           C\n",
      "\n",
      "ATOM    678    O THR A  39      57.209  47.295  57.209  1.00  5.99           O\n",
      "\n",
      "ATOM    679   CB THR A  39      57.452  44.917  57.452  1.00  5.99           C\n",
      "\n",
      "ATOM    680  OG1 THR A  39      56.941  43.504  56.941  1.00  5.99           O\n",
      "\n",
      "ATOM    681  CG2 THR A  39      58.936  44.983  58.936  1.00  5.99           C\n",
      "\n",
      "ATOM    690   OC THR A  39      59.267  46.529  59.267  1.00  5.99           O\n",
      "\n",
      "ATOM    692    N CYS A  40      59.161  46.370  59.161  1.00  5.99           N\n",
      "\n",
      "ATOM    693   CA CYS A  40      57.760  46.003  57.760  1.00  5.99           C\n",
      "\n",
      "ATOM    694    C CYS A  40      56.886  47.231  56.886  1.00  5.99           C\n",
      "\n",
      "ATOM    695    O CYS A  40      57.352  48.377  57.352  1.00  5.99           O\n",
      "\n",
      "ATOM    696   CB CYS A  40      57.197  45.760  57.197  1.00  5.99           C\n",
      "\n",
      "ATOM    697   SG CYS A  40      55.533  45.100  55.533  1.00  5.99           S\n",
      "\n",
      "ATOM    705   OC CYS A  40      55.489  46.994  55.489  1.00  5.99           O\n",
      "\n",
      "ATOM    706    N PRO A  41      55.605  46.873  55.605  1.00  5.99           N\n",
      "\n",
      "ATOM    707   CA PRO A  41      54.263  46.907  54.263  1.00  5.99           C\n",
      "\n",
      "ATOM    708    C PRO A  41      53.740  45.213  53.740  1.00  5.99           C\n",
      "\n",
      "ATOM    709    O PRO A  41      54.437  43.976  54.437  1.00  5.99           O\n",
      "\n",
      "ATOM    710   CB PRO A  41      54.550  47.908  54.550  1.00  5.99           C\n",
      "\n",
      "ATOM    711   CG PRO A  41      56.056  47.450  56.056  1.00  5.99           C\n",
      "\n",
      "ATOM    712   CD PRO A  41      56.704  47.451  56.704  1.00  5.99           C\n",
      "\n",
      "ATOM    721   OC PRO A  41      52.495  45.037  52.495  1.00  5.99           O\n",
      "\n",
      "ATOM    723    N GLY A  42      52.568  45.274  52.568  1.00  5.99           N\n",
      "\n",
      "ATOM    724   CA GLY A  42      52.337  45.208  52.337  1.00  5.99           C\n",
      "\n",
      "ATOM    725    C GLY A  42      50.854  45.127  50.854  1.00  5.99           C\n",
      "\n",
      "ATOM    726    O GLY A  42      49.998  45.122  49.998  1.00  5.99           O\n",
      "\n",
      "ATOM    731   OC GLY A  42      50.507  45.058  50.507  1.00  5.99           O\n",
      "\n",
      "ATOM    733    N ASP A  43      50.580  44.721  50.580  1.00  5.99           N\n",
      "\n",
      "ATOM    734   CA ASP A  43      52.006  45.093  52.006  1.00  5.99           C\n",
      "\n",
      "ATOM    735    C ASP A  43      52.394  46.615  52.394  1.00  5.99           C\n",
      "\n",
      "ATOM    736    O ASP A  43      51.621  47.526  51.621  1.00  5.99           O\n",
      "\n",
      "ATOM    737   CB ASP A  43      52.417  44.369  52.417  1.00  5.99           C\n",
      "\n",
      "ATOM    738   CG ASP A  43      53.943  44.224  53.943  1.00  5.99           C\n",
      "\n",
      "ATOM    739  OD1 ASP A  43      54.679  44.433  54.679  1.00  5.99           O\n",
      "\n",
      "ATOM    740  OD2 ASP A  43      54.239  43.902  54.239  1.00  5.99           O\n",
      "\n",
      "ATOM    746   OC ASP A  43      53.704  46.887  53.704  1.00  5.99           O\n",
      "\n",
      "ATOM    748    N TYR A  44      53.690  46.790  53.690  1.00  5.99           N\n",
      "\n",
      "ATOM    749   CA TYR A  44      54.373  47.445  54.373  1.00  5.99           C\n",
      "\n",
      "ATOM    750    C TYR A  44      53.608  48.580  53.608  1.00  5.99           C\n",
      "\n",
      "ATOM    751    O TYR A  44      52.550  48.869  52.550  1.00  5.99           O\n",
      "\n",
      "ATOM    752   CB TYR A  44      53.976  47.656  53.976  1.00  5.99           C\n",
      "\n",
      "ATOM    753   CG TYR A  44      54.734  46.609  54.734  1.00  5.99           C\n",
      "\n",
      "ATOM    754  CD1 TYR A  44      56.450  46.044  56.450  1.00  5.99           C\n",
      "\n",
      "ATOM    755  CD2 TYR A  44      53.705  46.207  53.705  1.00  5.99           C\n",
      "\n",
      "ATOM    756  CE1 TYR A  44      57.128  45.085  57.128  1.00  5.99           C\n",
      "\n",
      "ATOM    757  CE2 TYR A  44      54.385  45.250  54.385  1.00  5.99           C\n",
      "\n",
      "ATOM    758   CZ TYR A  44      56.097  44.689  56.097  1.00  5.99           C\n",
      "\n",
      "ATOM    770   OC TYR A  44      54.185  49.297  54.185  1.00  5.99           O\n",
      "\n",
      "ATOM    772    N ALA A  45      54.465  49.492  54.465  1.00  5.99           N\n",
      "\n",
      "ATOM    773   CA ALA A  45      55.540  49.662  55.540  1.00  5.99           C\n",
      "\n",
      "ATOM    774    C ALA A  45      56.361  51.228  56.361  1.00  5.99           C\n",
      "\n",
      "ATOM    775    O ALA A  45      56.118  52.188  56.118  1.00  5.99           O\n",
      "\n",
      "ATOM    776   CB ALA A  45      54.697  49.562  54.697  1.00  5.99           C\n",
      "\n",
      "ATOM    783   OC ALA A  45      57.448  51.526  57.448  1.00  5.99           O\n",
      "\n",
      "ATOM    785    N ASN A  46      57.398  51.251  57.398  1.00  5.99           N\n",
      "\n",
      "ATOM    786   CA ASN A  46      56.491  50.632  56.491  1.00  5.99           C\n",
      "\n",
      "ATOM    787    C ASN A  46      55.231  51.409  55.231  1.00  5.99           C\n",
      "\n",
      "ATOM    788    O ASN A  46      55.048  52.425  55.048  1.00  5.99           O\n",
      "\n",
      "ATOM    789   CB ASN A  46      55.737  50.283  55.737  1.00  5.99           C\n",
      "\n",
      "ATOM    790   CG ASN A  46      56.869  49.497  56.869  1.00  5.99           C\n",
      "\n",
      "ATOM    791  OD1 ASN A  46      57.060  49.919  57.060  1.00  5.99           O\n",
      "\n",
      "ATOM    792  ND2 ASN A  46      57.676  48.341  57.676  1.00  5.99           N\n",
      "\n",
      "ATOM    801   OC ASN A  46      54.264  50.894  54.264  1.00  5.99           O\n",
      "\n"
     ]
    }
   ],
   "source": [
    "atom_seq_number = 0\n",
    "\n",
    "with open('assets/final_inferred_molecule.pdb', 'w') as f:\n",
    "    for idx, aminoacid in enumerate(aminoacids_df):\n",
    "        for atom_row in aminoacid.iterrows():\n",
    "            write_atom_to_file(atom_row[1], idx, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## b) Receive PDB file and list $\\phi$ and $\\psi$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>element</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>h</th>\n",
       "      <th>residue_seq_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>THR</td>\n",
       "      <td>17.047</td>\n",
       "      <td>14.099</td>\n",
       "      <td>3.625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>16.967</td>\n",
       "      <td>12.784</td>\n",
       "      <td>4.338</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>THR</td>\n",
       "      <td>15.685</td>\n",
       "      <td>12.755</td>\n",
       "      <td>5.133</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name element aminoacid       x       y      z  h  residue_seq_number\n",
       "0    N       N       THR  17.047  14.099  3.625  1                   1\n",
       "1   CA       C       THR  16.967  12.784  4.338  1                   1\n",
       "2    C       C       THR  15.685  12.755  5.133  1                   1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crambin = read_molecule('assets/1crn.pdb')\n",
    "crambin[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$ ψ  = Normal(N_{i}, C_{\\alpha i}, C_{i}) \\^ Normal(C_{\\alpha i}, C_{i}, N_{i+1})  $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$φ = Normal(C_{i-1}, N_{i}, C_{\\alpha i}) \\^ Normal(N_{i}, C_{\\alpha i}, C_{i}) $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![assets/phi_psi.png](assets/phi_psi.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_psi(ni, cai, ci, ni1):\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    normal_ci = np.cross((cai - ci), (ni1 - ci))\n",
    "    return np.degrees(get_angle_rad(normal_cai, normal_ci))\n",
    "\n",
    "def compute_phi(ci_prev, ni, cai, ci):\n",
    "    normal_ni = np.cross((ci_prev - ni), (cai - ni))\n",
    "    normal_cai = np.cross((ni - cai), (ci - cai))\n",
    "    return np.degrees(get_angle_rad(normal_ni, normal_cai))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_phy_psi(molecule):\n",
    "    molecule['phi'] = None\n",
    "    molecule['psi'] = None\n",
    "\n",
    "    data = []\n",
    "\n",
    "    for i in range(1, len(molecule.residue_seq_number.unique())):\n",
    "        aminoacid_i = molecule[molecule.residue_seq_number == (i)]\n",
    "        ni = aminoacid_i[aminoacid_i.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        cai = aminoacid_i[aminoacid_i.name == 'CA'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        ci = aminoacid_i[aminoacid_i.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "        angle_data = {'residue_name': aminoacid_i.aminoacid.values[0], 'phi': None, 'psi': None}\n",
    "        try:\n",
    "            aminoacid_prev = molecule[molecule.residue_seq_number == (i-1)]\n",
    "            ci_prev = aminoacid_prev[aminoacid_prev.name == 'C'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['phi'] = compute_phi(ci_prev, ni, cai, ci)\n",
    "        except: pass\n",
    "\n",
    "        try:\n",
    "            aminoacid_i1 = molecule[molecule.residue_seq_number == (i+1)]\n",
    "            ni1 = aminoacid_i1[aminoacid_i1.name == 'N'][['x', 'y', 'z']].values.reshape(-1,)\n",
    "            angle_data['psi'] = compute_psi(ni, cai, ci, ni1)\n",
    "        except: pass\n",
    "\n",
    "        data.append(angle_data)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\python36\\lib\\site-packages\\ipykernel_launcher.py:8: RuntimeWarning: invalid value encountered in arccos\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>LEU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>179.999999</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ASP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>8.537736e-07</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1.800000e+02</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             phi         psi residue_name\n",
       "0   1.800000e+02  180.000000          THR\n",
       "1   1.800000e+02  179.999999          CYS\n",
       "2   0.000000e+00    0.000000          CYS\n",
       "3   1.800000e+02    0.000000          PRO\n",
       "4   1.800000e+02    0.000000          SER\n",
       "5   1.800000e+02  179.999999          ILE\n",
       "6   1.800000e+02  180.000000          VAL\n",
       "7   0.000000e+00  179.999999          ALA\n",
       "8   0.000000e+00  180.000000          ARG\n",
       "9   0.000000e+00  180.000000          SER\n",
       "10  1.800000e+02  179.999999          ASN\n",
       "11  1.800000e+02  180.000000          PHE\n",
       "12  0.000000e+00    0.000000          ASN\n",
       "13  0.000000e+00  180.000000          VAL\n",
       "14  0.000000e+00    0.000000          CYS\n",
       "15  1.800000e+02  180.000000          ARG\n",
       "16  1.800000e+02  180.000000          LEU\n",
       "17  0.000000e+00    0.000000          PRO\n",
       "18  0.000000e+00  180.000000          GLY\n",
       "19  1.800000e+02  180.000000          THR\n",
       "20  1.800000e+02  179.999999          PRO\n",
       "21  0.000000e+00  180.000000          GLU\n",
       "22  1.800000e+02  180.000000          ALA\n",
       "23  1.800000e+02  180.000000          ILE\n",
       "24  1.800000e+02  180.000000          CYS\n",
       "25  1.800000e+02  180.000000          ALA\n",
       "26  0.000000e+00    0.000000          THR\n",
       "27  1.800000e+02  180.000000          TYR\n",
       "28  0.000000e+00  179.999999          THR\n",
       "29  1.800000e+02  180.000000          GLY\n",
       "30  1.800000e+02    0.000000          CYS\n",
       "31  0.000000e+00  180.000000          ILE\n",
       "32  0.000000e+00  180.000000          ILE\n",
       "33  1.800000e+02  179.999999          ILE\n",
       "34  0.000000e+00    0.000000          PRO\n",
       "35  1.800000e+02  180.000000          GLY\n",
       "36  0.000000e+00    0.000000          ALA\n",
       "37  1.800000e+02  180.000000          THR\n",
       "38  1.800000e+02  180.000000          THR\n",
       "39  0.000000e+00  180.000000          CYS\n",
       "40  1.800000e+02  180.000000          PRO\n",
       "41  1.800000e+02  180.000000          GLY\n",
       "42  0.000000e+00    0.000000          ASP\n",
       "43  8.537736e-07  180.000000          TYR\n",
       "44  1.800000e+02  180.000000          ALA\n",
       "45  0.000000e+00         NaN          ASN"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/final_inferred_molecule.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## c) Ramachandran "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>residue_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>147.660026</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107.832573</td>\n",
       "      <td>144.346241</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>131.170389</td>\n",
       "      <td>133.291143</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>118.896523</td>\n",
       "      <td>151.202717</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>76.181506</td>\n",
       "      <td>18.979386</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>157.924288</td>\n",
       "      <td>166.027161</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>63.579524</td>\n",
       "      <td>42.100632</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>55.930329</td>\n",
       "      <td>44.561332</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>61.364845</td>\n",
       "      <td>43.779138</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>63.193412</td>\n",
       "      <td>43.303486</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>61.208800</td>\n",
       "      <td>42.430336</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>64.920400</td>\n",
       "      <td>39.489513</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>59.102393</td>\n",
       "      <td>47.211965</td>\n",
       "      <td>PHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>62.823308</td>\n",
       "      <td>35.231846</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>69.239667</td>\n",
       "      <td>41.218779</td>\n",
       "      <td>VAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>56.671614</td>\n",
       "      <td>36.000832</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>77.068519</td>\n",
       "      <td>16.126707</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>53.208796</td>\n",
       "      <td>46.167676</td>\n",
       "      <td>LEU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>77.189679</td>\n",
       "      <td>7.597741</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>106.257207</td>\n",
       "      <td>7.311793</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>52.674731</td>\n",
       "      <td>136.344219</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>56.977153</td>\n",
       "      <td>146.617514</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>56.412840</td>\n",
       "      <td>36.191815</td>\n",
       "      <td>GLU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>63.430344</td>\n",
       "      <td>34.860660</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>74.766620</td>\n",
       "      <td>37.894498</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>64.948033</td>\n",
       "      <td>31.686198</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>62.036226</td>\n",
       "      <td>54.031875</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>68.778289</td>\n",
       "      <td>25.486562</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>67.592836</td>\n",
       "      <td>36.300194</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>108.956767</td>\n",
       "      <td>18.473152</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>91.815951</td>\n",
       "      <td>3.067675</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>69.515392</td>\n",
       "      <td>164.375577</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>129.761553</td>\n",
       "      <td>157.032715</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>111.563161</td>\n",
       "      <td>129.588363</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>124.130558</td>\n",
       "      <td>158.127320</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>78.150912</td>\n",
       "      <td>24.351838</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>89.739760</td>\n",
       "      <td>161.711226</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>120.806412</td>\n",
       "      <td>1.238816</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>114.625917</td>\n",
       "      <td>104.511918</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>75.525053</td>\n",
       "      <td>145.356696</td>\n",
       "      <td>CYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>71.200027</td>\n",
       "      <td>162.749132</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>61.887021</td>\n",
       "      <td>23.112126</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>89.704786</td>\n",
       "      <td>2.008555</td>\n",
       "      <td>ASP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>130.332598</td>\n",
       "      <td>60.007908</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>88.769547</td>\n",
       "      <td>2.529881</td>\n",
       "      <td>ALA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           phi         psi residue_name\n",
       "0          NaN  147.660026          THR\n",
       "1   107.832573  144.346241          THR\n",
       "2   131.170389  133.291143          CYS\n",
       "3   118.896523  151.202717          CYS\n",
       "4    76.181506   18.979386          PRO\n",
       "5   157.924288  166.027161          SER\n",
       "6    63.579524   42.100632          ILE\n",
       "7    55.930329   44.561332          VAL\n",
       "8    61.364845   43.779138          ALA\n",
       "9    63.193412   43.303486          ARG\n",
       "10   61.208800   42.430336          SER\n",
       "11   64.920400   39.489513          ASN\n",
       "12   59.102393   47.211965          PHE\n",
       "13   62.823308   35.231846          ASN\n",
       "14   69.239667   41.218779          VAL\n",
       "15   56.671614   36.000832          CYS\n",
       "16   77.068519   16.126707          ARG\n",
       "17   53.208796   46.167676          LEU\n",
       "18   77.189679    7.597741          PRO\n",
       "19  106.257207    7.311793          GLY\n",
       "20   52.674731  136.344219          THR\n",
       "21   56.977153  146.617514          PRO\n",
       "22   56.412840   36.191815          GLU\n",
       "23   63.430344   34.860660          ALA\n",
       "24   74.766620   37.894498          ILE\n",
       "25   64.948033   31.686198          CYS\n",
       "26   62.036226   54.031875          ALA\n",
       "27   68.778289   25.486562          THR\n",
       "28   67.592836   36.300194          TYR\n",
       "29  108.956767   18.473152          THR\n",
       "30   91.815951    3.067675          GLY\n",
       "31   69.515392  164.375577          CYS\n",
       "32  129.761553  157.032715          ILE\n",
       "33  111.563161  129.588363          ILE\n",
       "34  124.130558  158.127320          ILE\n",
       "35   78.150912   24.351838          PRO\n",
       "36   89.739760  161.711226          GLY\n",
       "37  120.806412    1.238816          ALA\n",
       "38  114.625917  104.511918          THR\n",
       "39   75.525053  145.356696          CYS\n",
       "40   71.200027  162.749132          PRO\n",
       "41   61.887021   23.112126          GLY\n",
       "42   89.704786    2.008555          ASP\n",
       "43  130.332598   60.007908          TYR\n",
       "44   88.769547    2.529881          ALA"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecule = read_molecule('assets/1crn.pdb')\n",
    "phi_psi = pd.DataFrame(get_phy_psi(molecule))\n",
    "phi_psi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "phi_psi = phi_psi.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnIAAAJSCAYAAABHgiijAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHQZJREFUeJzt3W+MZfdd3/HPF68JaTYoQXYmJl5qUxlFSSQsPDKpKtCsCNhxHziO5MqpSgIBLaYJrdQ+aCI/IG20CKFCpNCQdGmsOEKwtaBOrGAIicWQVhAldnGDncSw64TsxlacfwU2gFknvz6Yu+10mdnde+OZc787r5c0mrnnnjvzm+9ezb73nDt7aowRAAD6+ZapFwAAwGKEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKb2Tb2A3XLZZZeNq666auplzO1rX/tanvOc50y9jHbMbX5mthhzm5+ZLcbc5td5Zg8++OCXxhiXn2+/PRNyV111VR544IGplzG39fX1rK2tTb2Mdsxtfma2GHObn5ktxtzm13lmVfXnF7KfU6sAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBT+6ZeAAB8M06cOLGjn//AgQNzP2beNS3yNXbaxfA9nD59+qL4Ps7FETkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0NS+qRcAAN+MZbzI+TKuaact48XpL7300ov+z8IROQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaWoqQq6o7q+rJqnp407a3VNXnq+qh2dtNm+57c1Udq6pHq+qGaVYNADCtpQi5JO9JcuMW2982xrh29nZfklTVS5LcluSls8f8SlVdsmsrBQBYEksRcmOMjyT5ygXufnOSo2OMp8YYn0lyLMn1O7Y4AIAlVWOMqdeQJKmqq5J8YIzxstnttyT5sSR/meSBJP92jPHVqvpPST46xvi12X7vTvI7Y4zf3OJzHkpyKElWVlauO3r06M5/I8+wU6dOZf/+/VMvox1zm5+ZLcbc5mdmi9ntuZ0+fXpHP/+ll166o58/6f1cO3jw4INjjNXz7bfM11p9Z5K3Jhmz97+Y5PVJaot9t6zRMcaRJEeSZHV1daytre3IQnfS+vp6Oq57auY2PzNbjLnNz8wWs9tzm/faqfPajWug7oXn2tKG3BjjC2c+rqpfTfKB2c2TSTb/6V+Z5PFdXBpAko2/6E6fPn3Bf+Fd7Bfvfqbs9MXXl/Hi7vPaje9hGb9v/r6leI3cVqrqik03b0ly5jda701yW1U9q6quTnJNko/t9voAAKa2FEfkquo3kqwluayqTib52SRrVXVtNk6bfjbJTyXJGOORqro7ySeTPJ3kDWOMr0+xbgCAKS1FyI0xXrPF5nefY//DSQ7v3IoAAJbf0p5aBQDg3IQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0txX8IDNDRgQMHcvz4cdekPI+zrwt6vuvT7vQ8L4Y/r2X8Hi6Ga9h25IgcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhq39QLAODidvbF0Y8fP+6C6bts3gvaJ/Nf1N6f6TQckQMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJpyrVUAuMi5DurFyxE5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATe2begEAwM46ceLE3I85cODADqyEZ5ojcgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJpaipCrqjur6smqenjTtu+oqg9V1Z/N3j9/tr2q6u1VdayqPlFV3zfdygEAprMUIZfkPUluPGvbm5LcP8a4Jsn9s9tJ8sok18zeDiV55y6tEQBgqSxFyI0xPpLkK2dtvjnJXbOP70ryqk3b3zs2fDTJ86rqit1ZKQDA8qgxxtRrSJJU1VVJPjDGeNns9v8eYzxv0/1fHWM8v6o+kOTnxxj/Y7b9/iT/bozxwBaf81A2jtplZWXluqNHj+78N/IMO3XqVPbv3z/1Mtoxt/mZ2WLMbX5mthhzm1/nmR08ePDBMcbq+fbbtxuLeYbVFtu2rNExxpEkR5JkdXV1rK2t7eCydsb6+no6rntq5jY/M1uMuc3PzBZjbvPbCzNbilOr2/jCmVOms/dPzrafTHJg035XJnl8l9cGADC5ZQ65e5O8bvbx65K8f9P2185+e/XlSf5ijPHEFAsEAJjSUpxararfSLKW5LKqOpnkZ5P8fJK7q+onknwuya2z3e9LclOSY0n+OsmP7/qCAQCWwFKE3BjjNdvc9UNb7DuSvGFnVwQAsPyW+dQqAADnIOQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICm9k29AACgvxMnTsy1/4EDB3ZoJXuLI3IAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoKl9Uy8AADZz8fWe/DlMwxE5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATS39fwhcVZ9N8ldJvp7k6THGalV9R5L/muSqJJ9N8s/GGF+dao0AAFPockTu4Bjj2jHG6uz2m5LcP8a4Jsn9s9sAAHtKl5A7281J7pp9fFeSV024FgCASSz9qdUkI8nvVdVI8p/HGEeSrIwxnkiSMcYTVfWCSVcIwLbOvnbq6dOnz3k91Xe9611zff7bb799rv1dE5SLSY0xpl7DOVXVd44xHp/F2oeS/EySe8cYz9u0z1fHGM/f4rGHkhxKkpWVleuOHj26W8t+xpw6dSr79++fehntmNv8zGwx5nZ+p0+f/v9u/+3f/m2+7du+bdv9v/jFL871+S+//PK59r/00kvn2n9ZeK7Nr/PMDh48+OCml5Rta+mPyI0xHp+9f7Kq7klyfZIvVNUVs6NxVyR5cpvHHklyJElWV1fH2traLq36mbO+vp6O656auc3PzBZjbud39tG3T3/603nxi1+87f7r6+tzff69ckTOc21+e2FmS/0auap6TlU998zHSX4kycNJ7k3yutlur0vy/mlWCAAwnWU/IreS5J6qSjbW+utjjN+tqo8nubuqfiLJ55LcOuEaAQAmsdQhN8Z4LMn3brH9y0l+aPdXBACwPJb61CoAANsTcgAATQk5AICmhBwAQFNCDgCgKSEHANDUUv/3IwD0d/aVFI4fP37Oqyv84R/+4Vyf/+d+7ufm2v9zn/vcXPt3vRIEe4MjcgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgqX1TLwAANvv93//9ufY/ceLEXPsfOHBgrv1hmTkiBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANOVaqwAslXmvnTqv++67b679b7rpph1ayeJcX5YzHJEDAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE3tm3oBALCZC7yfnxlxhiNyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU661CgDPsDvuuGOu/Q8fPjzX/idOnJhr/0WuzbobX4NvniNyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNtQ25qrqxqh6tqmNV9aap1wMAsNvOe4muqnrt7MNTSR4bYzy0s0s6v6q6JMk7kvxwkpNJPl5V944xPjntygAAds+FXGv14Oz9P0jy4qoaSX5yjPHAzi3rvK5PcmyM8ViSVNXRJDcnEXIAwJ5x3pAbY/z45ttV9U+TvK+qvneM8eUdW9m5vSjJ5qv5nkzy/ROtBYBGduPi7ocPH97Rz78b38NufA2+eTXGOPcOVf9mi82vT/LRJJ8cY/zSTizsXKrq1iQ3jDF+cnb7R5NcP8b4mbP2O5TkUJKsrKxcd/To0d1e6jft1KlT2b9//9TLaMfc5mdmizG3+ZnZYsxtfp1ndvDgwQfHGKvn2+9CTq0+d4tt35KNU61TTedkks3/VLgyyeNn7zTGOJLkSJKsrq6OtbW1XVncM2l9fT0d1z01c5ufmS3G3OZnZosxt/nthZldyKnVf7/5dlX9cDaOcv3rMcYXd2ph5/HxJNdU1dVJPp/ktiT/fKK1AABM4kJ+a/XO2YfPTvLi2WNunTDiMsZ4uqremOSDSS5JcucY45Gp1gMAMIULObX6B7P3X0vyWJI/Hud7Yd0uGGPcl+S+qdcBADCVCzm1etduLAQAgPm0vbIDAMBeJ+QAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaErIAQA0JeQAAJpa2pCrqrdU1eer6qHZ202b7ntzVR2rqker6oYp1wkAMJV9Uy/gPN42xviPmzdU1UuS3JbkpUm+M8mHq+p7xhhfn2KBAABTWdojcudwc5KjY4ynxhifSXIsyfUTrwkAYNcte8i9sao+UVV3VtXzZ9telOTEpn1OzrYBAOwpNcaY7otXfTjJC7e4644kH03ypSQjyVuTXDHGeH1VvSPJH40xfm32Od6d5L4xxm9t8fkPJTmUJCsrK9cdPXp0Z76RHXTq1Kns379/6mW0Y27zM7PFmNv8zGwx5ja/zjM7ePDgg2OM1fPtN+lr5MYYr7iQ/arqV5N8YHbzZJIDm+6+Msnj23z+I0mOJMnq6upYW1tbeK1TWV9fT8d1T83c5mdmizG3+ZnZYsxtfnthZkt7arWqrth085YkD88+vjfJbVX1rKq6Osk1ST622+sDAJjaMv/W6i9U1bXZOLX62SQ/lSRjjEeq6u4kn0zydJI3+I1VAGAvWtqQG2P86DnuO5zk8C4uBwBg6SztqVUAAM5NyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACamjTkqurWqnqkqr5RVatn3ffmqjpWVY9W1Q2btt8423asqt60+6sGAFgOUx+RezjJq5N8ZPPGqnpJktuSvDTJjUl+paouqapLkrwjySuTvCTJa2b7AgDsOfum/OJjjE8lSVWdfdfNSY6OMZ5K8pmqOpbk+tl9x8YYj80ed3S27yd3Z8UAAMtj0pA7hxcl+eim2ydn25LkxFnbv3+7T1JVh5IcSpKVlZWsr68/s6vcBadOnWq57qmZ2/zMbDHmNj8zW4y5zW8vzGzHQ66qPpzkhVvcdccY4/3bPWyLbSNbnwoe233tMcaRJEeSZHV1daytrZ17sUtofX09Hdc9NXObn5ktxtzmZ2aLMbf57YWZ7XjIjTFescDDTiY5sOn2lUken3283XYAgD1l6l922M69SW6rqmdV1dVJrknysSQfT3JNVV1dVd+ajV+IuHfCdQIATGbS18hV1S1JfjnJ5Ul+u6oeGmPcMMZ4pKruzsYvMTyd5A1jjK/PHvPGJB9MckmSO8cYj0y0fACASU39W6v3JLlnm/sOJzm8xfb7kty3w0sDAFh6y3pqFQCA8xByAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBADQl5AAAmhJyAABNTRpyVXVrVT1SVd+oqtVN26+qqr+pqodmb+/adN91VfUnVXWsqt5eVTXN6gEApjX1EbmHk7w6yUe2uO/4GOPa2dvtm7a/M8mhJNfM3m7c+WUCACyfSUNujPGpMcajF7p/VV2R5NvHGH80xhhJ3pvkVTu2QACAJTb1Eblzubqq/riq/qCqfmC27UVJTm7a5+RsGwDAnrNvp79AVX04yQu3uOuOMcb7t3nYE0m+a4zx5aq6Lsn7quqlSbZ6Pdw4x9c+lI3TsFlZWcn6+vpca18Gp06darnuqZnb/MxsMeY2PzNbjLnNby/MbMdDbozxigUe81SSp2YfP1hVx5N8TzaOwF25adcrkzx+js9zJMmRJFldXR1ra2vzLmVy6+vr6bjuqZnb/MxsMeY2PzNbjLnNby/MbClPrVbV5VV1yezj787GLzU8NsZ4IslfVdXLZ7+t+tok2x3VAwC4qE3934/cUlUnk/zjJL9dVR+c3fWDST5RVf8ryW8muX2M8ZXZfT+d5L8kOZbkeJLf2eVlAwAshR0/tXouY4x7ktyzxfbfSvJb2zzmgSQv2+GlAQAsvaU8tQoAwPkJOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaEnIAAE0JOQCApoQcAEBTQg4AoCkhBwDQlJADAGhKyAEANCXkAACaqjHG1GvYFVX1xSR/PvU6FnBZki9NvYiGzG1+ZrYYc5ufmS3G3ObXeWb/cIxx+fl22jMh11VVPTDGWJ16Hd2Y2/zMbDHmNj8zW4y5zW8vzMypVQCApoQcAEBTQm75HZl6AU2Z2/zMbDHmNj8zW4y5ze+in5nXyAEANOWIHABAU0JuSVTVrVX1SFV9o6pWN22/qqr+pqoemr29a9N911XVn1TVsap6e1XVNKufznZzm9335tlsHq2qGzZtv3G27VhVvWn3V71cquotVfX5Tc+xmzbdt+UM8TyaR1V9dvaz6qGqemC27Tuq6kNV9Wez98+fep1Tqqo7q+rJqnp407YtZ1Qb3j577n2iqr5vupVPa5u57amfaUJueTyc5NVJPrLFfcfHGNfO3m7ftP2dSQ4luWb2duPOL3PpbDm3qnpJktuSvDQbc/mVqrqkqi5J8o4kr0zykiSvme27171t03PsvmT7GU65yGXhebSQg7Pn15l/cL0pyf1jjGuS3D+7vZe9J3//Z/h2M3pl/t/P/UPZ+Ltgr3pPtv67b8/8TBNyS2KM8akxxqMXun9VXZHk28cYfzQ2Xuj43iSv2rEFLqlzzO3mJEfHGE+NMT6T5FiS62dvx8YYj40x/i7J0dm+/H3bzRDPo2fCzUnumn18V/bgz6/NxhgfSfKVszZvN6Obk7x3bPhokufN/k7Yc7aZ23Yuyp9pQq6Hq6vqj6vqD6rqB2bbXpTk5KZ9Ts62seFFSU5sun1mPttt3+veODtFc+emU1xmtT2zmc9I8ntV9WBVHZptWxljPJEks/cvmGx1y2u7GXn+nd+e+Zm2b+oF7CVV9eEkL9zirjvGGO/f5mFPJPmuMcaXq+q6JO+rqpcm2er1cBflryAvOLft5rPVP14uyrltdq4ZZuO0zFuzMYe3JvnFJK/PHnqOLcBs5vNPxhiPV9ULknyoqj499YKa8/w7tz31M03I7aIxxisWeMxTSZ6affxgVR1P8j3Z+JfElZt2vTLJ48/EOpfNInPLxnwObLq9eT7bbb9oXegMq+pXk3xgdvNcM9zrzGYOY4zHZ++frKp7snE66wtVdcUY44nZacEnJ13kctpuRp5/5zDG+MKZj/fCzzSnVpdcVV1+5sWYVfXd2Xhx62Ozw+x/VVUvn/226muTbHd0ai+6N8ltVfWsqro6G3P7WJKPJ7mmqq6uqm/Nxgtf751wnZM767U1t2TjF0iS7WeI59EFq6rnVNVzz3yc5Eey8Ry7N8nrZru9Ln5+bWW7Gd2b5LWz3159eZK/OHMKlr33M80RuSVRVbck+eUklyf57ap6aIxxQ5IfTPIfqurpJF9PcvsY48wLO386G7+x8+wkvzN721O2m9sY45GqujvJJ5M8neQNY4yvzx7zxiQfTHJJkjvHGI9MtPxl8QtVdW02TjF8NslPJcm5ZrjXjTGe9jy6YCtJ7tn492b2Jfn1McbvVtXHk9xdVT+R5HNJbp1wjZOrqt9Ispbksqo6meRnk/x8tp7RfUluysaL9f86yY/v+oKXxDZzW9tLP9Nc2QEAoCmnVgEAmhJyAABNCTkAgKaEHABAU0IOAKApIQcA0JSQAwBoSsgBLKiq3l1V/3LqdQB7l/8QGGBOVXVzkrcleW6S00k+n+THXN0B2G1CDmAOVfWPkvz3bFwW6F8leSDJXyY5nORlF8Mlf4A+nFoFmM+PJHnfGONPz2wYY/y3JN/IxkW4AXaNkAOY31anMmr2BrBrhBzAfD6c5JbZKdYk//c1c/uS/Om2jwLYAV4jBzCnqnp1kl9K8uwkf5fkK0n+xRjjTyZdGLDnCDmABVRVJbkzyf8cY/zy1OsB9iYhBwDQlNfIAQA0JeQAAJoScgAATQk5AICmhBwAQFNCDgCgKSEHANCUkAMAaOr/ANzZMHqux+yAAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "plt.hist2d(phi_psi.phi, phi_psi.psi, bins=(20, 20), cmap=plt.cm.Greys);\n",
    "plt.xlabel('$\\Phi$')\n",
    "plt.ylabel('$\\Psi$')\n",
    "plt.xlim([-180, 180])\n",
    "plt.ylim([-180, 180])\n",
    "plt.grid()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
